---
title: "Run on summary statistics data"
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```
# Purpose of the analysis

This workflow aims to use GWAS summary statistics data as input and implements the causal TWAS algorithm.  The "causal TWAS" method is developed to identify causal genes in GWAS studies. We improve the current TWAS analysis mainly by reducing false positives. We start from z scores for each SNP in a GWAS study and expression model trained by the [Fusion/TWAS software](http://gusevlab.org/projects/fusion/). In the end, we will provide a PIP (the posterior inclusion probality for being a causal gene) for each gene with an expression model. 

Below is the workflow for running `ctwas` for summary statistics data. We provided example input files with the package and you can use the following code as a toy pipeline. You can prepare your own input files following the examples. 

# Prepare Input

Our package is called `ctwas`.

```{r input}
library(ctwas)
```

The example inputs are given as below:

* z scores, a data frame:

```{r} 
# z scores
data(z_snp)
head(z_snp)
```
Here, A1 is effect allele. A2 is the other allele. z indicates the z score from GWAS.

* LD reference data. 

There are two ways to provide LD information for the functions in ctwas. One is to provide LD genotype data (in plink format), another way is to provide genetic correlation matrix (R matrix) for LD indpendent regions. 

  ** If you want to provide LD genotype type, then you need to provide a character vector of .pgen or .bed files. One file for one chromosome, in the order of 1 to 22. Therefore, the length of this vector needs to be 22. If .pgen files are given, then .pvar and .psam are assumed to present in the same directory. If .bed files are given, then .bim and .fam files are assumed to present in the same directory. For example:
```{r}
# genotype for LD reference
ld_pgenfs <- system.file("extdata/example_genotype_files", paste0("example_chr", 1:22, ".pgen"), package = "ctwas")
ld_pgenfs
```

  ** If you want to provide R matrices for LD independent regions. You need to provide a character vector of LD R information files, one for each chromosome, in the order of 1 to 22. For example:

```{r}
ld_Rfs <- system.file("extdata/example_ld_R", paste0("example_chr", 1:22, "_ld_R_info.txt"), package = "ctwas")
ld_Rfs
```
Each row in the LD R information file provides information for a LD independent region.  The file has these columns: "chrom" (chromosome), "region_name" (a name for the LD independent region, can be anything), "start" (start position of the LD independent region), "stop" (stop or end position of the LD independent region), "RDS_file" (the file path for the .RDS file containing the LD R matrix, more description below).

```{r}
head(read.table(ld_Rfs[1], header = T))
```
The "RDS_file", is a [R .RDS format](https://www.rdocumentation.org/packages/base/versions/3.3.2/topics/readRDS?tap_a=5644-dce66f&tap_s=10907-287229) file. It stores the LD correlation matrix for this region (a $p \times p$ matrix, $p$ is the number of SNPs in the region). We also require that for each "RDS_file"" we require that a file with the same stem name but ended with the suffix ".Rvar" to present in the same directory. This files gives SNPs information in the region and its order should match columns/rows in the .RDS file. An example, for region "1", the .RDS file is `/home/simingz/ctwas/inst/extdata/example_ld_R/example_chr1.R_snp.0_5.RDS`. the matching .Rvar file should be `/home/simingz/ctwas/inst/extdata/example_ld_R/example_chr1.R_snp.0_5.Rvar` and it looks like this:

```{r}
head(read.table("/home/simingz/ctwas/inst/extdata/example_ld_R/example_chr1.R_snp.0_5.Rvar", header = T))
```

Note: LD reference should contain as many GWAS SNPs as possible as only the overlapping SNPs are included in the analysis. Also only SNPs with z scores, in LD reference and are eQTLs (i.e. SNPs shared in GWAS, LD reference and weight files) are used in imputing gene expression z scores. Thus we suggest you to impute z scores for eQTLs if many are not available. The program will harmonize data internally. All coordinates are in hg19.

* Weight files. Please check out [Fusion/TWAS](http://gusevlab.org/projects/fusion/#computing-your-own-functional-weights) for the format of weights. Below is an example.
```{r}
# weight
weight <- system.file("extdata/example_fusion_weights", "Tissue", package = "ctwas")
```

* Regions file (Optional).

If you want to use your own regions file, please provide a file like this:
```{r regions}
regionsfile <- system.file("extdata", "example_regions.bed", package = "ctwas")
head(read.table(regionsfile, header = T))
```
The regions are left closed and right open, i.g. [start, stop). 

Otherwise, the package provided regions files generated by ldetect. You can specify the population that matches your GWAS data in `ctwas_rss` function. See below.

# Steps when using LD genotype data

## Step 1: impute expression z scores
In the following code, we compute expression z scores for each gene. We do this chromosome by chromosome. `impute_expr_z` will return z scores for each gene. It will also generate files containing the reference LD genotypes for eQTLs in the gene and the filename is returned. We will need these files in the `ctwas_rss` function in step 3.

When providing LD genotype data:
```{r impute, warning=F, message=F}
outputdir <- "~/temp"
ld_exprfs <- vector()
z_gene <- NULL
for (i in 1:22){
  ld_pgenf <- ld_pgenfs[i]
  res <- impute_expr_z(z_snp = z_snp, ld_pgenf = ld_pgenf, weight = weight,
                           method = "lasso", outputdir = outputdir,
                           outname = "test_ss")
  ld_exprfs[i] <- res$ld_exprf
  z_gene <- rbind(z_gene, res$z_gene)
}

head(z_gene)
ld_exprfs
```

## Step 2: run `ctwas_rss`. 
In this step we will perform the causal TWAS algorithm, the algorithm will run susie iteratively for parameter estimation and lastly provide PIPs for all genes and SNPs included in the analysis. If you don't want to define your own LD regions, then you can use the one defined by ldetect by simply specifying the population name using the `ld_regions` argument. One feature of the ctwas function is that it allows parallel computing. you can specify number of cores to use by the `ncore` argument. 

Run `ctwas_rss` using genotype data for LD:

```{r ctwas, warning=F, message=F}
pars <- ctwas_rss(z_gene = z_gene, z_snp = z_snp, ld_exprfs = ld_exprfs, ld_pgenfs = ld_pgenfs, ld_regions_custom = regionsfile, thin = 0.9, max_snp_region = 20, outputdir = outputdir, outname = "test_ss", ncore = 1)
```

# Steps when using LD correlation matrices

## Step 1: impute expression z scores
In the following code, we compute expression z scores for each gene. We do this chromosome by chromosome. `impute_expr_z` will return z scores for each gene. It will also generate files containing the reference LD genotypes for eQTLs in the gene and the filename is returned. We will need these files in the `ctwas_rss` function in step 3.

When providing LD R matrices:
```{r impute_R, warning=F, message=F}
outputdir <- "~/temp"
ld_exprfs <- vector()
z_gene <- NULL
for (i in 1:22){
  ld_Rf <- ld_Rfs[i]
  res <- impute_expr_z(z_snp = z_snp, ld_Rf = ld_Rf, weight = weight,
                           method = "lasso", outputdir = outputdir,
                           outname = "test_ss")
  ld_exprfs[i] <- res$ld_exprf
  z_gene <- rbind(z_gene, res$z_gene)
}

head(z_gene)
ld_exprfs
```

## Step 2: run `ctwas_rss`. 
In this step we will perform the causal TWAS algorithm, the algorithm will run susie iteratively for parameter estimation and lastly provide PIPs for all genes and SNPs included in the analysis. If you don't want to define your own LD regions, then you can use the one defined by ldetect by simply specifying the population name using the `ld_regions` argument. One feature of the ctwas function is that it allows parallel computing. you can specify number of cores to use by the `ncore` argument. 

When providing LD R matrices, use the following command:
```{r ctwas_R, warning=F, message=F}
pars <- ctwas_rss(z_gene = z_gene, z_snp = z_snp, ld_exprfs = ld_exprfs, ld_Rfs = ld_Rfs, ld_regions_custom = regionsfile, thin = 0.9, max_snp_region = 20, outputdir = outputdir, outname = "test_ss", ncore = 1)
```

# Output from ctwas
`ctwas_rss` returns the estimated parameters (in the order of gene, SNP):

```{r}
pars
```

PIP results are given in `outname.susieIrss.txt`. This file contains PIP for each gene and SNP, please check `susie_pip` column. 

```{r}
head(data.table::fread("~/temp/test_ss.susieIrss.txt"))[, c("chrom", "id", "pos", "type", "susie_pip")]
```

You will also notice a few auxilary/intermediate files produced by `ctwas_rss`. These files can be useful for diagnoiss.`.Rd` files contains parameters estimation updates with each iteration.

```{r}
sessionInfo()
```
