---
title: "Run on summary statistics data"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Run on summary statistics data}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```
# Requirements

LD reference should contain all SNPs given in z score data frame. The `id` column in z score data frame and LD reference SNP ID should have the same meaning (i.e., indicating the same SNP). Weights and LD refernce SNP ID should have same meaning and ref/alt allele info needs to match. All coordinates are in hg19.

# Prepare input

```{r input}
library(ctwas)

# z scores
data(zdf.snp)

# genotype for LD reference
ld_pgenfs <- system.file("extdata/example_genotype_files", paste0("example_chr", 1:22, ".pgen"), package = "ctwas")

# weight
weight <- system.file("extdata/example_fusion_weights", "Tissue", package = "ctwas")
```

# Impute expression z scores
```{r impute, warning=F, message=F}
outputdir <- "~"
ld_exprfs <- vector()
zdf.gene <- NULL
for (i in 1:22){
  ld_pgenf <- ld_pgenfs[i]
  res <- impute_expr_z(zdf.snp, ld_pgenf = ld_pgenf, weight = weight,
                           method = "lasso", outputdir = outputdir,
                           outname = "test_ss")
  ld_exprfs[i] <- res$ld_exprf
  zdf.gene <- rbind(zdf.gene, res$zdf)
}
zdf <- rbind(zdf.snp, zdf.gene)
```

# Run ctwas 
```{r ctwas, warning=F, message=F}
regionsfile <- system.file("extdata", "example_regions.bed", package = "ctwas")
ctwas_rss(zdf, ld_pgenfs, ld_exprfs, ld_regions_custom = regionsfile, thin = 1, outputdir = outputdir, outname = "test_ss", ncore = 1)
```
