---
title: "Run on individual-level data"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Run on individual-level data}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

# Requirements

Weights and LD refernce ID should have same meaning and ref/alt allele info needs to match. Y sample order and pgen sample order should match. All coordinates are in hg19.

# Prepare input

```{r setup}
library(ctwas)
getwd()
```

```{r input}
pgenfs <- system.file("extdata/example_genotype_files", paste0("example_chr", 1:22, ".pgen"), package = "ctwas")
weight <- system.file("extdata/example_fusion_weights", "Tissue", package = "ctwas")
```


# Impute expression
```{r impute}
outputdir <- "."
exprfs <- vector()
for (i in 1:22){
  pgenf <- pgenfs[i]
  exprfs[i] <- impute_expr(pgenf = pgenf, weight = weight,
                           method = "lasso", outputdir = outputdir,
                           outname = "test")
}
```

# Run ctwas 
```{r ctwas}
regionsfile <- system.file("extdata", "example_regions.bed", package = "ctwas")

X.g <- cbind(read_expr(exprfs[1]), read_expr(exprfs[2]))
X.g <- scale(X.g)

pgen <- prep_pgen(pgenfs[1], prep_pvar(pgenfs[1]))
X.s <- read_pgen(pgen, variantidx = 101:110)

X.s <- scale(X.s)
nsample <- nrow(X.g)

beta.gene <- c(0.5, 0.1)
beta.snp <- rep(0.2, 10)

Y <- X.g %*% beta.gene + X.s%*% beta.snp + rnorm(nsample, 0, 1)

ctwas(pgenfs, exprfs, Y, ld_regions_custom = regionsfile, thin = 1, outputdir = outputdir, outname = "test", ncore = 1)
```
