---
title: "Using multigroup cTWAS with summary statistics"
author: "Kaixuan Luo, Sheng Qian"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Using multigroup cTWAS with summary statistics}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(message = FALSE,
                      warning = FALSE,
                      fig.width = 6,
                      fig.height = 4,
                      fig.align = "center",
                      fig.cap = "&nbsp;",
                      dpi = 120)
```

This document shows the updated pipeline for running cTWAS with summary statistics using the `multi-group` version of the R package, 
which extends our earlier cTWAS method (single group version) to integrate multiple groups of prediction models, 
e.g. multiple tissues, cell types or conditions, or multiple molecular traits.

## Installation

Install the `multigroup` branch of `cTWAS` package

```{r install_package, eval=FALSE}
remotes::install_github("xinhe-lab/ctwas", ref = "multigroup_test")
```

Load the package

```{r load_package}
library(ctwas)
```

## Preparing input data

In this version of `cTWAS` package, we require first running the data preprocessing and harmonization steps 
before running parameter estimation and finemapping steps of cTWAS analysis. 

The inputs for the summary statistics version of cTWAS include GWAS summary statistics, 
prediction models in PredictDB format or FUSION format, and LD reference. 

### GWAS z-scores

We read the GWAS summary statistics as a data frame `z_snp`, with columns "id", "A1", "A2", "z", and each row is a variant. 
`A1` is the alternate allele, and `A2` is the reference allele. 

For this analysis, we will use summary statistics from a GWAS of LDL cholesterol in the UK Biobank. We will download the VCF from the IEU Open GWAS Project. 
```{r, eval=FALSE}
# set the working directory, download the summary statistics, and unzip the file.
dir.create("gwas_summary_stats")

system("wget https://gwas.mrcieu.ac.uk/files/ukb-d-30780_irnt/ukb-d-30780_irnt.vcf.gz -P gwas_summary_stats")
R.utils::gunzip("gwas_summary_stats/ukb-d-30780_irnt.vcf.gz")
```

Next, we will read the summary statistics. Then, we will compute the z-scores and format the input data. We will also collect the sample size, which will be useful later. We will save this output for convenience.

```{r, eval=FALSE}
# read the data using the VariantAnnotation package
z_snp <- VariantAnnotation::readVcf("gwas_summary_stats/ukb-d-30780_irnt.vcf")
z_snp <- as.data.frame(gwasvcf::vcf_to_tibble(z_snp))

# compute the z-scores
z_snp$Z <- z_snp$ES/z_snp$SE

# collect sample size (most frequent sample size for all variants)
gwas_n <- as.numeric(names(sort(table(z_snp$SS),decreasing=TRUE)[1]))
cat("gwas_n=", gwas_n, "\n")

# subset the columns and format the column names
z_snp <- z_snp[,c("rsid", "ALT", "REF", "Z")]
colnames(z_snp) <- c("id", "A1", "A2", "z")

# save the formatted z-scores and GWAS sample size
outputdir <- "/project2/xinhe/shared_data/multigroup_ctwas/LDL_multigroup_tutorial/"
dir.create(outputdir, showWarnings=F, recursive=T)
outname <- "example"

gwas_name <- "ukb-d-30780_irnt"
z_snp_outfile <- file.path(outputdir, paste0(gwas_name, ".z_snp.Rd"))
save(z_snp, gwas_n, file=z_snp_outfile)
```


### Prediction models

The `multigroup` version of cTWAS accepts multiple sets of prediction models in PredictDB or FUSION format. 

Please check [PredictDB](http://predictdb.org/) for the format of PredictDB weights. To specify weights in PredictDB format, provide the path to the `.db` file. 

Please check [Fusion/TWAS](http://gusevlab.org/projects/fusion/#compute-your-own-predictive-models) for the format of FUSION weights. 

### LD reference and region info

LD reference information can be provided as genetic correlation matrices (termed "R matrices") 
for regions that are approximately LD-independent. 

cTWAS performs its analysis region-by-region. The preferred way to run cTWAS is to provide pre-computed LD matrices for each region. 

*It is critical that the genome build (e.g. hg38) of the LD reference matches the genome build used to train the prediction models.*

The genome build of the GWAS summary statistics does not matter because variant positions are determined by the LD reference.

The choice of LD reference population is important for fine-mapping. Best practice for fine-mapping is to use in-sample LD reference (LD computed using the subjects in the GWAS sample). If in-sample LD reference is not an option, the LD reference should be as representative of the population in the GWAS sample as possible. Given that cTWAS is an extended fine-mapping algorithm, and that gene z-scores are computed using the observed GWAS z-scores, which reflect patterns of LD in the GWAS population, our recommendation is to match the LD reference to the GWAS population, not the population used to build the prediction models. 

#### Defining regions

cTWAS includes pre-defined regions based on European (EUR), Asian (ASN), or African (AFR) populations, 
using either genome build b38 or b37. 
These regions were previously generated using [LDetect](https://github.com/endrebak/ldetect). 

#### LD matrices

To use LD matrices for the LD reference, provide a directory containing all of the `.RDS` matrix files and matching `.Rvar` variant information files. 

We have precomputed reference LD matrices and the variant information tables accompanying the LD matrices for both UKB and 1000G European Phase3 references.
The complete LD matrices of European individuals from UK Biobank can be downloaded [here](https://uchicago.box.com/s/jqocacd2fulskmhoqnasrknbt59x3xkn). On the University of Chicago RCC cluster, the b38 reference is available at `/project2/mstephens/wcrouse/UKB_LDR_0.1/` and the b37 reference is available at `/project2/mstephens/wcrouse/UKB_LDR_0.1_b37/`.

#### Region info

In this version of the package, we also require a data frame `region_info` containing the following columns: 
"chrom", "start", "stop", "region_id", "LD_matrix", "SNP_info".

"chrom", "start", "stop" are the genomic coordinates of the regions, 
"region_id" contains the IDs of the region, and we use "chr:start-stop" format in region IDs by default.
"LD_matrix" stores the paths to the precomputed LD (R) matrices (`.RDS` files in our precomputed reference LD files),
"SNP_info" stores the paths to the variant information corresponding to the LD matrices (`.Rvar` files in our precomputed reference LD files). 

The `.RDS` file is [R .RDS format](https://www.rdocumentation.org/packages/base/versions/3.3.2/topics/readRDS?tap_a=5644-dce66f&tap_s=10907-287229). It stores the LD correlation matrix for a region (a $p \times p$ matrix, $p$ is the number of variants in the region). We require that for each `.RDS` file, in the same directory, there is a corresponding file with the same stem but ending with the suffix `.Rvar`. This `.Rvar` files includes variant information for the region, and the order of its rows must match the order of rows and columns in the `.RDS` file.

Here we use b38 European region file, which is included in the package.

```{r region_info, eval=FALSE}
genome_version <- "b38"
ld_R_dir <- "/project2/mstephens/wcrouse/UKB_LDR_0.1/"

region_file <- system.file("extdata/ldetect", paste0("EUR.", genome_version, ".bed"), package = "ctwas")
region_info <- read.table(region_file, header = TRUE)
colnames(region_info)[1:3] <- c("chrom", "start", "stop")
region_info$chrom <- as.numeric(gsub("chr", "", region_info$chrom))
region_info$region_id <- paste0(region_info$chr, ":", region_info$start, "-", region_info$stop)

filestem <- paste0("ukb_", genome_version, "_0.1")
ld_filestem <- sprintf("%s_chr%s.R_snp.%s_%s", filestem, region_info$chrom, region_info$start, region_info$stop)
region_info$LD_matrix <- file.path(ld_R_dir, paste0(ld_filestem, ".RDS"))
region_info$SNP_info <- file.path(ld_R_dir, paste0(ld_filestem, ".Rvar"))
# Check to make sure all the LD matrices and SNP info files are available.
stopifnot(all(file.exists(region_info$LD_matrix)))
stopifnot(all(file.exists(region_info$SNP_info)))
write.table(region_info, file = file.path(outputdir, "region_info.txt"), sep = "\t", col.names = T, row.names = F, quote = F)
```

The columns of the `.Rvar` file include information on chromosome, variant name, position in base pairs, and the alternative and reference alleles. The `variance` column is the variance of each variant prior to standardization; this is required for PredictDB weights but not FUSION weights. PredictDB weights should be scaled by the variance before imputing gene expression. This is because PredictDB weights assume that variant genotypes are not standardized before imputation, but our implementation assumes standardized variant genotypes. If variance information is missing, or if weights are in PredictDB format but are already on the standardized scale (e.g. if they were converted from FUSION to PredictDB format), this scaling can be turned off using the option `scale_by_ld_variance=FALSE` using the multigroup version of cTWAS. We've also include information on allele frequency in the variant info, but this is optional.

The naming convention for the LD matrices is `[filestem]_chr[#].R_snp.[start]_[end].RDS`. cTWAS expects that all `.RDS` and `.Rvar` files in the directory contain LD information, so no other files with these suffixes should be in the directory. Each variant should be uniquely assigned to a region, and the regions should be left closed and right open, i.e. [start, stop). The positions of the LD matrices must match exactly the positions specified by the region file. Do not include invariant or multiallelic variants in the LD reference. 

#### Convert reference genotype data to LD matrices

We could genotype data in PLINK format to LD matrices (.RDS) and variance information (.Rvar) files, using the `convert_geno_to_LD_matrix` function.
Below is an example of generating LD matrices using 1000G EUR Phase3 PLINK genotype data.

```{r convert_geno_to_LD_matrix, eval=FALSE}
# specify LD reference
ldref_dir <- "/project2/xinhe/shared_data/ldsc/sldsc/1000G_EUR_Phase3_plink"
# genotype files in PLINK .bed or .pgen format
genotype_files <- file.path(ldref_dir, paste0("1000G.EUR.QC.", 1:22, ".bed"))

# variance information files in PLINK .bim or .pvar format
# the output .Rvar files will use the positions and allele information in varinfo_files
varinfo_files <- file.path(ldref_dir, paste0("1000G.EUR.QC.", 1:22,".bim"))

# prepare a data frame region_info for LD regions with columns "chr", "start", and "stop",
# the positions should match those in snpinfo_files
genome_version = "b37"
region_file <- system.file("extdata/ldetect", paste0("EUR.", genome_version, ".bed"), package = "ctwas")
region_info <- read.table(region_file, header = T, stringsAsFactors = F)

# specify output
LD_outputdir <- "/project2/xinhe/shared_data/multigroup_ctwas/ref_LD_R/1000G_EUR_Phase3_b37/"
LD_outname <- "1000G_EUR_Phase3_b37"
region_info <- convert_geno_to_LD_matrix(region_info, genotype_files, varinfo_files, outputdir = LD_outputdir, outname = LD_outname)
```

### Data harmonization

These inputs should be harmonized prior to cTWAS analysis (i.e. the reference and alternative alleles for each variant 
should match across all three data sources). 

#### Harmonizing GWAS z-scores and LD reference

The `preprocess_z_snp()` function harmonizes GWAS z-scores and LD reference based on the included allele information.
If `drop_multiallelic = TRUE`, it will drop multiallelic variants. 
If `drop_strand_ambig = TRUE`, it will drop strand ambiguous variants.

```{r preprocess_z_snp, eval=FALSE}
# load the GWAS sumstats data
z_snp_outfile <- file.path(outputdir, paste0(gwas_name, ".z_snp.Rd"))
load(z_snp_outfile)

z_snp <- preprocess_z_snp(z_snp, region_info, gwas_n, drop_multiallelic = TRUE, drop_strand_ambig = TRUE)
```

#### Harmonizing prediction models and LD reference

The `preprocess_weight()` function harmonizes the PredictDB prediction models and LD reference. 
This only needs to be done once per combination of prediction models and LD reference. 

In this example, we use four weights, two eQTL models and two sQTLs models from liver and lung.
We preprocess each weight file separately and combine them in the end. 
We specify the `type` as the molecular QTL types, and `context` as the tissues, cell types, or conditions. 
For the expression models, we limit to protein coding genes, by setting `filter_protein_coding_genes = TRUE`.
We set `scale_by_ld_variance=TRUE` for PredictDB weights.

We use 6 cores to parallelize the computation over the weights.

```{r preprocess_weights, eval=FALSE}
weight_files <- c("/project/xinhe/shengqian/cTWAS_analysis/data/Liver_Expression.db",
                  "/project/xinhe/shengqian/cTWAS_analysis/data/Lung_Expression.db",
                  "/project/xinhe/shengqian/cTWAS_analysis/data/Liver_Splicing.db",
                  "/project/xinhe/shengqian/cTWAS_analysis/data/Lung_Splicing.db")
ncore <- 6

weights_liver_expression <- preprocess_weights(weight_file = weight_files[1],
                                               region_info,
                                               z_snp$id,
                                               type = "expression",
                                               context = "liver",
                                               weight_format = "PredictDB",
                                               ncore = ncore,
                                               drop_strand_ambig = TRUE,
                                               scale_by_ld_variance = TRUE,
                                               load_predictdb_LD = TRUE,
                                               filter_protein_coding_genes = TRUE)

weights_lung_expression <- preprocess_weights(weight_file = weight_files[2],
                                              region_info,
                                              z_snp$id,
                                              type = "expression",
                                              context = "lung",
                                              weight_format = "PredictDB",
                                              ncore = ncore,
                                              drop_strand_ambig = TRUE,
                                              scale_by_ld_variance = TRUE,
                                              load_predictdb_LD = TRUE,
                                              filter_protein_coding_genes = TRUE)

weights_liver_splicing <- preprocess_weights(weight_file = weight_files[3],
                                             region_info,
                                             z_snp$id,
                                             type = "splicing",
                                             context = "liver",
                                             weight_format = "PredictDB",
                                             ncore = ncore,
                                             drop_strand_ambig = TRUE,
                                             scale_by_ld_variance = TRUE,
                                             load_predictdb_LD = TRUE,
                                             filter_protein_coding_genes = FALSE)

weights_lung_splicing <- preprocess_weights(weight_file = weight_files[4],
                                            region_info,
                                            z_snp$id,
                                            type = "splicing",
                                            context = "lung",
                                            weight_format = "PredictDB",
                                            ncore = ncore,
                                            drop_strand_ambig = TRUE,
                                            scale_by_ld_variance = TRUE,
                                            load_predictdb_LD = TRUE,
                                            filter_protein_coding_genes = FALSE)

weights <- c(weights_liver_expression, weights_lung_expression, weights_liver_splicing, weights_lung_splicing)
```

## Running cTWAS analysis

We can run the entire process of cTWAS analysis from computing z-scores of molecular traits, 
estimating parameters to finemapping using the main `ctwas_sumstats` function.

```{r ctwas_sumstats, eval=FALSE}
res <- ctwas_sumstats(z_snp, weights, region_info, 
                      thin = 0.1,
                      niter1 = 3, niter2 = 30, 
                      L = 5, 
                      group_prior_var_structure = "independent", 
                      max_snp_region = 20000, 
                      min_nonSNP_PIP = 0.5,
                      ncore = ncore, 
                      save_cor = FALSE,
                      cor_dir = getwd(), 
                      logfile = NULL,
                      verbose = FALSE)

param <- res$param
finemap_res <- res$finemap_res
boundary_genes <- res$boundary_genes

```

We can also run each of the cTWAS modules separately. 
This would allow more flexible controls of the cTWAS modules with different settings 
(memory, parallelization, etc.). 
See [this tutorial][] for more details.

## Postprocessing

### Region merging

### Rerun finemapping with L = 1 for regions with problematic SNPs 

We will rerun the fine-mapping with L = 1 for the regions containing high PIP (>0.5) SNPs 
detected with LD mismatches, or high PIP genes with problematic SNPs in their weights. 

## cTWAS output

* Interpreting cTWAS output

* Summary of gene PIP

## Visualzing cTWAS results

We could make locus plot for regions of interest.


