---
title: "Simple cTWAS tutorial"
author: "Kaixuan Luo, Sheng Qian"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Simple cTWAS tutorial}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---



```{r setup, include=FALSE}
knitr::opts_chunk$set(message = FALSE,
                      warning = FALSE,
                      fig.align = "center",
                      fig.cap = "&nbsp;",
                      dpi = 120)
```



This document demonstrates how to run a simple analysis using cTWAS with GWAS summary statistics. Running cTWAS involves several steps: preparing input data, computing z-scores of molecular traits, assembling the input data for the regions, estimating parameters, screening regions with strong signals, and fine-mapping the variants and molecular traits. 

In this tutorial, we will show how to perform cTWAS main analysis with summary statistics without using LD matrices. 


Load the `ctwas` package.
```{r load_packages, message=FALSE}
library(ctwas)
```


We require first running the data preprocessing and harmonization steps before running cTWAS analysis. Please read the "Preparing cTWAS input data" tutorial to prepare cTWAS input data.

In this tutorial, we use sample data for chromosome 16 provided in the package. 

```{r load_z_snp}
# load z_snp sample data
z_snp <- readRDS(system.file("extdata/sample_data", "LDL_example.z_snp.RDS", package = "ctwas"))
```


We also need the GWAS sample size for summarizing the cTWAS result.
```{r gwas_n}
# GWAS sample size
gwas_n <- 343621
```


## Reference data

cTWAS assumes that the genome is partitioned into approximately independent LD regions. LD reference information can be provided as genetic correlation matrices (termed "R matrices") of these regions. 

Note: when running cTWAS without LD, you can skip "LD matrices". But you will still need to have a list of SNPs `snp_info` as a “reference”, with the information of these SNPs.  See the section "SNP info".

It is critical that the genome build (e.g. hg38) of the LD reference matches the genome build used to train the prediction models. The genome build of the GWAS summary statistics does not matter because variant positions are determined by the LD reference.

The choice of LD reference population is important for fine-mapping. Best practice for fine-mapping is to use an in-sample LD reference (LD computed using the subjects in the GWAS sample). If in-sample LD reference is not an option, the LD reference should be as representative of the population in the GWAS sample as possible.

### Region info and SNP info

We require a data frame `region_info` containing the region definitions, with the following columns: "chrom", "start", "stop", for the genomic coordinates of the regions, and "region_id" for the IDs of the regions (by default, we use [chrom_start_stop] as region IDs).

Here we use the b38 European LDetect blocks, included in the package.

```{r region_info}
region_file <- system.file("extdata/ldetect", "EUR.b38.bed", package = "ctwas")
region_info <- read.table(region_file, header = TRUE)
colnames(region_info)[1:3] <- c("chrom", "start", "stop")
region_info$chrom <- as.numeric(gsub("chr", "", region_info$chrom))
region_info$region_id <- paste0(region_info$chr, "_", region_info$start, "_", region_info$stop)
```



### LD and SNP info

We need a list of variants as a “reference”, with the positions and allele information of these variants. We provide a function, `preprocess_region_LD_snp_info()` to preprocess region definitions and map variants from the LD reference to regions, which takes input of `region_info` and a data frame `ref_snp_info` with variant information from the entire genome, and returns processed `region_info`, `snp_info`.

We have the lists of reference variant information from all the LD matrices in the genome in [hg38](https://uchicago.box.com/s/t089or92dkovv0epkrjvxq8r9db9ys99) and [hg19](https://uchicago.box.com/s/ufko2gjagcb693dob4khccqubuztb9pz).

In this example, we use variant information for chr16 in hg38.

```{r snp_info}
example_chrom <- 16
ref_snp_info_file <- system.file("extdata/sample_data", "ukb_b38_0.1_chr16_var_info.Rvar.gz", package = "ctwas")
ref_snp_info <- data.table::fread(ref_snp_info_file, sep = "\t")

res <- preprocess_region_LD_snp_info(region_info, 
                                     ref_snp_info = ref_snp_info, 
                                     chrom = example_chrom, 
                                     use_LD = FALSE)
region_info <- res$region_info
snp_info <- res$snp_info

head(region_info)
```


Note: in this simple tutorial, we use the "no-LD" version, so we use `use_LD = FALSE`. 
Please see the other tutorials for running full cTWAS analysis using LD. 

### Harmonizing GWAS z-scores and the reference data

The `preprocess_z_snp()` function harmonizes GWAS z-scores and the reference data in `snp_info`. We drop multiallelic variants and drop strand ambiguous variants, by default.

```{r preprocess_z_snp}
# preprocess z_snp
z_snp <- preprocess_z_snp(z_snp, 
                          snp_info, 
                          drop_multiallelic = TRUE, 
                          drop_strand_ambig = TRUE)
```


### Harmonizing prediction models and the reference data

The `preprocess_weight()` function harmonizes the PredictDB/FUSION prediction models and LD reference. 

In this version of cTWAS, we allow the joint analysis of multiple groups of molecular traits. This could be: eQTL of multiple tissues; or eQTLs, splicing QTLs and other types of QTLs in a single tissue. In a more complex setting, multiple types of QTL data from multiple tissues/cell types. Each group is defined by its “type” (kind of molecular traits), and “context” (tissue, cell type, condition, etc.).  So, we specify the `type` and `context` arguments for each weight file. 

In this example, we use liver and subcutaneous adipose gene expression models. We specify the `type` and `context` arguments for each weight file. 
We preprocess each of the weight files, and then simply concatenate them to get a list of all processed weights from different types or contexts.

```{r preprocess_weights}
# preprocess weights
weight_liver_file <- system.file("extdata/sample_data", "expression_Liver.db", package = "ctwas")
weights_liver <- preprocess_weights(weight_liver_file,
                                    region_info,
                                    z_snp$id,
                                    snp_info,
                                    type = "expression",
                                    context = "liver",
                                    weight_format = "PredictDB",
                                    drop_strand_ambig = TRUE,
                                    scale_predictdb_weights = TRUE,
                                    load_predictdb_LD = TRUE,
                                    filter_protein_coding_genes = TRUE)

weight_adipose_file <- system.file("extdata/sample_data", "expression_Adipose_Subcutaneous.db", package = "ctwas")
weights_adipose <- preprocess_weights(weight_adipose_file,
                                      region_info,
                                      z_snp$id,
                                      snp_info,
                                      type = "expression",
                                      context = "adipose",
                                      weight_format = "PredictDB",
                                      drop_strand_ambig = TRUE,
                                      scale_predictdb_weights = TRUE,
                                      load_predictdb_LD = TRUE,
                                      filter_protein_coding_genes = TRUE)

# concatenate weights from different types or contexts
weights <- c(weights_liver, weights_adipose)
```



## Running cTWAS analysis without LD

We use the function `ctwas_sumstats_noLD()` to run cTWAS analysis without LD. It takes preprocessed GWAS z-scores (`z_snp`), `weights`, `region_info`, `snp_info` as input, and will perform the main steps: computing z_scores of molecular traits, assembling input `region_data`, estimating parameters, screening regions, and finemapping. 

If `z_gene` is already computed when preparing cTWAS input data, we could specify the `z_gene` argument in the function, then it will skip computing z-scores of molecular traits.

```{r ctwas_sumstats_noLD, eval=FALSE}
ctwas_res <- ctwas_sumstats_noLD(z_snp,
                                 weights,
                                 region_info,
                                 snp_info,
                                 thin = 0.1,
                                 niter_prefit = 3,
                                 niter = 30,
                                 group_prior_var_structure = "shared_type",
                                 maxSNP = 20000,
                                 min_nonSNP_PIP = 0.5,
                                 ncore = 6)

# saveRDS(ctwas_res, file.path(outputdir, paste0(outname, ".ctwas_sumstats_noLD_res.RDS")))
```


```{r, include=FALSE}
ctwas_res <- readRDS(system.file("extdata/sample_data", "LDL_example.ctwas_sumstats_noLD_res.RDS", package = "ctwas"))
```


```{r}
param <- ctwas_res$param
finemap_res <- ctwas_res$finemap_res
boundary_genes <- ctwas_res$boundary_genes
z_gene <- ctwas_res$z_gene
region_data <- ctwas_res$region_data
screened_region_data <- ctwas_res$screened_region_data
```


Major argument settings:

-   `z_snp` is a data frame with preprocessed GWAS z-scores.
-   `weights` is a list of preprocessed weights from PredictDB or FUSION prediction models.
-   `region_info` is a data frame with genome coordinates, LD matrix and SNP info of all the regions
-   `snp_info` is a list of reference SNP info in each of the regions.
-   The `thin` argument randomly selects a subset of variants (10% when thin = 0.1) to use during the parameter estimation and screening regions, reducing computation.
-   The `niter_prefit` argument and `niter` argument sets the number of EM iterations for estimating parameters. We set `niter_prefit = 3`, and `niter = 30` by default.
-   There are several options for how to handle the prior effect size parameters by specifying the `group_prior_var_stucture`. "shared_type" (default option) allows all groups in one molecular QTL type to share the same variance parameter. "shared_context" allows all groups in one context (tissue, cell type, condition) to share the same variance parameter. "shared_nonSNP" allows all non-SNP groups to share the same variance parameter. "shared_all" allows all groups to share the same variance parameter. "independent" allows all groups to have their own separate variance parameters.
-   The `maxSNP` sets a maximum on the number of variants that can be in a single region to prevent memory issues during fine-mapping.
-   `min_nonSNP_PIP` is a threshold for selecting regions with strong non-SNP signals for finemapping. We set `min_nonSNP_PIP = 0.5` by default, so regions with total non-SNP signals greater than 0.5 will be selected for finemapping.
-   The `ncore` argument specifies the number of cores to use when parallelizing over regions.

The function returns a list including: 
estimated parameters, fine-mapping results, boundary genes, z-scores of molecular traits, assembled region data for all the regions and region data with full SNPs for selected regions.

## Summarizing cTWAS result

### Assessing parameters and computing PVE

We could use the function `summarize_param()` to assess the convergence of the estimated parameters and to compute the proportion of variance explained (PVE) by variants and genes. 
```{r summarize_param}
ctwas_parameters <- summarize_param(param, gwas_n)
```


```{r}
# Estimated prior inclusion probability
ctwas_parameters$group_prior

# Estimated prior effect size:
ctwas_parameters$group_prior_var

# Estimated enrichment of molecular traits over variants:
ctwas_parameters$enrichment

# PVE explained by molecular traits and variants:
ctwas_parameters$group_pve

# Total heritability (sum of PVE)
ctwas_parameters$total_pve

# Attributable heritability
ctwas_parameters$attributable_pve
```

Make convergence plots of estimated parameters
```{r convergence_plot, fig.width = 7, fig.height=5}
make_convergence_plots(param, gwas_n)
```


## Interpreting cTWAS results

To interpret and visualize the results, we first need to add gene annotations 
(gene names and gene types) to the finemapping result.

We need a user defined data frame `gene_annot` that should contain the following columns:
'chrom', 'start', 'end', 'gene_id', 'gene_name' and 'gene_type' for each gene or molecular trait.

If you only used gene expression data, you can use the following function 
to obtain the gene annotations from the Ensembl database.
We use `EnsDb.Hsapiens.v86` for the example data, please choose the Ensembl database for your specific data. 

```{r get_gene_annot, message=FALSE}
library(EnsDb.Hsapiens.v86)
ens_db <- EnsDb.Hsapiens.v86
finemap_gene_res <- finemap_res[finemap_res$type!="SNP",]
gene_ids <- unique(sapply(strsplit(finemap_gene_res$id, split = "[|]"), "[[", 1))
gene_annot <- get_gene_annot_from_ens_db(ens_db, gene_ids)
```


We then use `anno_finemap_res()` function to add gene annotations to the finemapping result,
and use gene mid-points to represent gene positions.
```{r anno_finemap_res}
# add gene annotations and use gene mid-points to represent gene positions
finemap_res <- anno_finemap_res(finemap_res, 
                                snp_info,
                                gene_annot,
                                use_gene_pos = "mid")
```


We list all molecular traits with PIP > 0.8, which is the threshold we used in the paper.
```{r interpret_res}
# select PIP > 0.8
sig_finemap_res <- finemap_res[finemap_res$type!="SNP" & finemap_res$susie_pip > 0.8,]

head(sig_finemap_res[order(-sig_finemap_res$susie_pip),])
```


When we have multiple contexts (tissues), 
it is useful to evaluate the combined PIP for each molecular trait across contexts:

```{r combine_pips}
# combine gene PIPs by context
combined_pip_by_context <- combine_gene_pips(finemap_res, 
                                             by = "context",
                                             filter_protein_coding_genes = TRUE)
head(combined_pip_by_context)

# select combined PIP > 0.8
sig_gene_pips_df <- combined_pip_by_context[combined_pip_by_context$combined_pip > 0.8, ]
head(sig_gene_pips_df)
```


## Visualizing cTWAS results

We could make locus plots for regions of interest. 
We could zoom in the region by specifying the `locus_range`.

```{r locus_plot, fig.width=10, fig.height=8}
make_locusplot(finemap_res,
               region_id = "16_71020125_72901251",
               weights = weights,
               ens_db = ens_db,
               locus_range = c(71.6e6,72.4e6),
               highlight_pip = 0.8,
               legend.position = "top")
```




