---
title: "Summarizing cTWAS results"
author: "Kaixuan Luo, Sheng Qian"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Summarizing cTWAS results}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(message = FALSE,
                      warning = FALSE,
                      fig.width = 6,
                      fig.height = 4,
                      fig.align = "center",
                      fig.cap = "&nbsp;",
                      dpi = 120)
```


In this document, we will show how to summarize and visualize cTWAS results.

Load the `ctwas` package. We also need packages `tidyverse`, `locuszoomr` for summarizing and visualizing cTWAS result
```{r load_packages, message=FALSE}
library(ctwas)
library(tidyverse)
library(locuszoomr)
```


We also need gene annotations from the Ensembl database. We use `EnsDb.Hsapiens.v86` for the example data, please choose the Ensembl database for your specific data. 
```{r load_ens_db, message=FALSE}
library(EnsDb.Hsapiens.v86)
```


Load the cTWAS results
```{r load_ctwas_res}
# GWAS sample size
gwas_n <- 343621
# Load weights
weights <- readRDS(system.file("extdata/sample_data", "LDL_example.preprocessed.weights.RDS", package = "ctwas"))

# Load cTWAS results
ctwas_res <- readRDS(system.file("extdata/sample_data", "LDL_example.ctwas_sumstats_res.RDS", package = "ctwas"))
param <- ctwas_res$param
finemap_res <- ctwas_res$finemap_res
boundary_genes <- ctwas_res$boundary_genes
z_gene <- ctwas_res$z_gene
region_data <- ctwas_res$region_data
screened_region_data <- ctwas_res$screened_region_data

# directory with computed correlation matrices
cor_dir <- system.file("extdata/sample_data", "cor_matrix", package = "ctwas")
```


### Assessing parameters and computing PVE

We could use the `summarize_param()` function to assess the convergence of the estimated parameters and 
to compute the proportion of variance explained (PVE) by variants and genes. 
```{r summarize_param}
ctwas_parameters <- summarize_param(param, gwas_n)
```


```{r}
# Estimated prior inclusion probability
ctwas_parameters$group_prior

# Estimated prior effect size:
ctwas_parameters$group_prior_var

# Estimated enrichment of molecular traits over variants:
ctwas_parameters$enrichment

# PVE explained by molecular traits and variants:
ctwas_parameters$group_pve

# Total heritability (sum of PVE)
ctwas_parameters$total_pve

# Attributable heritability
ctwas_parameters$attributable_pve
```


Plot convergence
```{r convergence_plot, fig.width = 7, fig.height=5}
ctwas_parameters$convergence_plot
```



## Summarizing cTWAS results

To interpret and visualize the results, we need to add gene names to the finemapping result, and use mid-points to represent gene positions.
```{r add_gene_names}
# load gene info database
ens_db <- EnsDb.Hsapiens.v86

# add gene names and use gene mid-points to represent gene positions
finemap_res <- anno_finemap_res(finemap_res, snp_info, ens_db, use_gene_pos = "mid")
```


## Interpreting cTWAS results

We list all molecular traits with PIP > 0.8, which is the threshold we used in the paper.
```{r interpret_res}
# select PIP > 0.8
sig_finemap_res <- finemap_res[finemap_res$type!="SNP" & finemap_res$susie_pip > 0.8,]
head(sig_finemap_res[order(-sig_finemap_res$susie_pip),])
```


When we have multiple contexts (tissues, conditions), 
it is useful to evaluate the combined PIP for each molecular trait over different contexts:


```{r combine_pips}
# combine PIPs across contexts (tissues)
combined_gene_pips <- sum_pip_across_contexts(finemap_res)

# select combined PIP > 0.8
sig_gene_pips_df <- combined_gene_pips[combined_gene_pips$combined_pip > 0.8, ]

# list genes with combined PIP > 0.8 or top 20 genes with highest combined PIPs
head(combined_gene_pips, max(nrow(sig_gene_pips_df), 20))
```



## Visualizing cTWAS results

We could make locus plots for regions of interest. We could zoom in the region by specifying the `locus_range`.

```{r locus_plot, fig.width=10, fig.height=8}
make_locusplot(finemap_res,
               region_id = "16_71020125_72901251",
               weights = weights,
               ens_db = ens_db, 
               use_LD = TRUE,
               cor_dir = cor_dir,
               pip_thresh = 0.8,
               locus_range = c(71.6e6,72.4e6),
               gene_biotype="protein_coding",
               legend.position = "right")
```


If we do not have precomputed correlation matrices, we could compute those by specifying `region_data`, `LD_info` and `snp_info`. Or we could skip coloring correlations by setting `use_LD = FALSE`. 

