---
title: "Summarizing and visualizing cTWAS results"
author: "Kaixuan Luo, Sheng Qian"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Summarizing and visualizing cTWAS results}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(message = FALSE,
                      warning = FALSE,
                      fig.width = 6,
                      fig.height = 4,
                      fig.align = "center",
                      fig.cap = "&nbsp;",
                      dpi = 120)
```


In this tutorial, we will show how to summarize and visualize cTWAS results.

Load the packages.
```{r load_packages, message=FALSE}
library(ctwas)
library(EnsDb.Hsapiens.v86)
```


We need the Ensembl database `EnsDb.Hsapiens.v86` (for hg38) in this tutorial, please choose the version of Ensembl database for your specific data (e.g. `EnsDb.Hsapiens.v75` for hg19). 

Let's first load the cTWAS input data needed to run this tutorial: 
GWAS sample size (`gwas_n`), prediction models of molecular traits (`weights`), the reference data: including information about regions (`region_info`), reference variant list (`snp_map`) and reference LD (`LD_map`).

```{r load_ctwas_input}
gwas_n <- 343621

weights <- readRDS(system.file("extdata/sample_data", "LDL_example.preprocessed.weights.RDS", package = "ctwas"))

region_info <- readRDS(system.file("extdata/sample_data", "LDL_example.region_info.RDS", package = "ctwas"))

snp_map <- readRDS(system.file("extdata/sample_data", "LDL_example.snp_map.RDS", package = "ctwas"))

LD_map <- readRDS(system.file("extdata/sample_data", "LDL_example.LD_map.RDS", package = "ctwas"))
```


We then load the cTWAS result for the sample data (chr16). 
Note, to better show the examples, here we use cTWAS results obtained using parameters estimated from the *entire genome*. 

```{r load_ctwas_res}
ctwas_res <- readRDS(system.file("extdata/sample_data", "LDL_example.ctwas_sumstats_res_param_allchrs.RDS", package = "ctwas"))
z_gene <- ctwas_res$z_gene
param <- ctwas_res$param
finemap_res <- ctwas_res$finemap_res
boundary_genes <- ctwas_res$boundary_genes
region_data <- ctwas_res$region_data
screen_res <- ctwas_res$screen_res
```


`ctwas_sumstats()` returns several main output: 

+ `z_gene`: the Z-scores of molecular traits, 

+ `param`: the estimated parameters,
 
+ `finemap_res`: fine-mapping results, a data frame of genes and variants, including their IDs, the regions they belong to (`region_id`), Z-scores (`z`), PIPs (`susie_pip`), credible set indices (`cs_index`), and estimated effect size variance (`mu2`).

```{r show_finemap_res}
head(finemap_res)
```


It also produces other output: 

+ `boundary_genes`: the genes/molecular traits that cross region boundaries (we will need these in the post-processing step). 

+ `region_data`: assembled region data, including Z scores of SNPs and molecular traits, for all the regions. Note that only data of the subset of SNPs used in screening regions are included. 

+ `screen_res`: screening regions results, including the data of all SNPs and molecular traits of the selected regions, estimated numbers of causal signals (`L`) for selected regions, and a data frame with estimated `L` and non-SNP PIPs for all regions.

We focus on examples from running cTWAS with LD here, but the steps below also work for running without LD.

## Assessing parameters and computing PVE

First, we make plots using the function `make_convergence_plots()` to see how estimated parameters converge during the execution of the program:

```{r convergence_plot, fig.width = 10, fig.height=5}
make_convergence_plots(param, gwas_n)
```


These plots show the estimated prior inclusion probability, prior effect size variance, and enrichment over the iterations of parameter estimation. The enrichment is defined as the ratio of the prior of molecular traits over the prior of variants. We generally expect molecular traits to have higher prior inclusion probability than variants. Enrichment values typically range from 20 - 100 for expresion traits. 

Then, we use `summarize_param()` to obtain estimated parameters (from the last iteration) and to compute the proportion of variance explained (PVE) by variants and molecular traits. These PVE values allow us to compute how heritability of the phenotype is partitioned among variants and groups of molecular traits. 

```{r summarize_param}
ctwas_parameters <- summarize_param(param, gwas_n)
ctwas_parameters
```


From our experience, genes typically explain around 5-15% of the total heritability of the trait. So if your number is much higher or lower, it may suggest some problems.

## Inspecting the cTWAS results

We can view the molecular traits above a certain PIP cutoff, e.g. PIP > 0.8. We also generally recommend limiting the results to credible sets. 

```{r interpret_res}
subset(finemap_res, group != "SNP" & susie_pip > 0.8 & cs_index != 0)
```


To help interpret the results, we could add gene annotations (e.g. gene names, gene types, etc.) to the fine-mapping result. 

To map between molecular traits and their corresponding genes, we need `mapping_table`, a  data frame containing IDs of molecular traits (`gene_id`), and the corresponding gene names (`gene_name`). Other gene annotation information, such as gene types (`gene_type`), and position information of the genes (`chrom`, `start`, `end`) could also be included and would be useful later when visualizing the results. 

We have a function `get_gene_annot_from_ens_db()` , which creates gene annotations from the Ensembl database.   
```{r get_gene_annot_from_ens_db, message=FALSE}
ens_db <- EnsDb.Hsapiens.v86
finemap_gene_res <- subset(finemap_res, group != "SNP")
gene_ids <- unique(sapply(strsplit(finemap_gene_res$id, split = "[|]"), "[[", 1))
gene_annot <- get_gene_annot_from_ens_db(ens_db, gene_ids)
mapping_table <- gene_annot
head(mapping_table)
```


Note: we used the gene annotations from `EnsDb.Hsapiens.v86` for the example data in hg38. You can choose the Ensembl database for the genome build of your own data (e.g. `EnsDb.Hsapiens.v75` for hg19)

For gene expression models, we could use the gene annotations as `mapping_table` to map gene IDs to corresponding gene names. If you have other molecular traits such as splicing, you would need to prepare your own `mapping_table` in similar format.

We then use the `anno_finemap_res()` function to map between molecular traits and their corresponding genes and add gene annotations to the fine-mapping result. If a molecular trait is mapped to multiple genes, it splits the PIP of the molecular trait to the target genes. 
```{r add_gene_names}
annotated_finemap_res <- anno_finemap_res(finemap_res,
                                          mapping_table,
                                          map_by = "gene_id",
                                          drop_unmapped_genes = TRUE)
```


We drop unannotated genes by setting `drop_unannotated_genes = TRUE` (default). 

We could limit to protein coding genes by setting `filter_protein_coding_genes = TRUE`, 
and limit genes to credible sets by setting `filter_cs_genes = TRUE`.

```{r filter_finemap_res}
annotated_finemap_res <- filter_finemap_res(annotated_finemap_res, 
                                            filter_protein_coding_genes = TRUE, 
                                            filter_cs_genes = TRUE)
```


We can now run the command at the beginning of this section to view the genes with additional information.

```{r interpret_res_2}
subset(annotated_finemap_res, group != "SNP" & susie_pip > 0.8 & cs_index != 0)
```


In this example, HPR gene in liver, HP gene in adipose, and CETP gene in adipose are the only three genes that satisfy this threshold.

When we have multiple contexts or different types of molecular traits, it is useful to evaluate the total evidence of a gene being causal, *combining evidence of all the molecular traits affecting this gene across all types and contexts.* We compute combined PIPs of all these molecular traits of the same gene, using the following formula: $1 - \prod_k (1 - \text{PIP}_k)$, where $\text{PIP}_k$ is the PIP of the $k$-th molecular trait of a gene.
 
For example, we can combine PIPs of the same gene by context:
 
```{r combine_pips_by_context}
combined_pip_by_context <- combine_gene_pips(annotated_finemap_res, 
                                             by = "context")
subset(combined_pip_by_context, combined_pip > 0.8)
```


In this example, we can see the causal signal of HPR is specific in liver,  and the causal signals of HP and CETP genes are specific in adipose. 

Similarly, we can also combine PIPs of the same gene by types:  

```{r combine_pips_by_type}
combined_pip_by_type <- combine_gene_pips(annotated_finemap_res, 
                                          by = "type")
subset(combined_pip_by_type, combined_pip > 0.8)
```


We can also combine PIPs of the same gene by groups, which includes all the combinations of contexts and types.  

```{r combine_pips_by_group}
combined_pip_by_group <- combine_gene_pips(annotated_finemap_res, 
                                           by = "group")
subset(combined_pip_by_group, combined_pip > 0.8)
```



## Visualizing the cTWAS results

It is often useful to visualize the results of cTWAS. This would help one understand the rationale of how cTWAS chooses a particular molecular trait and identify issues when cTWAS doesnâ€™t behave properly. 

Before we make the plots,  we use the function `add_pos_to_finemap_res()` to add position information for the genes and SNPs to the fine-mapping result. Here we use the midpoint position to represent genes (`use_gene_pos = "mid"`). The function needs annotated fine-mapping results, the list of reference SNPs (`snp_map`) which contains IDs and positions of SNPs, and the gene annotations (`gene_annot`) which contain the names and positions of genes. 
```{r add_pos}
annotated_finemap_res <- add_pos_to_finemap_res(annotated_finemap_res,
                                                snp_map,
                                                gene_annot,
                                                use_gene_pos = "mid")
```


The plotting function could color the molecular traits and SNPs based on the correlations among them. If we have saved correlation matrices in `cor_dir` when running cTWAS with LD, we could simply load those precomputed correlation matrices of the region using function `load_region_cor()`, and then make the locus plot. 

```{r load_region_cor_res}
region_id <- "16_71020125_72901251"
cor_dir <- system.file("extdata/sample_data", "cor_matrix", package = "ctwas")
cor_res <- load_region_cor(region_id, cor_dir = cor_dir)
```


If we don't have precomputed correlation matrices, we could compute the correlation matrices of the region using the function `get_region_cor()`. We will need the region data with all SNPs (`screened_region_data`), `LD_map` and `weights`.

```{r compute_region_cor_res, eval=FALSE}
screened_region_data <- screen_res$screened_region_data
region_id <- "16_71020125_72901251"

cor_res <- get_region_cor(region_id,
                          sids = screened_region_data[[region_id]]$sid,
                          gids = screened_region_data[[region_id]]$gid,
                          LD_map = LD_map,
                          weights = weights)
```


We could now use the `make_locusplot()` function to make locus plots for regions of interest. We need the annotated fine-mapping result with position information, region ID, preprocessed weights (to plot QTL locations), and Ensembl gene annotation database (to plot the gene track). 

```{r locus_plot, fig.width=10, fig.height=8}
make_locusplot(annotated_finemap_res,
               region_id = "16_71020125_72901251",
               weights = weights,
               ens_db = ens_db, 
               locus_range = c(71.6e6,72.4e6),
               highlight_pip = 0.8,
               R_snp_gene = cor_res$R_snp_gene,
               R_gene = cor_res$R_gene)
```


The locus plot shows several tracks in the example region in chromosome 6. We could zoom in a region by specifying the `locus_range` (71.6 - 72.4 Mb). The top one shows -log10(p-value) of the association of variants and molecular traits with the phenotype. The next track shows the PIPs of variants and molecular traits.  The data points are colored by the correlations with the focal gene (the gene with the highest PIP, HPR in this case). We can draw a red dotted line to show the PIP cutoff of 0.8 by setting `highlight_pip = 0.8`. The next track shows the locations of QTLs of the focal gene (by default, the focal gene is the gene with the highest PIP). The bottom is the gene track.

In this plot, one can see that there are several genes with good associations with the phenotype, but only HPR in liver and HP in adipose are selected by cTWAS. This is because once HPR, HP and other likely causal variants are selected by cTWAS, there is no remaining association left in other molecular traits. 

For the "no-LD" version, we could still make the locus plot. The difference is that the plot will not color the data points based on the correlations to the focal gene.

```{r locus_plot_noLD, fig.width=10, fig.height=8}
make_locusplot(annotated_finemap_res,
               region_id = "16_71020125_72901251",
               weights = weights,
               ens_db = ens_db,
               locus_range = c(71.6e6,72.4e6),
               highlight_pip = 0.8)
```


If you get an error message `Viewport has zero dimension(s)` when making locus plots, it usually suggests the R viewer window is too small. You can try increasing the size of the viewer window size and try again. 
