---
title: "Summarizing cTWAS results"
author: "Kaixuan Luo, Sheng Qian"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Summarizing cTWAS results}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(message = FALSE,
                      warning = FALSE,
                      fig.width = 6,
                      fig.height = 4,
                      fig.align = "center",
                      fig.cap = "&nbsp;",
                      dpi = 120)
```


In this document, we will show how to summarize and visualize cTWAS results.

Load the packages.
```{r load_packages, message=FALSE}
library(ctwas)
```

Load the cTWAS results
```{r load_ctwas_res}
# GWAS sample size
gwas_n <- 343621

# Load weights
weights <- readRDS(system.file("extdata/sample_data", "LDL_example.preprocessed.weights.RDS", package = "ctwas"))

# Load region_info, snp_info and LD_info
region_info <- readRDS(system.file("extdata/sample_data", "LDL_example.region_info.RDS", package = "ctwas"))
snp_info <- readRDS(system.file("extdata/sample_data", "LDL_example.snp_info.RDS", package = "ctwas"))
LD_info <- readRDS(system.file("extdata/sample_data", "LDL_example.LD_info.RDS", package = "ctwas"))

# Load cTWAS results
ctwas_res <- readRDS(system.file("extdata/sample_data", "LDL_example.ctwas_sumstats_res.RDS", package = "ctwas"))
param <- ctwas_res$param
finemap_res <- ctwas_res$finemap_res
boundary_genes <- ctwas_res$boundary_genes
z_gene <- ctwas_res$z_gene
region_data <- ctwas_res$region_data
screened_region_data <- ctwas_res$screened_region_data

# directory with computed correlation matrices
cor_dir <- system.file("extdata/sample_data", "cor_matrix", package = "ctwas")
```


### Assessing parameters and computing PVE

We could use the `summarize_param()` function to assess the convergence of the estimated parameters and 
to compute the proportion of variance explained (PVE) by variants and genes. 
```{r summarize_param}
ctwas_parameters <- summarize_param(param, gwas_n)
```


```{r}
# Estimated prior inclusion probability
ctwas_parameters$group_prior

# Estimated prior effect size:
ctwas_parameters$group_prior_var

# Estimated enrichment of molecular traits over variants:
ctwas_parameters$enrichment

# PVE explained by molecular traits and variants:
ctwas_parameters$group_pve

# Total heritability (sum of PVE)
ctwas_parameters$total_pve

# Attributable heritability
ctwas_parameters$attributable_pve
```


Plot convergence
```{r convergence_plot, fig.width = 7, fig.height=5}
ctwas_parameters$convergence_plot
```



## Summarizing cTWAS results

To interpret and visualize the results, we first need to add gene annotations 
(gene names and gene types) to the finemapping result.

We need a user defined data frame of `gene_annot` that contains the following columns:
'chrom', 'start', 'end', 'gene_id', 'gene_name' and 'gene_type' for each gene or molecular trait.

If you only used gene expression data, you can use the following function 
to obtain the gene annotations from the Ensembl database.

We use `EnsDb.Hsapiens.v86` for the example data, 
please choose the Ensembl database for your specific data. 
```{r get_gene_annot, message=FALSE}
library(EnsDb.Hsapiens.v86)
ens_db <- EnsDb.Hsapiens.v86
finemap_gene_res <- finemap_res[finemap_res$type!="SNP",]
gene_ids <- unique(sapply(strsplit(finemap_gene_res$id, split = "[|]"), "[[", 1))
gene_annot <- get_gene_annot_from_ens_db(ens_db, gene_ids)
```

We then use `anno_finemap_res()` function to add gene annotations to the finemapping result,
and use gene mid-points to represent gene positions.
```{r anno_finemap_res}
# add gene annotations and use gene mid-points to represent gene positions
finemap_res <- anno_finemap_res(finemap_res, 
                                snp_info,
                                gene_annot,
                                use_gene_pos = "mid")
```

## Interpreting cTWAS results

We list all molecular traits with PIP > 0.8, which is the threshold we used in the paper.
```{r select_sig_res}
# select gene PIP > 0.8
sig_finemap_res <- finemap_res[finemap_res$type!="SNP" & finemap_res$susie_pip > 0.8,]
head(sig_finemap_res[order(-sig_finemap_res$susie_pip),])
```


When we have multiple contexts (e.g. tissues) or types (e.g. expression, splicing), 
it is useful to evaluate the combined PIP for each gene by contexts, types, or groups.
We could further limit results in credible sets and/or protein coding genes.
```{r combine_pips}
# combine gene PIPs by context
combined_pip_by_context <- combine_gene_pips(finemap_res, 
                                             by = "context",
                                             filter_cs = TRUE,
                                             filter_protein_coding_genes = TRUE)
print(combined_pip_by_context)

# combine gene PIPs by type
combined_pip_by_type <- combine_gene_pips(finemap_res, 
                                          by = "type",
                                          filter_cs = TRUE,
                                          filter_protein_coding_genes = TRUE)

# combine gene PIPs by group
combined_pip_by_group <- combine_gene_pips(finemap_res, 
                                            by = "group",
                                            filter_cs = TRUE,
                                            filter_protein_coding_genes = TRUE)

# select combined PIP > 0.8
sig_gene_pips_df <- combined_pip_by_context[combined_pip_by_context$combined_pip > 0.8, ]
print(sig_gene_pips_df)
```


## Visualizing cTWAS results

We could make locus plots for regions of interest. 
We could zoom in the region by specifying the `locus_range`.

If we want to show the correlations with the focus gene,
we need correlation matrices of the region. 
Because we have computed and saved correlation matrices during fine-mapping step,
we could simply load those precomputed correlation matrices.
```{r get_region_cor_res}
cor_res <- get_region_cor(region_id = "16_71020125_72901251", cor_dir = cor_dir)
```

```{r locus_plot, fig.width=10, fig.height=8}
make_locusplot(finemap_res,
               region_id = "16_71020125_72901251",
               weights = weights,
               ens_db = ens_db, 
               R_snp_gene = cor_res$R_snp_gene,
               R_gene = cor_res$R_gene,
               locus_range = c(71.6e6,72.4e6),
               filter_gene_biotype = "protein_coding",
               highlight_pip = 0.8,
               legend.position = "right")
```

