<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Run on individual-level data â€¢ ctwas</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Run on individual-level data">
<meta property="og:description" content="ctwas">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ctwas</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.8</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/ctwas_idv_wkfl.html">Run on individual-level data</a>
    </li>
    <li>
      <a href="../articles/ctwas_rss_wkfl.html">Run on summary statistics data</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Run on individual-level data</h1>
            
      
      
      <div class="hidden name"><code>ctwas_idv_wkfl.Rmd</code></div>

    </div>

    
    
<div id="requirements" class="section level1">
<h1 class="hasAnchor">
<a href="#requirements" class="anchor"></a>Requirements</h1>
<p>Weights and LD refernce ID should have same meaning and ref/alt allele info needs to match. Y sample order and pgen sample order (as defined in .psam for plink2 files or .fam for plink1 files) should match. All coordinates are in hg19.</p>
</div>
<div id="prepare-input" class="section level1">
<h1 class="hasAnchor">
<a href="#prepare-input" class="anchor"></a>Prepare input</h1>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">ctwas</span><span class="op">)</span></code></pre></div>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># phenotype</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="va">Y</span><span class="op">)</span>

<span class="co"># individual level genotype data</span>
<span class="va">pgenfs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata/example_genotype_files"</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"example_chr"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">22</span>, <span class="st">".pgen"</span><span class="op">)</span>, package <span class="op">=</span> <span class="st">"ctwas"</span><span class="op">)</span>

<span class="co"># weight</span>
<span class="va">weight</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata/example_fusion_weights"</span>, <span class="st">"Tissue"</span>, package <span class="op">=</span> <span class="st">"ctwas"</span><span class="op">)</span></code></pre></div>
</div>
<div id="impute-expression" class="section level1">
<h1 class="hasAnchor">
<a href="#impute-expression" class="anchor"></a>Impute expression</h1>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">outputdir</span> <span class="op">&lt;-</span> <span class="st">"~/temp"</span>
<span class="va">exprfs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/vector.html">vector</a></span><span class="op">(</span><span class="op">)</span>
<span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fl">22</span><span class="op">)</span><span class="op">{</span>
  <span class="va">pgenf</span> <span class="op">&lt;-</span> <span class="va">pgenfs</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>
  <span class="va">exprfs</span><span class="op">[</span><span class="va">i</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/impute_expr.html">impute_expr</a></span><span class="op">(</span>pgenf <span class="op">=</span> <span class="va">pgenf</span>, weight <span class="op">=</span> <span class="va">weight</span>,
                           method <span class="op">=</span> <span class="st">"lasso"</span>, outputdir <span class="op">=</span> <span class="va">outputdir</span>,
                           outname <span class="op">=</span> <span class="st">"test"</span><span class="op">)</span>
<span class="op">}</span>
<span class="co">#&gt; 2021-01-31 13:15:18 INFO::expression inmputation started for chr 1.</span>
<span class="co">#&gt; 2021-01-31 13:15:18 INFO::number of genes with weights provided: 20</span>
<span class="co">#&gt; 2021-01-31 13:15:18 INFO::Number of genes with imputed expression: 1 for chr 1</span>
<span class="co">#&gt; 2021-01-31 13:15:18 INFO::expression inmputation done for chr 1.</span>
<span class="co">#&gt; 2021-01-31 13:15:19 INFO::expression inmputation started for chr 2.</span>
<span class="co">#&gt; 2021-01-31 13:15:19 INFO::number of genes with weights provided: 20</span>
<span class="co">#&gt; 2021-01-31 13:15:19 INFO::Number of genes with imputed expression: 1 for chr 2</span>
<span class="co">#&gt; 2021-01-31 13:15:19 INFO::expression inmputation done for chr 2.</span>
<span class="co">#&gt; 2021-01-31 13:15:19 INFO::expression inmputation started for chr 3.</span>
<span class="co">#&gt; 2021-01-31 13:15:19 INFO::number of genes with weights provided: 20</span>
<span class="co">#&gt; 2021-01-31 13:15:19 INFO::Number of genes with imputed expression: 1 for chr 3</span>
<span class="co">#&gt; 2021-01-31 13:15:19 INFO::expression inmputation done for chr 3.</span>
<span class="co">#&gt; 2021-01-31 13:15:19 INFO::expression inmputation started for chr 4.</span>
<span class="co">#&gt; 2021-01-31 13:15:19 INFO::number of genes with weights provided: 20</span>
<span class="co">#&gt; 2021-01-31 13:15:19 INFO::Number of genes with imputed expression: 1 for chr 4</span>
<span class="co">#&gt; 2021-01-31 13:15:19 INFO::expression inmputation done for chr 4.</span>
<span class="co">#&gt; 2021-01-31 13:15:19 INFO::expression inmputation started for chr 5.</span>
<span class="co">#&gt; 2021-01-31 13:15:19 INFO::number of genes with weights provided: 20</span>
<span class="co">#&gt; 2021-01-31 13:15:19 INFO::Number of genes with imputed expression: 1 for chr 5</span>
<span class="co">#&gt; 2021-01-31 13:15:19 INFO::expression inmputation done for chr 5.</span>
<span class="co">#&gt; 2021-01-31 13:15:19 INFO::expression inmputation started for chr 6.</span>
<span class="co">#&gt; 2021-01-31 13:15:19 INFO::number of genes with weights provided: 20</span>
<span class="co">#&gt; 2021-01-31 13:15:19 INFO::Number of genes with imputed expression: 1 for chr 6</span>
<span class="co">#&gt; 2021-01-31 13:15:19 INFO::expression inmputation done for chr 6.</span>
<span class="co">#&gt; 2021-01-31 13:15:19 INFO::expression inmputation started for chr 7.</span>
<span class="co">#&gt; 2021-01-31 13:15:19 INFO::number of genes with weights provided: 20</span>
<span class="co">#&gt; 2021-01-31 13:15:19 INFO::Number of genes with imputed expression: 1 for chr 7</span>
<span class="co">#&gt; 2021-01-31 13:15:19 INFO::expression inmputation done for chr 7.</span>
<span class="co">#&gt; 2021-01-31 13:15:19 INFO::expression inmputation started for chr 8.</span>
<span class="co">#&gt; 2021-01-31 13:15:19 INFO::number of genes with weights provided: 20</span>
<span class="co">#&gt; 2021-01-31 13:15:19 INFO::Number of genes with imputed expression: 1 for chr 8</span>
<span class="co">#&gt; 2021-01-31 13:15:19 INFO::expression inmputation done for chr 8.</span>
<span class="co">#&gt; 2021-01-31 13:15:19 INFO::expression inmputation started for chr 9.</span>
<span class="co">#&gt; 2021-01-31 13:15:19 INFO::number of genes with weights provided: 20</span>
<span class="co">#&gt; 2021-01-31 13:15:19 INFO::Number of genes with imputed expression: 1 for chr 9</span>
<span class="co">#&gt; 2021-01-31 13:15:19 INFO::expression inmputation done for chr 9.</span>
<span class="co">#&gt; 2021-01-31 13:15:19 INFO::expression inmputation started for chr 10.</span>
<span class="co">#&gt; 2021-01-31 13:15:19 INFO::number of genes with weights provided: 20</span>
<span class="co">#&gt; 2021-01-31 13:15:19 INFO::Number of genes with imputed expression: 1 for chr 10</span>
<span class="co">#&gt; 2021-01-31 13:15:19 INFO::expression inmputation done for chr 10.</span>
<span class="co">#&gt; 2021-01-31 13:15:19 INFO::expression inmputation started for chr 11.</span>
<span class="co">#&gt; 2021-01-31 13:15:19 INFO::number of genes with weights provided: 20</span>
<span class="co">#&gt; 2021-01-31 13:15:19 INFO::Number of genes with imputed expression: 1 for chr 11</span>
<span class="co">#&gt; 2021-01-31 13:15:19 INFO::expression inmputation done for chr 11.</span>
<span class="co">#&gt; 2021-01-31 13:15:19 INFO::expression inmputation started for chr 12.</span>
<span class="co">#&gt; 2021-01-31 13:15:19 INFO::number of genes with weights provided: 20</span>
<span class="co">#&gt; 2021-01-31 13:15:19 INFO::Number of genes with imputed expression: 1 for chr 12</span>
<span class="co">#&gt; 2021-01-31 13:15:19 INFO::expression inmputation done for chr 12.</span>
<span class="co">#&gt; 2021-01-31 13:15:19 INFO::expression inmputation started for chr 13.</span>
<span class="co">#&gt; 2021-01-31 13:15:19 INFO::number of genes with weights provided: 20</span>
<span class="co">#&gt; 2021-01-31 13:15:19 INFO::Number of genes with imputed expression: 1 for chr 13</span>
<span class="co">#&gt; 2021-01-31 13:15:19 INFO::expression inmputation done for chr 13.</span>
<span class="co">#&gt; 2021-01-31 13:15:19 INFO::expression inmputation started for chr 14.</span>
<span class="co">#&gt; 2021-01-31 13:15:19 INFO::number of genes with weights provided: 20</span>
<span class="co">#&gt; 2021-01-31 13:15:19 INFO::Number of genes with imputed expression: 1 for chr 14</span>
<span class="co">#&gt; 2021-01-31 13:15:19 INFO::expression inmputation done for chr 14.</span>
<span class="co">#&gt; 2021-01-31 13:15:19 INFO::expression inmputation started for chr 15.</span>
<span class="co">#&gt; 2021-01-31 13:15:19 INFO::number of genes with weights provided: 20</span>
<span class="co">#&gt; 2021-01-31 13:15:19 INFO::Number of genes with imputed expression: 1 for chr 15</span>
<span class="co">#&gt; 2021-01-31 13:15:19 INFO::expression inmputation done for chr 15.</span>
<span class="co">#&gt; 2021-01-31 13:15:19 INFO::expression inmputation started for chr 16.</span>
<span class="co">#&gt; 2021-01-31 13:15:19 INFO::number of genes with weights provided: 20</span>
<span class="co">#&gt; 2021-01-31 13:15:19 INFO::Number of genes with imputed expression: 1 for chr 16</span>
<span class="co">#&gt; 2021-01-31 13:15:19 INFO::expression inmputation done for chr 16.</span>
<span class="co">#&gt; 2021-01-31 13:15:19 INFO::expression inmputation started for chr 17.</span>
<span class="co">#&gt; 2021-01-31 13:15:19 INFO::number of genes with weights provided: 20</span>
<span class="co">#&gt; 2021-01-31 13:15:19 INFO::Number of genes with imputed expression: 1 for chr 17</span>
<span class="co">#&gt; 2021-01-31 13:15:19 INFO::expression inmputation done for chr 17.</span>
<span class="co">#&gt; 2021-01-31 13:15:19 INFO::expression inmputation started for chr 18.</span>
<span class="co">#&gt; 2021-01-31 13:15:19 INFO::number of genes with weights provided: 20</span>
<span class="co">#&gt; 2021-01-31 13:15:19 INFO::Number of genes with imputed expression: 1 for chr 18</span>
<span class="co">#&gt; 2021-01-31 13:15:19 INFO::expression inmputation done for chr 18.</span>
<span class="co">#&gt; 2021-01-31 13:15:20 INFO::expression inmputation started for chr 19.</span>
<span class="co">#&gt; 2021-01-31 13:15:20 INFO::number of genes with weights provided: 20</span>
<span class="co">#&gt; 2021-01-31 13:15:20 INFO::Number of genes with imputed expression: 1 for chr 19</span>
<span class="co">#&gt; 2021-01-31 13:15:20 INFO::expression inmputation done for chr 19.</span>
<span class="co">#&gt; 2021-01-31 13:15:20 INFO::expression inmputation started for chr 20.</span>
<span class="co">#&gt; 2021-01-31 13:15:20 INFO::number of genes with weights provided: 20</span>
<span class="co">#&gt; 2021-01-31 13:15:20 INFO::Number of genes with imputed expression: 1 for chr 20</span>
<span class="co">#&gt; 2021-01-31 13:15:20 INFO::expression inmputation done for chr 20.</span>
<span class="co">#&gt; 2021-01-31 13:15:20 INFO::expression inmputation started for chr 21.</span>
<span class="co">#&gt; 2021-01-31 13:15:20 INFO::number of genes with weights provided: 20</span>
<span class="co">#&gt; 2021-01-31 13:15:20 INFO::Number of genes with imputed expression: 0 for chr 21</span>
<span class="co">#&gt; 2021-01-31 13:15:20 INFO::expression inmputation done for chr 21.</span>
<span class="co">#&gt; 2021-01-31 13:15:20 INFO::expression inmputation started for chr 22.</span>
<span class="co">#&gt; 2021-01-31 13:15:20 INFO::number of genes with weights provided: 20</span>
<span class="co">#&gt; 2021-01-31 13:15:20 INFO::Number of genes with imputed expression: 0 for chr 22</span>
<span class="co">#&gt; 2021-01-31 13:15:20 INFO::expression inmputation done for chr 22.</span></code></pre></div>
</div>
<div id="run-ctwas" class="section level1">
<h1 class="hasAnchor">
<a href="#run-ctwas" class="anchor"></a>Run ctwas</h1>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">regionsfile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"example_regions.bed"</span>, package <span class="op">=</span> <span class="st">"ctwas"</span><span class="op">)</span>

<span class="fu"><a href="../reference/ctwas.html">ctwas</a></span><span class="op">(</span><span class="va">pgenfs</span>, <span class="va">exprfs</span>, <span class="va">Y</span>, ld_regions_custom <span class="op">=</span> <span class="va">regionsfile</span>, thin <span class="op">=</span> <span class="fl">1</span>, outputdir <span class="op">=</span> <span class="va">outputdir</span>, outname <span class="op">=</span> <span class="st">"test"</span>, ncore <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>
<span class="co">#&gt; 2021-01-31 13:15:20 INFO::ctwas started ...</span>
<span class="co">#&gt; 2021-01-31 13:15:20 INFO::LD region file: /tmp/RtmpMImhSh/temp_libpath7ebf78aade23/ctwas/extdata/example_regions.bed</span>
<span class="co">#&gt; 2021-01-31 13:15:20 INFO::Run susie iteratively, getting rough estimate ...</span>
<span class="co">#&gt; 2021-01-31 13:15:20 INFO::No. LD regions: 374</span>
<span class="co">#&gt; 2021-01-31 13:15:20 INFO::No. regions with at least one SNP/gene for chr1: 17</span>
<span class="co">#&gt; 2021-01-31 13:15:20 INFO::No. regions with at least one SNP/gene for chr2: 17</span>
<span class="co">#&gt; 2021-01-31 13:15:20 INFO::No. regions with at least one SNP/gene for chr3: 17</span>
<span class="co">#&gt; 2021-01-31 13:15:20 INFO::No. regions with at least one SNP/gene for chr4: 17</span>
<span class="co">#&gt; 2021-01-31 13:15:20 INFO::No. regions with at least one SNP/gene for chr5: 17</span>
<span class="co">#&gt; 2021-01-31 13:15:20 INFO::No. regions with at least one SNP/gene for chr6: 17</span>
<span class="co">#&gt; 2021-01-31 13:15:20 INFO::No. regions with at least one SNP/gene for chr7: 17</span>
<span class="co">#&gt; 2021-01-31 13:15:20 INFO::No. regions with at least one SNP/gene for chr8: 17</span>
<span class="co">#&gt; 2021-01-31 13:15:20 INFO::No. regions with at least one SNP/gene for chr9: 17</span>
<span class="co">#&gt; 2021-01-31 13:15:20 INFO::No. regions with at least one SNP/gene for chr10: 17</span>
<span class="co">#&gt; 2021-01-31 13:15:20 INFO::No. regions with at least one SNP/gene for chr11: 17</span>
<span class="co">#&gt; 2021-01-31 13:15:20 INFO::No. regions with at least one SNP/gene for chr12: 17</span>
<span class="co">#&gt; 2021-01-31 13:15:20 INFO::No. regions with at least one SNP/gene for chr13: 17</span>
<span class="co">#&gt; 2021-01-31 13:15:20 INFO::No. regions with at least one SNP/gene for chr14: 17</span>
<span class="co">#&gt; 2021-01-31 13:15:20 INFO::No. regions with at least one SNP/gene for chr15: 17</span>
<span class="co">#&gt; 2021-01-31 13:15:20 INFO::No. regions with at least one SNP/gene for chr16: 17</span>
<span class="co">#&gt; 2021-01-31 13:15:20 INFO::No. regions with at least one SNP/gene for chr17: 17</span>
<span class="co">#&gt; 2021-01-31 13:15:20 INFO::No. regions with at least one SNP/gene for chr18: 17</span>
<span class="co">#&gt; 2021-01-31 13:15:20 INFO::No. regions with at least one SNP/gene for chr19: 17</span>
<span class="co">#&gt; 2021-01-31 13:15:20 INFO::No. regions with at least one SNP/gene for chr20: 17</span>
<span class="co">#&gt; 2021-01-31 13:15:20 INFO::No. regions with at least one SNP/gene for chr21: 17</span>
<span class="co">#&gt; 2021-01-31 13:15:20 INFO::No. regions with at least one SNP/gene for chr22: 17</span>
<span class="co">#&gt; 2021-01-31 13:15:20 INFO::run iteration 1</span>
<span class="co">#&gt; 2021-01-31 13:15:31 INFO::After iteration 1, gene prior 0.253154795140839:, SNP prior:0.0358330326434716</span>
<span class="co">#&gt; 2021-01-31 13:15:31 INFO::run iteration 2</span>
<span class="co">#&gt; 2021-01-31 13:15:40 INFO::After iteration 2, gene prior 0.132051318381414:, SNP prior:0.0344389231166242</span>
<span class="co">#&gt; 2021-01-31 13:15:40 INFO::run iteration 3</span>
<span class="co">#&gt; 2021-01-31 13:15:49 INFO::After iteration 3, gene prior 0.0804519667580253:, SNP prior:0.0343732967104774</span>
<span class="co">#&gt; 2021-01-31 13:15:50 INFO::Blocks are filtered: 22 blocks left</span>
<span class="co">#&gt; 2021-01-31 13:15:50 INFO::Run susie iteratively, getting accurate estimate ...</span>
<span class="co">#&gt; 2021-01-31 13:15:50 INFO::run iteration 1</span>
<span class="co">#&gt; 2021-01-31 13:15:53 INFO::After iteration 1, gene prior 0.0642345701692352:, SNP prior:0.0274211548626107</span>
<span class="co">#&gt; 2021-01-31 13:15:53 INFO::run iteration 2</span>
<span class="co">#&gt; 2021-01-31 13:15:55 INFO::After iteration 2, gene prior 0.0599204896969733:, SNP prior:0.0244099580229773</span>
<span class="co">#&gt; 2021-01-31 13:15:55 INFO::run iteration 3</span>
<span class="co">#&gt; 2021-01-31 13:15:57 INFO::After iteration 3, gene prior 0.0588888738156299:, SNP prior:0.0225452632402439</span>
<span class="co">#&gt; 2021-01-31 13:15:57 INFO::run iteration 4</span>
<span class="co">#&gt; 2021-01-31 13:15:59 INFO::After iteration 4, gene prior 0.0586465815699955:, SNP prior:0.0212202700261521</span>
<span class="co">#&gt; 2021-01-31 13:15:59 INFO::run iteration 5</span>
<span class="co">#&gt; 2021-01-31 13:16:01 INFO::After iteration 5, gene prior 0.0585893484787186:, SNP prior:0.0202031796666298</span>
<span class="co">#&gt; 2021-01-31 13:16:01 INFO::run iteration 6</span>
<span class="co">#&gt; 2021-01-31 13:16:04 INFO::After iteration 6, gene prior 0.0585760534778286:, SNP prior:0.0193822761085657</span>
<span class="co">#&gt; 2021-01-31 13:16:04 INFO::run iteration 7</span>
<span class="co">#&gt; 2021-01-31 13:16:06 INFO::After iteration 7, gene prior 0.0585736054718825:, SNP prior:0.0186961687319586</span>
<span class="co">#&gt; 2021-01-31 13:16:06 INFO::run iteration 8</span>
<span class="co">#&gt; 2021-01-31 13:16:08 INFO::After iteration 8, gene prior 0.0585740223982371:, SNP prior:0.0181079279857141</span>
<span class="co">#&gt; 2021-01-31 13:16:08 INFO::run iteration 9</span>
<span class="co">#&gt; 2021-01-31 13:16:10 INFO::After iteration 9, gene prior 0.0585752605251688:, SNP prior:0.0175937758624876</span>
<span class="co">#&gt; 2021-01-31 13:16:10 INFO::run iteration 10</span>
<span class="co">#&gt; 2021-01-31 13:16:12 INFO::After iteration 10, gene prior 0.0585767390161605:, SNP prior:0.017137569030242</span>
<span class="co">#&gt; 2021-01-31 13:16:12 INFO::run iteration 11</span>
<span class="co">#&gt; 2021-01-31 13:16:14 INFO::After iteration 11, gene prior 0.0585782666560106:, SNP prior:0.0167278749458581</span>
<span class="co">#&gt; 2021-01-31 13:16:14 INFO::run iteration 12</span>
<span class="co">#&gt; 2021-01-31 13:16:16 INFO::After iteration 12, gene prior 0.0585797702619266:, SNP prior:0.0163563152076408</span>
<span class="co">#&gt; 2021-01-31 13:16:16 INFO::run iteration 13</span>
<span class="co">#&gt; 2021-01-31 13:16:18 INFO::After iteration 13, gene prior 0.0585812190273481:, SNP prior:0.0160165737969517</span>
<span class="co">#&gt; 2021-01-31 13:16:18 INFO::run iteration 14</span>
<span class="co">#&gt; 2021-01-31 13:16:20 INFO::After iteration 14, gene prior 0.0585826001219523:, SNP prior:0.01570377546209</span>
<span class="co">#&gt; 2021-01-31 13:16:20 INFO::run iteration 15</span>
<span class="co">#&gt; 2021-01-31 13:16:22 INFO::After iteration 15, gene prior 0.0585839093124729:, SNP prior:0.015414080735491</span>
<span class="co">#&gt; 2021-01-31 13:16:22 INFO::run iteration 16</span>
<span class="co">#&gt; 2021-01-31 13:16:24 INFO::After iteration 16, gene prior 0.0585851467380859:, SNP prior:0.0151444132585396</span>
<span class="co">#&gt; 2021-01-31 13:16:24 INFO::run iteration 17</span>
<span class="co">#&gt; 2021-01-31 13:16:26 INFO::After iteration 17, gene prior 0.0585863147840351:, SNP prior:0.0148922709197058</span>
<span class="co">#&gt; 2021-01-31 13:16:26 INFO::run iteration 18</span>
<span class="co">#&gt; 2021-01-31 13:16:28 INFO::After iteration 18, gene prior 0.0585874169398054:, SNP prior:0.0146555917904577</span>
<span class="co">#&gt; 2021-01-31 13:16:28 INFO::run iteration 19</span>
<span class="co">#&gt; 2021-01-31 13:16:30 INFO::After iteration 19, gene prior 0.0585884571660951:, SNP prior:0.0144326568892288</span>
<span class="co">#&gt; 2021-01-31 13:16:30 INFO::run iteration 20</span>
<span class="co">#&gt; 2021-01-31 13:16:32 INFO::After iteration 20, gene prior 0.0585894395406753:, SNP prior:0.0142220183033906</span>
<span class="co">#&gt; 2021-01-31 13:16:32 INFO::run iteration 21</span>
<span class="co">#&gt; 2021-01-31 13:16:34 INFO::After iteration 21, gene prior 0.0585903680625106:, SNP prior:0.0140224451502488</span>
<span class="co">#&gt; 2021-01-31 13:16:34 INFO::run iteration 22</span>
<span class="co">#&gt; 2021-01-31 13:16:36 INFO::After iteration 22, gene prior 0.0585912465474825:, SNP prior:0.0138328823300905</span>
<span class="co">#&gt; 2021-01-31 13:16:36 INFO::run iteration 23</span>
<span class="co">#&gt; 2021-01-31 13:16:38 INFO::After iteration 23, gene prior 0.0585920785776198:, SNP prior:0.0136524186111371</span>
<span class="co">#&gt; 2021-01-31 13:16:38 INFO::run iteration 24</span>
<span class="co">#&gt; 2021-01-31 13:16:41 INFO::After iteration 24, gene prior 0.0585928674815719:, SNP prior:0.0134802616285783</span>
<span class="co">#&gt; 2021-01-31 13:16:41 INFO::run iteration 25</span>
<span class="co">#&gt; 2021-01-31 13:16:43 INFO::After iteration 25, gene prior 0.0585936163331067:, SNP prior:0.0133157180788571</span>
<span class="co">#&gt; 2021-01-31 13:16:43 INFO::run iteration 26</span>
<span class="co">#&gt; 2021-01-31 13:16:45 INFO::After iteration 26, gene prior 0.0585943279597108:, SNP prior:0.0131581778680279</span>
<span class="co">#&gt; 2021-01-31 13:16:45 INFO::run iteration 27</span>
<span class="co">#&gt; 2021-01-31 13:16:47 INFO::After iteration 27, gene prior 0.0585950049565232:, SNP prior:0.0130071013050308</span>
<span class="co">#&gt; 2021-01-31 13:16:47 INFO::run iteration 28</span>
<span class="co">#&gt; 2021-01-31 13:16:49 INFO::After iteration 28, gene prior 0.0585956497027374:, SNP prior:0.0128620086651556</span>
<span class="co">#&gt; 2021-01-31 13:16:49 INFO::run iteration 29</span>
<span class="co">#&gt; 2021-01-31 13:16:51 INFO::After iteration 29, gene prior 0.058596264378766:, SNP prior:0.0127224716168898</span>
<span class="co">#&gt; 2021-01-31 13:16:51 INFO::run iteration 30</span>
<span class="co">#&gt; 2021-01-31 13:16:53 INFO::After iteration 30, gene prior 0.0585968509831772:, SNP prior:0.0125881061272171</span>
<span class="co">#&gt; 2021-01-31 13:16:53 INFO::Run susie for all regions.</span>
<span class="co">#&gt; 2021-01-31 13:16:53 INFO::No. LD regions: 374</span>
<span class="co">#&gt; 2021-01-31 13:16:53 INFO::No. regions with at least one SNP/gene for chr1: 17</span>
<span class="co">#&gt; 2021-01-31 13:16:53 INFO::No. regions with at least one SNP/gene for chr2: 17</span>
<span class="co">#&gt; 2021-01-31 13:16:53 INFO::No. regions with at least one SNP/gene for chr3: 17</span>
<span class="co">#&gt; 2021-01-31 13:16:53 INFO::No. regions with at least one SNP/gene for chr4: 17</span>
<span class="co">#&gt; 2021-01-31 13:16:53 INFO::No. regions with at least one SNP/gene for chr5: 17</span>
<span class="co">#&gt; 2021-01-31 13:16:53 INFO::No. regions with at least one SNP/gene for chr6: 17</span>
<span class="co">#&gt; 2021-01-31 13:16:53 INFO::No. regions with at least one SNP/gene for chr7: 17</span>
<span class="co">#&gt; 2021-01-31 13:16:53 INFO::No. regions with at least one SNP/gene for chr8: 17</span>
<span class="co">#&gt; 2021-01-31 13:16:53 INFO::No. regions with at least one SNP/gene for chr9: 17</span>
<span class="co">#&gt; 2021-01-31 13:16:53 INFO::No. regions with at least one SNP/gene for chr10: 17</span>
<span class="co">#&gt; 2021-01-31 13:16:53 INFO::No. regions with at least one SNP/gene for chr11: 17</span>
<span class="co">#&gt; 2021-01-31 13:16:53 INFO::No. regions with at least one SNP/gene for chr12: 17</span>
<span class="co">#&gt; 2021-01-31 13:16:53 INFO::No. regions with at least one SNP/gene for chr13: 17</span>
<span class="co">#&gt; 2021-01-31 13:16:53 INFO::No. regions with at least one SNP/gene for chr14: 17</span>
<span class="co">#&gt; 2021-01-31 13:16:53 INFO::No. regions with at least one SNP/gene for chr15: 17</span>
<span class="co">#&gt; 2021-01-31 13:16:53 INFO::No. regions with at least one SNP/gene for chr16: 17</span>
<span class="co">#&gt; 2021-01-31 13:16:53 INFO::No. regions with at least one SNP/gene for chr17: 17</span>
<span class="co">#&gt; 2021-01-31 13:16:53 INFO::No. regions with at least one SNP/gene for chr18: 17</span>
<span class="co">#&gt; 2021-01-31 13:16:53 INFO::No. regions with at least one SNP/gene for chr19: 17</span>
<span class="co">#&gt; 2021-01-31 13:16:53 INFO::No. regions with at least one SNP/gene for chr20: 17</span>
<span class="co">#&gt; 2021-01-31 13:16:53 INFO::No. regions with at least one SNP/gene for chr21: 17</span>
<span class="co">#&gt; 2021-01-31 13:16:53 INFO::No. regions with at least one SNP/gene for chr22: 17</span>
<span class="co">#&gt; 2021-01-31 13:16:53 INFO::run iteration 1</span>
<span class="co">#&gt; 2021-01-31 13:17:09 INFO::After iteration 1, gene prior 0.084251176908959:, SNP prior:0.0631871447349</span>
<span class="co">#&gt; $group_prior</span>
<span class="co">#&gt; [1] 0.05859685 0.01258811</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $group_prior_var</span>
<span class="co">#&gt; [1] 0.2031186515 0.0002644459</span></code></pre></div>
<div class="sourceCode"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; R version 3.6.1 (2019-07-05)</span>
<span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span>
<span class="co">#&gt; Running under: Scientific Linux 7.4 (Nitrogen)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Matrix products: default</span>
<span class="co">#&gt; BLAS/LAPACK: /software/openblas-0.2.19-el7-x86_64/lib/libopenblas_haswellp-r0.2.19.so</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; locale:</span>
<span class="co">#&gt;  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span>
<span class="co">#&gt;  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span>
<span class="co">#&gt;  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span>
<span class="co">#&gt;  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span>
<span class="co">#&gt;  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span>
<span class="co">#&gt; [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; attached base packages:</span>
<span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; other attached packages:</span>
<span class="co">#&gt; [1] ctwas_0.1.8</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; loaded via a namespace (and not attached):</span>
<span class="co">#&gt;  [1] Rcpp_1.0.5        compiler_3.6.1    pillar_1.4.2     </span>
<span class="co">#&gt;  [4] iterators_1.0.10  tools_3.6.1       digest_0.6.20    </span>
<span class="co">#&gt;  [7] evaluate_0.14     memoise_1.1.0     tibble_2.1.3     </span>
<span class="co">#&gt; [10] lattice_0.20-38   pkgconfig_2.0.2   rlang_0.4.6      </span>
<span class="co">#&gt; [13] Matrix_1.2-18     foreach_1.4.4     yaml_2.2.0       </span>
<span class="co">#&gt; [16] parallel_3.6.1    pkgdown_1.6.1     xfun_0.8         </span>
<span class="co">#&gt; [19] stringr_1.4.0     dplyr_0.8.3       knitr_1.23       </span>
<span class="co">#&gt; [22] pgenlibr_0.2      desc_1.2.0        fs_1.3.1         </span>
<span class="co">#&gt; [25] vctrs_0.3.1       rprojroot_1.3-2   grid_3.6.1       </span>
<span class="co">#&gt; [28] tidyselect_1.1.0  glue_1.3.1        data.table_1.13.2</span>
<span class="co">#&gt; [31] R6_2.4.0          rmarkdown_1.13    purrr_0.3.4      </span>
<span class="co">#&gt; [34] magrittr_1.5      backports_1.1.4   codetools_0.2-16 </span>
<span class="co">#&gt; [37] htmltools_0.3.6   assertthat_0.2.1  logging_0.10-108 </span>
<span class="co">#&gt; [40] stringi_1.4.3     doParallel_1.0.14 crayon_1.3.4</span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Siming Zhao.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
