<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Run on summary statistics data • ctwas</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Run on summary statistics data">
<meta property="og:description" content="ctwas">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ctwas</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.11</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/ctwas_idv_wkfl.html">Run on individual-level data</a>
    </li>
    <li>
      <a href="../articles/ctwas_rss_wkfl.html">Run on summary statistics data</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/simingz/ctwas">Source</a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="ctwas_rss_wkfl_files/header-attrs-2.6/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Run on summary statistics data</h1>
            
      
      
      <div class="hidden name"><code>ctwas_rss_wkfl.Rmd</code></div>

    </div>

    
    
<div id="purpose-of-the-analysis" class="section level1">
<h1 class="hasAnchor">
<a href="#purpose-of-the-analysis" class="anchor"></a>Purpose of the analysis</h1>
<p>This workflow aims to use GWAS summary statistics data as input and implements the causal TWAS algorithm. The “causal TWAS” method is developed to identify causal genes in GWAS studies. We improve the current TWAS analysis mainly by reducing false positives. We start from z scores for each SNP in a GWAS study and expression model trained by the <a href="http://gusevlab.org/projects/fusion/">Fusion/TWAS software</a>. In the end, we will provide a PIP (the posterior inclusion probality for being a causal gene) for each gene with an expression model.</p>
<p>Below is the workflow for running <code>ctwas</code> for summary statistics data. I provided example input files with the package and you can use the following code as a toy pipeline. You can prepare your own input files following the examples.</p>
</div>
<div id="step-1-prepare-input" class="section level1">
<h1 class="hasAnchor">
<a href="#step-1-prepare-input" class="anchor"></a>step 1: Prepare input</h1>
<p>Our package is called <code>ctwas</code>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">ctwas</span><span class="op">)</span></code></pre></div>
<p>The input are given below: z scores, a data frame:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># z scores</span>
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="va">zdf.snp</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">zdf.snp</span><span class="op">)</span>
<span class="co">#&gt;       id          z</span>
<span class="co">#&gt; 1 rs1001  -6.644254</span>
<span class="co">#&gt; 2 rs1002  14.369485</span>
<span class="co">#&gt; 3 rs1003 -14.180414</span>
<span class="co">#&gt; 4 rs1004  -6.739014</span>
<span class="co">#&gt; 5 rs1005 -14.049735</span>
<span class="co">#&gt; 6 rs1006  -2.176475</span></code></pre></div>
<p>The genotype files for LD reference data. This is a character vector of .pgen or .bed files. One file for one chromosome, in the order of 1 to 22. Therefore, the length of this vector needs to be 22. If .pgen files are given, then .pvar and .psam are assumed to present in the same directory. If .bed files are given, then .bim and .fam files are assumed to present in the same directory.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># genotype for LD reference</span>
<span class="va">ld_pgenfs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata/example_genotype_files"</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"example_chr"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">22</span>, <span class="st">".pgen"</span><span class="op">)</span>, package <span class="op">=</span> <span class="st">"ctwas"</span><span class="op">)</span>

<span class="va">ld_pgenfs</span>
<span class="co">#&gt;  [1] "/tmp/RtmpfUkBMG/temp_libpath6611c462696/ctwas/extdata/example_genotype_files/example_chr1.pgen" </span>
<span class="co">#&gt;  [2] "/tmp/RtmpfUkBMG/temp_libpath6611c462696/ctwas/extdata/example_genotype_files/example_chr2.pgen" </span>
<span class="co">#&gt;  [3] "/tmp/RtmpfUkBMG/temp_libpath6611c462696/ctwas/extdata/example_genotype_files/example_chr3.pgen" </span>
<span class="co">#&gt;  [4] "/tmp/RtmpfUkBMG/temp_libpath6611c462696/ctwas/extdata/example_genotype_files/example_chr4.pgen" </span>
<span class="co">#&gt;  [5] "/tmp/RtmpfUkBMG/temp_libpath6611c462696/ctwas/extdata/example_genotype_files/example_chr5.pgen" </span>
<span class="co">#&gt;  [6] "/tmp/RtmpfUkBMG/temp_libpath6611c462696/ctwas/extdata/example_genotype_files/example_chr6.pgen" </span>
<span class="co">#&gt;  [7] "/tmp/RtmpfUkBMG/temp_libpath6611c462696/ctwas/extdata/example_genotype_files/example_chr7.pgen" </span>
<span class="co">#&gt;  [8] "/tmp/RtmpfUkBMG/temp_libpath6611c462696/ctwas/extdata/example_genotype_files/example_chr8.pgen" </span>
<span class="co">#&gt;  [9] "/tmp/RtmpfUkBMG/temp_libpath6611c462696/ctwas/extdata/example_genotype_files/example_chr9.pgen" </span>
<span class="co">#&gt; [10] "/tmp/RtmpfUkBMG/temp_libpath6611c462696/ctwas/extdata/example_genotype_files/example_chr10.pgen"</span>
<span class="co">#&gt; [11] "/tmp/RtmpfUkBMG/temp_libpath6611c462696/ctwas/extdata/example_genotype_files/example_chr11.pgen"</span>
<span class="co">#&gt; [12] "/tmp/RtmpfUkBMG/temp_libpath6611c462696/ctwas/extdata/example_genotype_files/example_chr12.pgen"</span>
<span class="co">#&gt; [13] "/tmp/RtmpfUkBMG/temp_libpath6611c462696/ctwas/extdata/example_genotype_files/example_chr13.pgen"</span>
<span class="co">#&gt; [14] "/tmp/RtmpfUkBMG/temp_libpath6611c462696/ctwas/extdata/example_genotype_files/example_chr14.pgen"</span>
<span class="co">#&gt; [15] "/tmp/RtmpfUkBMG/temp_libpath6611c462696/ctwas/extdata/example_genotype_files/example_chr15.pgen"</span>
<span class="co">#&gt; [16] "/tmp/RtmpfUkBMG/temp_libpath6611c462696/ctwas/extdata/example_genotype_files/example_chr16.pgen"</span>
<span class="co">#&gt; [17] "/tmp/RtmpfUkBMG/temp_libpath6611c462696/ctwas/extdata/example_genotype_files/example_chr17.pgen"</span>
<span class="co">#&gt; [18] "/tmp/RtmpfUkBMG/temp_libpath6611c462696/ctwas/extdata/example_genotype_files/example_chr18.pgen"</span>
<span class="co">#&gt; [19] "/tmp/RtmpfUkBMG/temp_libpath6611c462696/ctwas/extdata/example_genotype_files/example_chr19.pgen"</span>
<span class="co">#&gt; [20] "/tmp/RtmpfUkBMG/temp_libpath6611c462696/ctwas/extdata/example_genotype_files/example_chr20.pgen"</span>
<span class="co">#&gt; [21] "/tmp/RtmpfUkBMG/temp_libpath6611c462696/ctwas/extdata/example_genotype_files/example_chr21.pgen"</span>
<span class="co">#&gt; [22] "/tmp/RtmpfUkBMG/temp_libpath6611c462696/ctwas/extdata/example_genotype_files/example_chr22.pgen"</span></code></pre></div>
<p>Please check out <a href="http://gusevlab.org/projects/fusion/#computing-your-own-functional-weights">Fusion/TWAS</a> for the format of weights. Below is an example.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># weight</span>
<span class="va">weight</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata/example_fusion_weights"</span>, <span class="st">"Tissue"</span>, package <span class="op">=</span> <span class="st">"ctwas"</span><span class="op">)</span></code></pre></div>
<p>Note: LD reference should contain all SNPs given in z score data frame. The <code>id</code> column in z score data frame and LD reference SNP ID should have the same meaning (i.e., indicating the same SNP). Weights and LD refernce SNP ID should have same meaning and ref/alt allele info needs to match. All coordinates are in hg19.</p>
</div>
<div id="step2-impute-expression-z-scores" class="section level1">
<h1 class="hasAnchor">
<a href="#step2-impute-expression-z-scores" class="anchor"></a>step2: Impute expression z scores</h1>
<p>In the following code, we compute expression z scores for each gene. We do this chromosome by chromosome. <code>impute_expr_z</code> will return z scores for each gene. It will also generate files containing the reference LD genotypes for eQTLs in the gene and the filename is returned. We will need these files in the <code>ctwas_rss</code> function in step 3.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">outputdir</span> <span class="op">&lt;-</span> <span class="st">"~/temp"</span>
<span class="va">ld_exprfs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/vector.html">vector</a></span><span class="op">(</span><span class="op">)</span>
<span class="va">zdf.gene</span> <span class="op">&lt;-</span> <span class="cn">NULL</span>
<span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fl">22</span><span class="op">)</span><span class="op">{</span>
  <span class="va">ld_pgenf</span> <span class="op">&lt;-</span> <span class="va">ld_pgenfs</span><span class="op">[</span><span class="va">i</span><span class="op">]</span>
  <span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/impute_expr_z.html">impute_expr_z</a></span><span class="op">(</span><span class="va">zdf.snp</span>, ld_pgenf <span class="op">=</span> <span class="va">ld_pgenf</span>, weight <span class="op">=</span> <span class="va">weight</span>,
                           method <span class="op">=</span> <span class="st">"lasso"</span>, outputdir <span class="op">=</span> <span class="va">outputdir</span>,
                           outname <span class="op">=</span> <span class="st">"test_ss"</span><span class="op">)</span>
  <span class="va">ld_exprfs</span><span class="op">[</span><span class="va">i</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">res</span><span class="op">$</span><span class="va">ld_exprf</span>
  <span class="va">zdf.gene</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html">rbind</a></span><span class="op">(</span><span class="va">zdf.gene</span>, <span class="va">res</span><span class="op">$</span><span class="va">zdf</span><span class="op">)</span>
<span class="op">}</span>
<span class="co">#&gt; 2021-03-04 14:03:47 INFO::expression z score inmputation started for chr 1.</span>
<span class="co">#&gt; 2021-03-04 14:03:47 INFO::number of genes with weights provided: 20</span>
<span class="co">#&gt; 2021-03-04 14:03:47 INFO::Number of genes with imputed expression: 1 for chr 1</span>
<span class="co">#&gt; 2021-03-04 14:03:47 INFO::expression inmputation done for chr 1.</span>
<span class="co">#&gt; 2021-03-04 14:03:47 INFO::expression z score inmputation started for chr 2.</span>
<span class="co">#&gt; 2021-03-04 14:03:47 INFO::number of genes with weights provided: 20</span>
<span class="co">#&gt; 2021-03-04 14:03:47 INFO::Number of genes with imputed expression: 1 for chr 2</span>
<span class="co">#&gt; 2021-03-04 14:03:47 INFO::expression inmputation done for chr 2.</span>
<span class="co">#&gt; 2021-03-04 14:03:48 INFO::expression z score inmputation started for chr 3.</span>
<span class="co">#&gt; 2021-03-04 14:03:48 INFO::number of genes with weights provided: 20</span>
<span class="co">#&gt; 2021-03-04 14:03:48 INFO::Number of genes with imputed expression: 1 for chr 3</span>
<span class="co">#&gt; 2021-03-04 14:03:48 INFO::expression inmputation done for chr 3.</span>
<span class="co">#&gt; 2021-03-04 14:03:48 INFO::expression z score inmputation started for chr 4.</span>
<span class="co">#&gt; 2021-03-04 14:03:48 INFO::number of genes with weights provided: 20</span>
<span class="co">#&gt; 2021-03-04 14:03:48 INFO::Number of genes with imputed expression: 1 for chr 4</span>
<span class="co">#&gt; 2021-03-04 14:03:48 INFO::expression inmputation done for chr 4.</span>
<span class="co">#&gt; 2021-03-04 14:03:48 INFO::expression z score inmputation started for chr 5.</span>
<span class="co">#&gt; 2021-03-04 14:03:48 INFO::number of genes with weights provided: 20</span>
<span class="co">#&gt; 2021-03-04 14:03:48 INFO::Number of genes with imputed expression: 1 for chr 5</span>
<span class="co">#&gt; 2021-03-04 14:03:48 INFO::expression inmputation done for chr 5.</span>
<span class="co">#&gt; 2021-03-04 14:03:48 INFO::expression z score inmputation started for chr 6.</span>
<span class="co">#&gt; 2021-03-04 14:03:48 INFO::number of genes with weights provided: 20</span>
<span class="co">#&gt; 2021-03-04 14:03:48 INFO::Number of genes with imputed expression: 1 for chr 6</span>
<span class="co">#&gt; 2021-03-04 14:03:48 INFO::expression inmputation done for chr 6.</span>
<span class="co">#&gt; 2021-03-04 14:03:48 INFO::expression z score inmputation started for chr 7.</span>
<span class="co">#&gt; 2021-03-04 14:03:48 INFO::number of genes with weights provided: 20</span>
<span class="co">#&gt; 2021-03-04 14:03:48 INFO::Number of genes with imputed expression: 1 for chr 7</span>
<span class="co">#&gt; 2021-03-04 14:03:48 INFO::expression inmputation done for chr 7.</span>
<span class="co">#&gt; 2021-03-04 14:03:48 INFO::expression z score inmputation started for chr 8.</span>
<span class="co">#&gt; 2021-03-04 14:03:48 INFO::number of genes with weights provided: 20</span>
<span class="co">#&gt; 2021-03-04 14:03:48 INFO::Number of genes with imputed expression: 1 for chr 8</span>
<span class="co">#&gt; 2021-03-04 14:03:48 INFO::expression inmputation done for chr 8.</span>
<span class="co">#&gt; 2021-03-04 14:03:48 INFO::expression z score inmputation started for chr 9.</span>
<span class="co">#&gt; 2021-03-04 14:03:48 INFO::number of genes with weights provided: 20</span>
<span class="co">#&gt; 2021-03-04 14:03:48 INFO::Number of genes with imputed expression: 1 for chr 9</span>
<span class="co">#&gt; 2021-03-04 14:03:48 INFO::expression inmputation done for chr 9.</span>
<span class="co">#&gt; 2021-03-04 14:03:48 INFO::expression z score inmputation started for chr 10.</span>
<span class="co">#&gt; 2021-03-04 14:03:48 INFO::number of genes with weights provided: 20</span>
<span class="co">#&gt; 2021-03-04 14:03:48 INFO::Number of genes with imputed expression: 1 for chr 10</span>
<span class="co">#&gt; 2021-03-04 14:03:48 INFO::expression inmputation done for chr 10.</span>
<span class="co">#&gt; 2021-03-04 14:03:48 INFO::expression z score inmputation started for chr 11.</span>
<span class="co">#&gt; 2021-03-04 14:03:48 INFO::number of genes with weights provided: 20</span>
<span class="co">#&gt; 2021-03-04 14:03:48 INFO::Number of genes with imputed expression: 1 for chr 11</span>
<span class="co">#&gt; 2021-03-04 14:03:48 INFO::expression inmputation done for chr 11.</span>
<span class="co">#&gt; 2021-03-04 14:03:48 INFO::expression z score inmputation started for chr 12.</span>
<span class="co">#&gt; 2021-03-04 14:03:48 INFO::number of genes with weights provided: 20</span>
<span class="co">#&gt; 2021-03-04 14:03:48 INFO::Number of genes with imputed expression: 1 for chr 12</span>
<span class="co">#&gt; 2021-03-04 14:03:49 INFO::expression inmputation done for chr 12.</span>
<span class="co">#&gt; 2021-03-04 14:03:49 INFO::expression z score inmputation started for chr 13.</span>
<span class="co">#&gt; 2021-03-04 14:03:49 INFO::number of genes with weights provided: 20</span>
<span class="co">#&gt; 2021-03-04 14:03:49 INFO::Number of genes with imputed expression: 1 for chr 13</span>
<span class="co">#&gt; 2021-03-04 14:03:49 INFO::expression inmputation done for chr 13.</span>
<span class="co">#&gt; 2021-03-04 14:03:49 INFO::expression z score inmputation started for chr 14.</span>
<span class="co">#&gt; 2021-03-04 14:03:49 INFO::number of genes with weights provided: 20</span>
<span class="co">#&gt; 2021-03-04 14:03:49 INFO::Number of genes with imputed expression: 1 for chr 14</span>
<span class="co">#&gt; 2021-03-04 14:03:49 INFO::expression inmputation done for chr 14.</span>
<span class="co">#&gt; 2021-03-04 14:03:49 INFO::expression z score inmputation started for chr 15.</span>
<span class="co">#&gt; 2021-03-04 14:03:49 INFO::number of genes with weights provided: 20</span>
<span class="co">#&gt; 2021-03-04 14:03:49 INFO::Number of genes with imputed expression: 1 for chr 15</span>
<span class="co">#&gt; 2021-03-04 14:03:49 INFO::expression inmputation done for chr 15.</span>
<span class="co">#&gt; 2021-03-04 14:03:49 INFO::expression z score inmputation started for chr 16.</span>
<span class="co">#&gt; 2021-03-04 14:03:49 INFO::number of genes with weights provided: 20</span>
<span class="co">#&gt; 2021-03-04 14:03:49 INFO::Number of genes with imputed expression: 1 for chr 16</span>
<span class="co">#&gt; 2021-03-04 14:03:49 INFO::expression inmputation done for chr 16.</span>
<span class="co">#&gt; 2021-03-04 14:03:49 INFO::expression z score inmputation started for chr 17.</span>
<span class="co">#&gt; 2021-03-04 14:03:49 INFO::number of genes with weights provided: 20</span>
<span class="co">#&gt; 2021-03-04 14:03:49 INFO::Number of genes with imputed expression: 1 for chr 17</span>
<span class="co">#&gt; 2021-03-04 14:03:49 INFO::expression inmputation done for chr 17.</span>
<span class="co">#&gt; 2021-03-04 14:03:49 INFO::expression z score inmputation started for chr 18.</span>
<span class="co">#&gt; 2021-03-04 14:03:49 INFO::number of genes with weights provided: 20</span>
<span class="co">#&gt; 2021-03-04 14:03:49 INFO::Number of genes with imputed expression: 1 for chr 18</span>
<span class="co">#&gt; 2021-03-04 14:03:49 INFO::expression inmputation done for chr 18.</span>
<span class="co">#&gt; 2021-03-04 14:03:49 INFO::expression z score inmputation started for chr 19.</span>
<span class="co">#&gt; 2021-03-04 14:03:49 INFO::number of genes with weights provided: 20</span>
<span class="co">#&gt; 2021-03-04 14:03:49 INFO::Number of genes with imputed expression: 1 for chr 19</span>
<span class="co">#&gt; 2021-03-04 14:03:49 INFO::expression inmputation done for chr 19.</span>
<span class="co">#&gt; 2021-03-04 14:03:49 INFO::expression z score inmputation started for chr 20.</span>
<span class="co">#&gt; 2021-03-04 14:03:49 INFO::number of genes with weights provided: 20</span>
<span class="co">#&gt; 2021-03-04 14:03:49 INFO::Number of genes with imputed expression: 1 for chr 20</span>
<span class="co">#&gt; 2021-03-04 14:03:49 INFO::expression inmputation done for chr 20.</span>
<span class="co">#&gt; 2021-03-04 14:03:49 INFO::expression z score inmputation started for chr 21.</span>
<span class="co">#&gt; 2021-03-04 14:03:49 INFO::number of genes with weights provided: 20</span>
<span class="co">#&gt; 2021-03-04 14:03:49 INFO::Number of genes with imputed expression: 0 for chr 21</span>
<span class="co">#&gt; 2021-03-04 14:03:49 INFO::expression inmputation done for chr 21.</span>
<span class="co">#&gt; 2021-03-04 14:03:49 INFO::expression z score inmputation started for chr 22.</span>
<span class="co">#&gt; 2021-03-04 14:03:49 INFO::number of genes with weights provided: 20</span>
<span class="co">#&gt; 2021-03-04 14:03:49 INFO::Number of genes with imputed expression: 0 for chr 22</span>
<span class="co">#&gt; 2021-03-04 14:03:49 INFO::expression inmputation done for chr 22.</span>
<span class="va">zdf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html">rbind</a></span><span class="op">(</span><span class="va">zdf.snp</span>, <span class="va">zdf.gene</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">zdf</span><span class="op">)</span>
<span class="co">#&gt;       id          z</span>
<span class="co">#&gt; 1 rs1001  -6.644254</span>
<span class="co">#&gt; 2 rs1002  14.369485</span>
<span class="co">#&gt; 3 rs1003 -14.180414</span>
<span class="co">#&gt; 4 rs1004  -6.739014</span>
<span class="co">#&gt; 5 rs1005 -14.049735</span>
<span class="co">#&gt; 6 rs1006  -2.176475</span>
<span class="va">ld_exprfs</span>
<span class="co">#&gt;  [1] "~/temp/test_ss_chr1.expr.gz"  "~/temp/test_ss_chr2.expr.gz" </span>
<span class="co">#&gt;  [3] "~/temp/test_ss_chr3.expr.gz"  "~/temp/test_ss_chr4.expr.gz" </span>
<span class="co">#&gt;  [5] "~/temp/test_ss_chr5.expr.gz"  "~/temp/test_ss_chr6.expr.gz" </span>
<span class="co">#&gt;  [7] "~/temp/test_ss_chr7.expr.gz"  "~/temp/test_ss_chr8.expr.gz" </span>
<span class="co">#&gt;  [9] "~/temp/test_ss_chr9.expr.gz"  "~/temp/test_ss_chr10.expr.gz"</span>
<span class="co">#&gt; [11] "~/temp/test_ss_chr11.expr.gz" "~/temp/test_ss_chr12.expr.gz"</span>
<span class="co">#&gt; [13] "~/temp/test_ss_chr13.expr.gz" "~/temp/test_ss_chr14.expr.gz"</span>
<span class="co">#&gt; [15] "~/temp/test_ss_chr15.expr.gz" "~/temp/test_ss_chr16.expr.gz"</span>
<span class="co">#&gt; [17] "~/temp/test_ss_chr17.expr.gz" "~/temp/test_ss_chr18.expr.gz"</span>
<span class="co">#&gt; [19] "~/temp/test_ss_chr19.expr.gz" "~/temp/test_ss_chr20.expr.gz"</span>
<span class="co">#&gt; [21] "~/temp/test_ss_chr21.expr.gz" "~/temp/test_ss_chr22.expr.gz"</span></code></pre></div>
</div>
<div id="step-3-run-ctwas_rss-" class="section level1">
<h1 class="hasAnchor">
<a href="#step-3-run-ctwas_rss-" class="anchor"></a>step 3: Run <code>ctwas_rss</code>.</h1>
<p>In this step we will perform the causal TWAS algorithm, the algorithm will run susie iteratively for parameter estimation and lastly provide PIPs for all genes and SNPs included in the analysis. If you don’t want to define your own LD regions, then you can use the one defined by ldetect by simply specifying the population name using the <code>ld_regions</code> argument. One feature of the ctwas function is that it allows parallel computing. you can specify number of cores to use by the <code>ncore</code> argument.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">regionsfile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"example_regions.bed"</span>, package <span class="op">=</span> <span class="st">"ctwas"</span><span class="op">)</span>
<span class="va">pars</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ctwas_rss.html">ctwas_rss</a></span><span class="op">(</span><span class="va">zdf</span>, <span class="va">ld_pgenfs</span>, <span class="va">ld_exprfs</span>, ld_regions_custom <span class="op">=</span> <span class="va">regionsfile</span>, thin <span class="op">=</span> <span class="fl">1</span>, outputdir <span class="op">=</span> <span class="va">outputdir</span>, outname <span class="op">=</span> <span class="st">"test_ss"</span>, ncore <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>
<span class="co">#&gt; 2021-03-04 14:03:49 INFO::ctwas started ...</span>
<span class="co">#&gt; 2021-03-04 14:03:49 INFO::LD region file: /tmp/RtmpfUkBMG/temp_libpath6611c462696/ctwas/extdata/example_regions.bed</span>
<span class="co">#&gt; 2021-03-04 14:03:49 INFO::No. LD regions: 374</span>
<span class="co">#&gt; 2021-03-04 14:03:49 INFO::No. regions with at least one SNP/gene for chr1: 17</span>
<span class="co">#&gt; 2021-03-04 14:03:49 INFO::Merge regions for chr1</span>
<span class="co">#&gt; 2021-03-04 14:03:49 INFO::No. regions with at least one SNP/gene for chr1 after merging: 17</span>
<span class="co">#&gt; 2021-03-04 14:03:49 INFO::No. regions with at least one SNP/gene for chr2: 17</span>
<span class="co">#&gt; 2021-03-04 14:03:49 INFO::Merge regions for chr2</span>
<span class="co">#&gt; 2021-03-04 14:03:49 INFO::No. regions with at least one SNP/gene for chr2 after merging: 17</span>
<span class="co">#&gt; 2021-03-04 14:03:49 INFO::No. regions with at least one SNP/gene for chr3: 17</span>
<span class="co">#&gt; 2021-03-04 14:03:49 INFO::Merge regions for chr3</span>
<span class="co">#&gt; 2021-03-04 14:03:49 INFO::No. regions with at least one SNP/gene for chr3 after merging: 17</span>
<span class="co">#&gt; 2021-03-04 14:03:49 INFO::No. regions with at least one SNP/gene for chr4: 17</span>
<span class="co">#&gt; 2021-03-04 14:03:49 INFO::Merge regions for chr4</span>
<span class="co">#&gt; 2021-03-04 14:03:49 INFO::No. regions with at least one SNP/gene for chr4 after merging: 16</span>
<span class="co">#&gt; 2021-03-04 14:03:49 INFO::No. regions with at least one SNP/gene for chr5: 17</span>
<span class="co">#&gt; 2021-03-04 14:03:49 INFO::Merge regions for chr5</span>
<span class="co">#&gt; 2021-03-04 14:03:49 INFO::No. regions with at least one SNP/gene for chr5 after merging: 17</span>
<span class="co">#&gt; 2021-03-04 14:03:49 INFO::No. regions with at least one SNP/gene for chr6: 17</span>
<span class="co">#&gt; 2021-03-04 14:03:49 INFO::Merge regions for chr6</span>
<span class="co">#&gt; 2021-03-04 14:03:49 INFO::No. regions with at least one SNP/gene for chr6 after merging: 17</span>
<span class="co">#&gt; 2021-03-04 14:03:49 INFO::No. regions with at least one SNP/gene for chr7: 17</span>
<span class="co">#&gt; 2021-03-04 14:03:49 INFO::Merge regions for chr7</span>
<span class="co">#&gt; 2021-03-04 14:03:49 INFO::No. regions with at least one SNP/gene for chr7 after merging: 16</span>
<span class="co">#&gt; 2021-03-04 14:03:49 INFO::No. regions with at least one SNP/gene for chr8: 17</span>
<span class="co">#&gt; 2021-03-04 14:03:49 INFO::Merge regions for chr8</span>
<span class="co">#&gt; 2021-03-04 14:03:49 INFO::No. regions with at least one SNP/gene for chr8 after merging: 17</span>
<span class="co">#&gt; 2021-03-04 14:03:49 INFO::No. regions with at least one SNP/gene for chr9: 17</span>
<span class="co">#&gt; 2021-03-04 14:03:49 INFO::Merge regions for chr9</span>
<span class="co">#&gt; 2021-03-04 14:03:49 INFO::No. regions with at least one SNP/gene for chr9 after merging: 17</span>
<span class="co">#&gt; 2021-03-04 14:03:49 INFO::No. regions with at least one SNP/gene for chr10: 17</span>
<span class="co">#&gt; 2021-03-04 14:03:49 INFO::Merge regions for chr10</span>
<span class="co">#&gt; 2021-03-04 14:03:49 INFO::No. regions with at least one SNP/gene for chr10 after merging: 17</span>
<span class="co">#&gt; 2021-03-04 14:03:50 INFO::No. regions with at least one SNP/gene for chr11: 17</span>
<span class="co">#&gt; 2021-03-04 14:03:50 INFO::Merge regions for chr11</span>
<span class="co">#&gt; 2021-03-04 14:03:50 INFO::No. regions with at least one SNP/gene for chr11 after merging: 16</span>
<span class="co">#&gt; 2021-03-04 14:03:50 INFO::No. regions with at least one SNP/gene for chr12: 17</span>
<span class="co">#&gt; 2021-03-04 14:03:50 INFO::Merge regions for chr12</span>
<span class="co">#&gt; 2021-03-04 14:03:50 INFO::No. regions with at least one SNP/gene for chr12 after merging: 16</span>
<span class="co">#&gt; 2021-03-04 14:03:50 INFO::No. regions with at least one SNP/gene for chr13: 17</span>
<span class="co">#&gt; 2021-03-04 14:03:50 INFO::Merge regions for chr13</span>
<span class="co">#&gt; 2021-03-04 14:03:50 INFO::No. regions with at least one SNP/gene for chr13 after merging: 16</span>
<span class="co">#&gt; 2021-03-04 14:03:50 INFO::No. regions with at least one SNP/gene for chr14: 17</span>
<span class="co">#&gt; 2021-03-04 14:03:50 INFO::Merge regions for chr14</span>
<span class="co">#&gt; 2021-03-04 14:03:50 INFO::No. regions with at least one SNP/gene for chr14 after merging: 16</span>
<span class="co">#&gt; 2021-03-04 14:03:50 INFO::No. regions with at least one SNP/gene for chr15: 17</span>
<span class="co">#&gt; 2021-03-04 14:03:50 INFO::Merge regions for chr15</span>
<span class="co">#&gt; 2021-03-04 14:03:50 INFO::No. regions with at least one SNP/gene for chr15 after merging: 17</span>
<span class="co">#&gt; 2021-03-04 14:03:50 INFO::No. regions with at least one SNP/gene for chr16: 17</span>
<span class="co">#&gt; 2021-03-04 14:03:50 INFO::Merge regions for chr16</span>
<span class="co">#&gt; 2021-03-04 14:03:50 INFO::No. regions with at least one SNP/gene for chr16 after merging: 17</span>
<span class="co">#&gt; 2021-03-04 14:03:50 INFO::No. regions with at least one SNP/gene for chr17: 17</span>
<span class="co">#&gt; 2021-03-04 14:03:50 INFO::Merge regions for chr17</span>
<span class="co">#&gt; 2021-03-04 14:03:50 INFO::No. regions with at least one SNP/gene for chr17 after merging: 16</span>
<span class="co">#&gt; 2021-03-04 14:03:50 INFO::No. regions with at least one SNP/gene for chr18: 17</span>
<span class="co">#&gt; 2021-03-04 14:03:50 INFO::Merge regions for chr18</span>
<span class="co">#&gt; 2021-03-04 14:03:50 INFO::No. regions with at least one SNP/gene for chr18 after merging: 16</span>
<span class="co">#&gt; 2021-03-04 14:03:50 INFO::No. regions with at least one SNP/gene for chr19: 17</span>
<span class="co">#&gt; 2021-03-04 14:03:50 INFO::Merge regions for chr19</span>
<span class="co">#&gt; 2021-03-04 14:03:50 INFO::No. regions with at least one SNP/gene for chr19 after merging: 17</span>
<span class="co">#&gt; 2021-03-04 14:03:50 INFO::No. regions with at least one SNP/gene for chr20: 17</span>
<span class="co">#&gt; 2021-03-04 14:03:50 INFO::Merge regions for chr20</span>
<span class="co">#&gt; 2021-03-04 14:03:50 INFO::No. regions with at least one SNP/gene for chr20 after merging: 17</span>
<span class="co">#&gt; 2021-03-04 14:03:50 INFO::No. regions with at least one SNP/gene for chr21: 17</span>
<span class="co">#&gt; 2021-03-04 14:03:50 INFO::Merge regions for chr21</span>
<span class="co">#&gt; 2021-03-04 14:03:50 INFO::No. regions with at least one SNP/gene for chr21 after merging: 17</span>
<span class="co">#&gt; 2021-03-04 14:03:50 INFO::No. regions with at least one SNP/gene for chr22: 17</span>
<span class="co">#&gt; 2021-03-04 14:03:50 INFO::Merge regions for chr22</span>
<span class="co">#&gt; 2021-03-04 14:03:50 INFO::No. regions with at least one SNP/gene for chr22 after merging: 17</span>
<span class="co">#&gt; 2021-03-04 14:03:50 INFO::Run susie iteratively, getting rough estimate ...</span>
<span class="co">#&gt; 2021-03-04 14:03:50 INFO::run iteration 1</span>
<span class="co">#&gt; 2021-03-04 14:04:01 INFO::After iteration 1, gene prior 0.170771013961618:, SNP prior:0.0352160625214421</span>
<span class="co">#&gt; 2021-03-04 14:04:01 INFO::run iteration 2</span>
<span class="co">#&gt; 2021-03-04 14:04:12 INFO::After iteration 2, gene prior 0.0978925854786318:, SNP prior:0.0342703285454733</span>
<span class="co">#&gt; 2021-03-04 14:04:12 INFO::run iteration 3</span>
<span class="co">#&gt; 2021-03-04 14:04:24 INFO::After iteration 3, gene prior 0.0709623978338629:, SNP prior:0.0340568433278576</span>
<span class="co">#&gt; 2021-03-04 14:04:24 INFO::Blocks are filtered: 14 blocks left</span>
<span class="co">#&gt; 2021-03-04 14:04:24 INFO::Run susie iteratively, getting accurate estimate ...</span>
<span class="co">#&gt; 2021-03-04 14:04:24 INFO::run iteration 1</span>
<span class="co">#&gt; 2021-03-04 14:04:27 INFO::After iteration 1, gene prior 0.0922549265126892:, SNP prior:0.0242396402631829</span>
<span class="co">#&gt; 2021-03-04 14:04:27 INFO::run iteration 2</span>
<span class="co">#&gt; 2021-03-04 14:04:31 INFO::After iteration 2, gene prior 0.0936321739769035:, SNP prior:0.0214034800382348</span>
<span class="co">#&gt; 2021-03-04 14:04:31 INFO::run iteration 3</span>
<span class="co">#&gt; 2021-03-04 14:04:34 INFO::After iteration 3, gene prior 0.0938633693720293:, SNP prior:0.0200262928505434</span>
<span class="co">#&gt; 2021-03-04 14:04:34 INFO::run iteration 4</span>
<span class="co">#&gt; 2021-03-04 14:04:38 INFO::After iteration 4, gene prior 0.0939156377753681:, SNP prior:0.0192424959046532</span>
<span class="co">#&gt; 2021-03-04 14:04:38 INFO::run iteration 5</span>
<span class="co">#&gt; 2021-03-04 14:04:41 INFO::After iteration 5, gene prior 0.0939356324035656:, SNP prior:0.0187672758461306</span>
<span class="co">#&gt; 2021-03-04 14:04:41 INFO::run iteration 6</span>
<span class="co">#&gt; 2021-03-04 14:04:44 INFO::After iteration 6, gene prior 0.0939479739427962:, SNP prior:0.0184753869552136</span>
<span class="co">#&gt; 2021-03-04 14:04:44 INFO::run iteration 7</span>
<span class="co">#&gt; 2021-03-04 14:04:48 INFO::After iteration 7, gene prior 0.0939572297378678:, SNP prior:0.018302197689861</span>
<span class="co">#&gt; 2021-03-04 14:04:48 INFO::run iteration 8</span>
<span class="co">#&gt; 2021-03-04 14:04:51 INFO::After iteration 8, gene prior 0.093964568255844:, SNP prior:0.0182106789103659</span>
<span class="co">#&gt; 2021-03-04 14:04:51 INFO::run iteration 9</span>
<span class="co">#&gt; 2021-03-04 14:04:55 INFO::After iteration 9, gene prior 0.0939705041226744:, SNP prior:0.0181779988423127</span>
<span class="co">#&gt; 2021-03-04 14:04:55 INFO::run iteration 10</span>
<span class="co">#&gt; 2021-03-04 14:04:58 INFO::After iteration 10, gene prior 0.0939753641184598:, SNP prior:0.0181892874585848</span>
<span class="co">#&gt; 2021-03-04 14:04:58 INFO::run iteration 11</span>
<span class="co">#&gt; 2021-03-04 14:05:02 INFO::After iteration 11, gene prior 0.0939793822443231:, SNP prior:0.0182344387413349</span>
<span class="co">#&gt; 2021-03-04 14:05:02 INFO::run iteration 12</span>
<span class="co">#&gt; 2021-03-04 14:05:05 INFO::After iteration 12, gene prior 0.0939827316695113:, SNP prior:0.0183063456143277</span>
<span class="co">#&gt; 2021-03-04 14:05:05 INFO::run iteration 13</span>
<span class="co">#&gt; 2021-03-04 14:05:08 INFO::After iteration 13, gene prior 0.0939855424970787:, SNP prior:0.0183998677483807</span>
<span class="co">#&gt; 2021-03-04 14:05:08 INFO::run iteration 14</span>
<span class="co">#&gt; 2021-03-04 14:05:12 INFO::After iteration 14, gene prior 0.0939879137476652:, SNP prior:0.018511199064225</span>
<span class="co">#&gt; 2021-03-04 14:05:12 INFO::run iteration 15</span>
<span class="co">#&gt; 2021-03-04 14:05:15 INFO::After iteration 15, gene prior 0.093989921767648:, SNP prior:0.0186374648551941</span>
<span class="co">#&gt; 2021-03-04 14:05:16 INFO::run iteration 16</span>
<span class="co">#&gt; 2021-03-04 14:05:19 INFO::After iteration 16, gene prior 0.0939916261765161:, SNP prior:0.0187764566421134</span>
<span class="co">#&gt; 2021-03-04 14:05:19 INFO::run iteration 17</span>
<span class="co">#&gt; 2021-03-04 14:05:22 INFO::After iteration 17, gene prior 0.0939930741026556:, SNP prior:0.0189264527160328</span>
<span class="co">#&gt; 2021-03-04 14:05:22 INFO::run iteration 18</span>
<span class="co">#&gt; 2021-03-04 14:05:26 INFO::After iteration 18, gene prior 0.0939943032288664:, SNP prior:0.01908609368778</span>
<span class="co">#&gt; 2021-03-04 14:05:26 INFO::run iteration 19</span>
<span class="co">#&gt; 2021-03-04 14:05:29 INFO::After iteration 19, gene prior 0.0939953440069081:, SNP prior:0.0192542943240877</span>
<span class="co">#&gt; 2021-03-04 14:05:29 INFO::run iteration 20</span>
<span class="co">#&gt; 2021-03-04 14:05:33 INFO::After iteration 20, gene prior 0.0939962212872596:, SNP prior:0.01943017990068</span>
<span class="co">#&gt; 2021-03-04 14:05:33 INFO::run iteration 21</span>
<span class="co">#&gt; 2021-03-04 14:05:36 INFO::After iteration 21, gene prior 0.0939969555330488:, SNP prior:0.0196130394751356</span>
<span class="co">#&gt; 2021-03-04 14:05:36 INFO::run iteration 22</span>
<span class="co">#&gt; 2021-03-04 14:05:40 INFO::After iteration 22, gene prior 0.0939975637349052:, SNP prior:0.0198022910596808</span>
<span class="co">#&gt; 2021-03-04 14:05:40 INFO::run iteration 23</span>
<span class="co">#&gt; 2021-03-04 14:05:43 INFO::After iteration 23, gene prior 0.0939980601081758:, SNP prior:0.019997455307034</span>
<span class="co">#&gt; 2021-03-04 14:05:43 INFO::run iteration 24</span>
<span class="co">#&gt; 2021-03-04 14:05:47 INFO::After iteration 24, gene prior 0.093998456629938:, SNP prior:0.0201981353808248</span>
<span class="co">#&gt; 2021-03-04 14:05:47 INFO::run iteration 25</span>
<span class="co">#&gt; 2021-03-04 14:05:50 INFO::After iteration 25, gene prior 0.0939987634567644:, SNP prior:0.0204040013823012</span>
<span class="co">#&gt; 2021-03-04 14:05:50 INFO::run iteration 26</span>
<span class="co">#&gt; 2021-03-04 14:05:54 INFO::After iteration 26, gene prior 0.0939989892527785:, SNP prior:0.0206147781769054</span>
<span class="co">#&gt; 2021-03-04 14:05:54 INFO::run iteration 27</span>
<span class="co">#&gt; 2021-03-04 14:05:57 INFO::After iteration 27, gene prior 0.0939991414495369:, SNP prior:0.020830235787722</span>
<span class="co">#&gt; 2021-03-04 14:05:57 INFO::run iteration 28</span>
<span class="co">#&gt; 2021-03-04 14:06:01 INFO::After iteration 28, gene prior 0.0939992264536088:, SNP prior:0.0210501817479241</span>
<span class="co">#&gt; 2021-03-04 14:06:01 INFO::run iteration 29</span>
<span class="co">#&gt; 2021-03-04 14:06:05 INFO::After iteration 29, gene prior 0.0939992498136655:, SNP prior:0.0212744549632837</span>
<span class="co">#&gt; 2021-03-04 14:06:05 INFO::run iteration 30</span>
<span class="co">#&gt; 2021-03-04 14:06:08 INFO::After iteration 30, gene prior 0.0939992163559557:, SNP prior:0.021502920749501</span>
<span class="co">#&gt; 2021-03-04 14:06:08 INFO::Run susie for all regions.</span>
<span class="co">#&gt; 2021-03-04 14:06:08 INFO::run iteration 1</span>
<span class="co">#&gt; 2021-03-04 14:06:22 INFO::After iteration 1, gene prior 0.0978585128100543:, SNP prior:0.104147349326915</span></code></pre></div>
<p><code>ctwas_rss</code> returns the estimated parameters ( in the order of gene, SNP):</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pars</span>
<span class="co">#&gt; $group_prior</span>
<span class="co">#&gt; [1] 0.09399922 0.02150292</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $group_prior_var</span>
<span class="co">#&gt; [1] 301.4180820   0.2283366</span></code></pre></div>
<p>PIP results are given in <code>outname.susieIres.txt</code>. This file contains PIP for each gene and SNP, please check <code>susie_pip</code> column.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu">data.table</span><span class="fu">::</span><span class="fu"><a href="https://Rdatatable.gitlab.io/data.table/reference/fread.html">fread</a></span><span class="op">(</span><span class="st">"~/temp/test_ss.susieIrss.txt"</span><span class="op">)</span><span class="op">)</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"chrom"</span>, <span class="st">"id"</span>, <span class="st">"pos"</span>, <span class="st">"type"</span>, <span class="st">"susie_pip"</span><span class="op">)</span><span class="op">]</span>
<span class="co">#&gt;    chrom     id pos type   susie_pip</span>
<span class="co">#&gt; 1:     1  gene1   1 gene 1.000000000</span>
<span class="co">#&gt; 2:     1 rs1001   1  SNP 0.105157370</span>
<span class="co">#&gt; 3:     1 rs1002   2  SNP 0.087961560</span>
<span class="co">#&gt; 4:     1 rs1003   3  SNP 0.090608972</span>
<span class="co">#&gt; 5:     1 rs1004   4  SNP 0.102215921</span>
<span class="co">#&gt; 6:     1 rs1006   6  SNP 0.003594507</span></code></pre></div>
<p>You will also notice a few auxilary/intermediate files produced by <code>ctwas_rss</code>. These files can be useful for diagnoiss.<code>.Rd</code> files contains parameters estimation updates with each iteration.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span><span class="op">(</span><span class="op">)</span>
<span class="co">#&gt; R version 3.6.3 (2020-02-29)</span>
<span class="co">#&gt; Platform: x86_64-conda_cos6-linux-gnu (64-bit)</span>
<span class="co">#&gt; Running under: Scientific Linux 7.4 (Nitrogen)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Matrix products: default</span>
<span class="co">#&gt; BLAS:   /scratch/midway2/simingz/conda-envs/twas/lib/libblas.so.3.6.1</span>
<span class="co">#&gt; LAPACK: /scratch/midway2/simingz/conda-envs/twas/lib/liblapack.so.3.6.1</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; locale:</span>
<span class="co">#&gt;  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span>
<span class="co">#&gt;  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span>
<span class="co">#&gt;  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span>
<span class="co">#&gt;  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span>
<span class="co">#&gt;  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span>
<span class="co">#&gt; [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; attached base packages:</span>
<span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; other attached packages:</span>
<span class="co">#&gt; [1] ctwas_0.1.11</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; loaded via a namespace (and not attached):</span>
<span class="co">#&gt;  [1] Rcpp_1.0.5        compiler_3.6.3    pillar_1.4.6      iterators_1.0.12 </span>
<span class="co">#&gt;  [5] tools_3.6.3       digest_0.6.25     evaluate_0.14     memoise_1.1.0    </span>
<span class="co">#&gt;  [9] lifecycle_0.2.0   tibble_3.0.3      lattice_0.20-41   pkgconfig_2.0.3  </span>
<span class="co">#&gt; [13] rlang_0.4.7       Matrix_1.2-18     foreach_1.5.0     parallel_3.6.3   </span>
<span class="co">#&gt; [17] yaml_2.2.1        pkgdown_1.6.1     xfun_0.19         stringr_1.4.0    </span>
<span class="co">#&gt; [21] dplyr_1.0.0       knitr_1.29        pgenlibr_0.2      desc_1.2.0       </span>
<span class="co">#&gt; [25] generics_0.0.2    fs_1.4.2          vctrs_0.3.2       systemfonts_0.3.2</span>
<span class="co">#&gt; [29] rprojroot_2.0.2   grid_3.6.3        tidyselect_1.1.0  glue_1.4.1       </span>
<span class="co">#&gt; [33] data.table_1.12.8 R6_2.4.1          rmarkdown_2.6     purrr_0.3.4      </span>
<span class="co">#&gt; [37] magrittr_1.5      codetools_0.2-16  htmltools_0.5.0   ellipsis_0.3.1   </span>
<span class="co">#&gt; [41] assertthat_0.2.1  logging_0.10-108  ragg_0.3.1        stringi_1.4.6    </span>
<span class="co">#&gt; [45] doParallel_1.0.15 crayon_1.3.4</span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Siming Zhao.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
