<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Using cTWAS modules • ctwas</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Using cTWAS modules">
<meta property="og:description" content="ctwas">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ctwas</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.2.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorial
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/preparing_ctwas_input_data.html">Preparing cTWAS input data</a>
    </li>
    <li>
      <a href="../articles/ctwas_main_function.html">Using cTWAS main function</a>
    </li>
    <li>
      <a href="../articles/ctwas_modules.html">Using cTWAS modules</a>
    </li>
    <li>
      <a href="../articles/summarizing_ctwas_results.html">Summarizing cTWAS results</a>
    </li>
    <li>
      <a href="../articles/postprocessing_ctwas_results.html">Post-processing cTWAS results</a>
    </li>
    <li>
      <a href="../articles/FAQ.html">FAQ</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/xinhe-lab/ctwas/tree/multigroup_test" class="external-link">Source</a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="ctwas_modules_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Using cTWAS modules</h1>
                        <h4 data-toc-skip class="author">Kaixuan Luo, Sheng Qian</h4>
            
            <h4 data-toc-skip class="date">2024-05-10</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/xinhe-lab/ctwas/blob/multigroup_test/vignettes/ctwas_modules.Rmd" class="external-link"><code>vignettes/ctwas_modules.Rmd</code></a></small>
      <div class="hidden name"><code>ctwas_modules.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="using-ctwas-modules">Using cTWAS modules<a class="anchor" aria-label="anchor" href="#using-ctwas-modules"></a>
</h2>
<p>In this tutorial, we will show how to perform cTWAS analysis with summary statistics by running each of the cTWAS modules separately. This would allow more flexible controls of the cTWAS modules with different settings (memory, parallelization, etc.). We could save the results from each module for future access.</p>
<p>We require first running the data preprocessing and harmonization steps before running cTWAS main analysis.</p>
<p>Please follow the tutorial <a href="">Preparing cTWAS input data</a> to prepare cTWAS input data.</p>
<p>Load the packages</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/xinhe-lab/ctwas" class="external-link">ctwas</a></span><span class="op">)</span></span></code></pre></div>
<p>Load the prepared input data</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># output directory</span></span>
<span><span class="va">outputdir</span> <span class="op">&lt;-</span> <span class="st">"/project2/xinhe/shared_data/multigroup_ctwas/tutorial/LDL_multitissue_tutorial/output/"</span></span>
<span><span class="va">outname</span> <span class="op">&lt;-</span> <span class="st">"LDL.Liver_Adipose"</span></span>
<span></span>
<span><span class="co"># Load z_snp</span></span>
<span><span class="va">z_snp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">outputdir</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">outname</span>, <span class="st">".preprocessed.z_snp.RDS"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Load weights</span></span>
<span><span class="va">weights</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">outputdir</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">outname</span>, <span class="st">".preprocessed.weights.RDS"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Load region info</span></span>
<span><span class="va">region_info</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">outputdir</span>, <span class="st">"region_info.RDS"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Load gene z-scores</span></span>
<span><span class="va">z_gene</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">outputdir</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">outname</span>, <span class="st">".z_gene.RDS"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="assemble-input-data-for-the-regions">Assemble input data for the regions<a class="anchor" aria-label="anchor" href="#assemble-input-data-for-the-regions"></a>
</h3>
<p>After we have done data preprocessing and computed z-scores of molecular traits, we assemble the input data for all the regions using the <code><a href="../reference/assemble_region_data.html">assemble_region_data()</a></code> function .<br>
It assigns genes (molecular traits), SNPs and their z-score data for each region. <code>region_info</code> is a data frame with genome coordinates, LD matrix and SNP info of all the regions. <code>z_snp</code> is a data frame with preprocessed GWAS z-scores. <code>z_gene</code> is a data frame with computed z-scores of molecular traits. <code>weights</code> is a list of preprocessed weights from PredictDB or FUSION prediction models. The <code>thin</code> argument randomly selects a subset of variants (10% when thin = 0.1) to use during the parameter estimation and screening regions, reducing computation. If <code>thin = 1</code>, it will use all the SNPs. If <code>adjust_boundary_genes = TRUE</code>, it will identify the genes (molecular traits) that cross boundaries of multiple regions, and reassign them to one of the regions with the largest effect according to the eQTL effect sizes. <code>maxSNP</code> sets a maximum on the number of variants that can be in a single region to prevent memory issues during fine-mapping. If the number of SNPs in a region is more than <code>maxSNP</code>, It will trim the SNPs (randomly, by default) in the region. <code>ncore</code> specifies the number of cores to use when parallelizing over regions.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/assemble_region_data.html">assemble_region_data</a></span><span class="op">(</span><span class="va">region_info</span>, </span>
<span>                            <span class="va">z_snp</span>, </span>
<span>                            <span class="va">z_gene</span>, </span>
<span>                            <span class="va">weights</span>,</span>
<span>                            thin <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>                            maxSNP <span class="op">=</span> <span class="fl">20000</span>,</span>
<span>                            adjust_boundary_genes <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                            ncore <span class="op">=</span> <span class="va">ncore</span><span class="op">)</span></span>
<span><span class="va">region_data</span> <span class="op">&lt;-</span> <span class="va">res</span><span class="op">$</span><span class="va">region_data</span></span>
<span><span class="va">boundary_genes</span> <span class="op">&lt;-</span> <span class="va">res</span><span class="op">$</span><span class="va">boundary_genes</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS</a></span><span class="op">(</span><span class="va">res</span>, <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">outputdir</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">outname</span>, <span class="st">".region_data.thin"</span>, <span class="va">thin</span>, <span class="st">".RDS"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>This functions returns a list with <code>region_data</code>, assembled input data for all the regions, as well as a data frame containing the region information for each of cross-boundary genes (or molecular traits).</p>
</div>
<div class="section level3">
<h3 id="estimating-parameters">Estimating parameters<a class="anchor" aria-label="anchor" href="#estimating-parameters"></a>
</h3>
<p>We use the <code><a href="../reference/est_param.html">est_param()</a></code> function to estimate two sets of parameters: the prior inclusion probabilities and the prior effect size variance. It will take the assembled <code>region_data</code> as input.</p>
<p>There are several options for how to handle the prior effect size parameters by specifying the <code>group_prior_var_stucture</code>:</p>
<ul>
<li>“shared_type” (default option) allows all groups in one molecular QTL type to share the same variance parameter.</li>
<li>“shared_context” allows all groups in one molecular QTL type to share the same variance parameter.</li>
<li>“shared_nonSNP” allows all non-SNP groups to share the same variance parameter.</li>
<li>“shared_all” allows all groups to share the same variance parameter.</li>
<li>“independent” allows all groups to have their own separate variance parameters.</li>
</ul>
<p>This step will run the EM algorithm to estimate parameters and return estimate parameters (<code>group_prior</code>, <code>group_prior_var</code>, etc.). It will run two rounds of EM algorithm, first round uses fewer iterations (<code>niter_prefit=3</code> by default) to get rough parameter estimates, and using the estimated priors in the first round, we select regions with single effect to run the second round. We use more iterations in the second round (<code>niter=30</code> by default) to get accurate parameter estimates.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">param</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/est_param.html">est_param</a></span><span class="op">(</span><span class="va">region_data</span>, </span>
<span>                   group_prior_var_structure <span class="op">=</span> <span class="st">"shared_type"</span>, </span>
<span>                   niter_prefit <span class="op">=</span> <span class="fl">3</span>,</span>
<span>                   niter <span class="op">=</span> <span class="fl">30</span>, </span>
<span>                   ncore <span class="op">=</span> <span class="va">ncore</span><span class="op">)</span></span>
<span><span class="va">group_prior</span> <span class="op">&lt;-</span> <span class="va">param</span><span class="op">$</span><span class="va">group_prior</span></span>
<span><span class="va">group_prior_var</span> <span class="op">&lt;-</span> <span class="va">param</span><span class="op">$</span><span class="va">group_prior_var</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS</a></span><span class="op">(</span><span class="va">param</span>, <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">outputdir</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">outname</span>, <span class="st">".param.RDS"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="screening-regions">Screening regions<a class="anchor" aria-label="anchor" href="#screening-regions"></a>
</h3>
<p>After parameter estimation, we use the <code><a href="../reference/screen_regions.html">screen_regions()</a></code> function to perform a screening process to fine-map all regions using thinned variants, and select regions with strong signals (total non-SNP PIP greater than 0.5, by default). It returns a data frame with region IDs and nonSNP PIPs for each region. We could further narrow down the list of regions by using more stringent <code>min_nonSNP_PIP</code> cutoffs.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">screen_regions_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/screen_regions.html">screen_regions</a></span><span class="op">(</span><span class="va">region_data</span>,</span>
<span>                                     <span class="va">region_info</span>,</span>
<span>                                     <span class="va">weights</span>,</span>
<span>                                     L <span class="op">=</span> <span class="fl">5</span>,</span>
<span>                                     group_prior <span class="op">=</span> <span class="va">group_prior</span>,</span>
<span>                                     group_prior_var <span class="op">=</span> <span class="va">group_prior_var</span>,</span>
<span>                                     min_nonSNP_PIP <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>                                     ncore <span class="op">=</span> <span class="va">ncore</span>,</span>
<span>                                     verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">screened_region_data</span> <span class="op">&lt;-</span> <span class="va">screen_regions_res</span><span class="op">$</span><span class="va">screened_region_data</span></span>
<span><span class="va">region_nonSNP_PIP_df</span> <span class="op">&lt;-</span> <span class="va">screen_regions_res</span><span class="op">$</span><span class="va">region_nonSNP_PIP_df</span></span></code></pre></div>
<p>After selecting regions with strong signals, we expand the screened regions with full sets of SNPs for final finemapping. If the number of SNPs in a region is more than <code>maxSNP</code>, It will trim the SNPs in the region (according to the z-scores, by default). This step is not needed if <code>thin = 1</code>.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="va">thin</span> <span class="op">&lt;</span> <span class="fl">1</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">screened_region_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/expand_region_data.html">expand_region_data</a></span><span class="op">(</span><span class="va">screened_region_data</span>,</span>
<span>                                             <span class="va">region_info</span>,</span>
<span>                                             <span class="va">z_snp</span>,</span>
<span>                                             <span class="va">z_gene</span>,</span>
<span>                                             trim_by <span class="op">=</span> <span class="st">"z"</span>,</span>
<span>                                             maxSNP <span class="op">=</span> <span class="fl">20000</span>,</span>
<span>                                             ncore <span class="op">=</span> <span class="va">ncore</span><span class="op">)</span></span>
<span></span>
<span>  <span class="co"># update data in screened regions with screened_region_data</span></span>
<span>  <span class="va">region_data</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">screened_region_data</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">screened_region_data</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS</a></span><span class="op">(</span><span class="va">region_data</span>, <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">outputdir</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">outname</span>, <span class="st">".region_data.RDS"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS</a></span><span class="op">(</span><span class="va">screened_region_data</span>, <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">outputdir</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">outname</span>, <span class="st">".screened_region_data.RDS"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="fine-mapping-screened-regions">Fine-mapping screened regions<a class="anchor" aria-label="anchor" href="#fine-mapping-screened-regions"></a>
</h3>
<p>We now perform the fine-mapping step for each of the screened regions with the estimated parameters.</p>
<p>The <code><a href="../reference/finemap_regions.html">finemap_regions()</a></code> function performs fine-mapping for the regions in the <code>screened_region_data</code>.</p>
<p>It first computes correlation matrices (SNP-SNP, SNP-gene, and gene-gene correlation matrices), and then runs fine-mapping with estimated parameters (<code>group_prior</code> and <code>group_prior_var</code>), and returns a data frame with fine-mapping results for the regions.</p>
<p>If <code>save_cor = TRUE</code>, it will save the computed correlation matrices to <code>cor_dir</code> for future access.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">finemap_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/finemap_regions.html">finemap_regions</a></span><span class="op">(</span><span class="va">screened_region_data</span>,</span>
<span>                               <span class="va">region_info</span>,</span>
<span>                               <span class="va">weights</span>,</span>
<span>                               group_prior <span class="op">=</span> <span class="va">group_prior</span>,</span>
<span>                               group_prior_var <span class="op">=</span> <span class="va">group_prior_var</span>,</span>
<span>                               L <span class="op">=</span> <span class="fl">5</span>,</span>
<span>                               save_cor <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                               cor_dir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">outputdir</span>, <span class="st">"cor_matrix"</span><span class="op">)</span>,</span>
<span>                               ncore <span class="op">=</span> <span class="va">ncore</span>,</span>
<span>                               verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS</a></span><span class="op">(</span><span class="va">finemap_res</span>, <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">outputdir</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">outname</span>, <span class="st">".ctwas.res.RDS"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="fine-mapping-a-single-region">Fine-mapping a single region<a class="anchor" aria-label="anchor" href="#fine-mapping-a-single-region"></a>
</h3>
<p>We could also run finemapping for a single region (or multiple regions) with precomputed parameters.</p>
<p>Here we show an example for finemapping a single region “16:71020125-72901251”.</p>
<p>We will first expand the region of interest with a full set of SNPs using the preassembled <code>region_data</code>.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">region_id</span> <span class="op">&lt;-</span> <span class="st">"16:71020125-72901251"</span></span>
<span><span class="va">selected_region_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/expand_region_data.html">expand_region_data</a></span><span class="op">(</span><span class="va">region_data</span><span class="op">[</span><span class="va">region_id</span><span class="op">]</span>,</span>
<span>                                           <span class="va">region_info</span>,</span>
<span>                                           <span class="va">z_snp</span>,</span>
<span>                                           <span class="va">z_gene</span>,</span>
<span>                                           maxSNP <span class="op">=</span> <span class="fl">20000</span>,</span>
<span>                                           trim_by <span class="op">=</span> <span class="st">"z"</span><span class="op">)</span></span></code></pre></div>
<p>We could then perform finemapping for this region of interest:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">finemap_region_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/finemap_regions.html">finemap_regions</a></span><span class="op">(</span><span class="va">selected_region_data</span>,</span>
<span>                                      <span class="va">region_info</span>,</span>
<span>                                      <span class="va">weights</span>,</span>
<span>                                      group_prior <span class="op">=</span> <span class="va">group_prior</span>,</span>
<span>                                      group_prior_var <span class="op">=</span> <span class="va">group_prior_var</span>,</span>
<span>                                      L <span class="op">=</span> <span class="fl">5</span>,</span>
<span>                                      save_cor <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                                      cor_dir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">outputdir</span>, <span class="st">"cor_matrix"</span><span class="op">)</span>,</span>
<span>                                      verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="running-ctwas-without-ld-matrices">Running cTWAS without LD matrices<a class="anchor" aria-label="anchor" href="#running-ctwas-without-ld-matrices"></a>
</h2>
<p>In case you do not have LD matrices, it is possible to run cTWAS without LD matrices.</p>
<p>To do this, simply set <code>L = 1</code> in <code><a href="../reference/screen_regions.html">screen_regions()</a></code> and <code><a href="../reference/finemap_regions.html">finemap_regions()</a></code> and also set <code>cor_dir = NULL</code> in <code><a href="../reference/finemap_regions.html">finemap_regions()</a></code> . Then it will run these steps with <code>L = 1</code> without LD matrices.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Siming Zhao, Wesley Crouse, Sheng Qian, Kaixuan Luo, Matthew Stephens, Xin He.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
