<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Using cTWAS main function • ctwas</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Using cTWAS main function">
<meta property="og:description" content="ctwas">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ctwas</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.2.46</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorial
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/simple_ctwas_tutorial.html">Simple cTWAS tutorial</a>
    </li>
    <li>
      <a href="../articles/preparing_ctwas_input_data.html">Preparing cTWAS input data</a>
    </li>
    <li>
      <a href="../articles/ctwas_main_function.html">Using cTWAS main function</a>
    </li>
    <li>
      <a href="../articles/ctwas_modules.html">Using cTWAS modules</a>
    </li>
    <li>
      <a href="../articles/summarizing_ctwas_results.html">Summarizing cTWAS results</a>
    </li>
    <li>
      <a href="../articles/postprocessing_ctwas_results.html">Post-processing cTWAS results</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/xinhe-lab/ctwas/tree/multigroup_test" class="external-link">Source</a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="ctwas_main_function_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Using cTWAS main function</h1>
                        <h4 data-toc-skip class="author">Kaixuan Luo, Sheng Qian</h4>
            
            <h4 data-toc-skip class="date">2024-06-28</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/xinhe-lab/ctwas/blob/multigroup_test/vignettes/ctwas_main_function.Rmd" class="external-link"><code>vignettes/ctwas_main_function.Rmd</code></a></small>
      <div class="hidden name"><code>ctwas_main_function.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="using-ctwas-main-function">Using cTWAS main function<a class="anchor" aria-label="anchor" href="#using-ctwas-main-function"></a>
</h2>
<p>In this tutorial, we will show how to perform cTWAS main analysis with summary statistics using the <code><a href="../reference/ctwas_sumstats.html">ctwas_sumstats()</a></code> function.</p>
<p>We require first running the data preprocessing and harmonization steps before running cTWAS main analysis.</p>
<p>Please follow the “Preparing cTWAS input data” tutorial to prepare cTWAS input data.</p>
<p>Load the packages</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/xinhe-lab/ctwas/" class="external-link">ctwas</a></span><span class="op">)</span></span></code></pre></div>
<p>Load the prepared input data.</p>
<p>Note: you would need to update the paths to LD matrices in LD_info.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load z_snp</span></span>
<span><span class="va">z_snp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/sample_data"</span>, <span class="st">"LDL_example.preprocessed.z_snp.RDS"</span>, package <span class="op">=</span> <span class="st">"ctwas"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Load weights</span></span>
<span><span class="va">weights</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/sample_data"</span>, <span class="st">"LDL_example.preprocessed.weights.RDS"</span>, package <span class="op">=</span> <span class="st">"ctwas"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Load region_info, and snp_info</span></span>
<span><span class="va">region_info</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/sample_data"</span>, <span class="st">"LDL_example.region_info.RDS"</span>, package <span class="op">=</span> <span class="st">"ctwas"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">snp_info</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/sample_data"</span>, <span class="st">"LDL_example.snp_info.RDS"</span>, package <span class="op">=</span> <span class="st">"ctwas"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Load LD_info</span></span>
<span><span class="co"># need to update the paths to LD matrices in LD_info</span></span>
<span><span class="va">LD_info</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/sample_data"</span>, <span class="st">"LDL_example.LD_info.RDS"</span>, package <span class="op">=</span> <span class="st">"ctwas"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Let’s set the cTWAS output directory and output name, and we use 6 cores to parallelize computing over regions.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">outputdir</span> <span class="op">&lt;-</span> <span class="st">"./ctwas_output"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="va">outputdir</span>, showWarnings<span class="op">=</span><span class="cn">F</span>, recursive<span class="op">=</span><span class="cn">T</span><span class="op">)</span></span>
<span></span>
<span><span class="va">outname</span> <span class="op">&lt;-</span> <span class="st">"LDL_example"</span></span>
<span><span class="va">ncore</span> <span class="op">&lt;-</span> <span class="fl">6</span></span></code></pre></div>
<p>The <code><a href="../reference/ctwas_sumstats.html">ctwas_sumstats()</a></code> function is a main function to run cTWAS analysis with summary statistics. It takes preprocessed GWAS z-scores (<code>z_snp</code>), <code>weights</code>, <code>region_info</code>, and <code>snp_info</code> as input, and will perform the main steps: computing z_scores of molecular traits, assembling input <code>region_data</code>, estimating parameters, screening regions, and finemapping. If <code>z_gene</code> is already computed when preparing cTWAS input data, we could specify the <code>z_gene</code> argument in the function, then it will skip computing z_scores of molecular traits.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cor_dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">outputdir</span>, <span class="st">"cor_matrix"</span><span class="op">)</span></span>
<span><span class="va">log_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">outputdir</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">outname</span>, <span class="st">".ctwas_sumstats.log"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">ctwas_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ctwas_sumstats.html">ctwas_sumstats</a></span><span class="op">(</span><span class="va">z_snp</span>,</span>
<span>                            <span class="va">weights</span>,</span>
<span>                            region_info <span class="op">=</span> <span class="va">region_info</span>,</span>
<span>                            snp_info <span class="op">=</span> <span class="va">snp_info</span>,</span>
<span>                            LD_info <span class="op">=</span> <span class="va">LD_info</span>,</span>
<span>                            thin <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>                            niter_prefit <span class="op">=</span> <span class="fl">3</span>,</span>
<span>                            niter <span class="op">=</span> <span class="fl">30</span>,</span>
<span>                            L <span class="op">=</span> <span class="fl">5</span>,</span>
<span>                            group_prior_var_structure <span class="op">=</span> <span class="st">"shared_type"</span>,</span>
<span>                            screen_method <span class="op">=</span> <span class="st">"nonSNP_PIP"</span>,</span>
<span>                            maxSNP <span class="op">=</span> <span class="fl">20000</span>,</span>
<span>                            min_nonSNP_PIP <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>                            ncore <span class="op">=</span> <span class="va">ncore</span>,</span>
<span>                            save_cor <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                            cor_dir <span class="op">=</span> <span class="va">cor_dir</span>,</span>
<span>                            outputdir <span class="op">=</span> <span class="va">outputdir</span>,</span>
<span>                            outname <span class="op">=</span> <span class="va">outname</span>,</span>
<span>                            logfile <span class="op">=</span> <span class="va">log_file</span>,</span>
<span>                            verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">param</span> <span class="op">&lt;-</span> <span class="va">ctwas_res</span><span class="op">$</span><span class="va">param</span></span>
<span><span class="va">finemap_res</span> <span class="op">&lt;-</span> <span class="va">ctwas_res</span><span class="op">$</span><span class="va">finemap_res</span></span>
<span><span class="va">boundary_genes</span> <span class="op">&lt;-</span> <span class="va">ctwas_res</span><span class="op">$</span><span class="va">boundary_genes</span></span>
<span><span class="va">z_gene</span> <span class="op">&lt;-</span> <span class="va">ctwas_res</span><span class="op">$</span><span class="va">z_gene</span></span>
<span><span class="va">region_data</span> <span class="op">&lt;-</span> <span class="va">ctwas_res</span><span class="op">$</span><span class="va">region_data</span></span>
<span><span class="va">screened_region_data</span> <span class="op">&lt;-</span> <span class="va">ctwas_res</span><span class="op">$</span><span class="va">screened_region_data</span></span>
<span></span>
<span><span class="co"># saveRDS(ctwas_res, file.path(outputdir, paste0(outname, ".ctwas_sumstats_res.RDS")))</span></span></code></pre></div>
<p>Major argument settings:</p>
<ul>
<li>
<code>z_snp</code> is a data frame with preprocessed GWAS z-scores.</li>
<li>
<code>weights</code> is a list of preprocessed weights from PredictDB or FUSION prediction models.</li>
<li>
<code>region_info</code> is a data frame of region definitions.</li>
<li>
<code>snp_info</code> is a list of SNP info data frames for LD reference.</li>
<li>
<code>LD_info</code> is a list of paths to LD matrices for each of the regions.</li>
<li>The <code>thin</code> argument randomly selects a subset of variants (10% when thin = 0.1) to use during the parameter estimation and screening regions, reducing computation.</li>
<li>The <code>niter_prefit</code> argument and <code>niter</code> argument sets the number of EM iterations for estimating parameters. We set <code>niter_prefit = 3</code>, and <code>niter = 30</code> by default.</li>
<li>The <code>L</code> argument sets the number of effects for SuSiE during the screening regions and fine mapping steps. We set <code>L = 5</code> by default.</li>
<li>There are several options for how to handle the prior effect size parameters by specifying the <code>group_prior_var_stucture</code>. “shared_type” (default option) allows all groups in one molecular QTL type to share the same variance parameter. “shared_context” allows all groups in one context (tissue, cell type, condition) to share the same variance parameter. “shared_nonSNP” allows all non-SNP groups to share the same variance parameter. “shared_all” allows all groups to share the same variance parameter. “independent” allows all groups to have their own separate variance parameters.</li>
<li>We can choose the method for screening regions by specifying the <code>screen_method</code> argument: “nonSNP_PIP” (by total non-SNP PIPs in a region) or “cs” (by the number of credible sets in a region).</li>
<li>The <code>maxSNP</code> sets a maximum on the number of variants that can be in a single region to prevent memory issues during fine-mapping.</li>
<li>
<code>min_nonSNP_PIP</code> is a threshold for selecting regions with strong non-SNP signals for finemapping. We set <code>min_nonSNP_PIP = 0.5</code> by default, so regions with total non-SNP signals greater than 0.5 will be selected for finemapping.</li>
<li>If the <code>save_cor</code> argument is TRUE, it will save the correlation matrices (SNP-SNP, SNP-gene, and gene-gene correlation matrices) in the <code>cor_dir</code> directory for future access.</li>
<li>If <code>outputdir</code> is not NULL, it will save the output from each step as a <code>.RDS</code> file to the <code>outputdir</code> directory. The <code>.RDS</code> files will have <code>outname</code> as the prefix.</li>
<li>The <code>ncore</code> argument specifies the number of cores to use when parallelizing over regions.</li>
<li>If the <code>logfile</code> argument is specified, it will write log messages to the log file.</li>
<li>If the <code>verbose</code> argument is TRUE, it will print more detailed messages.</li>
</ul>
<p>The function returns a list including: estimated parameters, fine-mapping results, boundary genes, z_scores of molecular traits, assembled region data for all the regions, and region data with full SNPs for screened regions.</p>
</div>
<div class="section level2">
<h2 id="running-ctwas-without-ld-matrices">Running cTWAS without LD matrices<a class="anchor" aria-label="anchor" href="#running-ctwas-without-ld-matrices"></a>
</h2>
<p>In case you do not have LD matrices, it is possible to run cTWAS without LD matrices.</p>
<p>To do this, we could use the main function <code><a href="../reference/ctwas_sumstats_noLD.html">ctwas_sumstats_noLD()</a></code>, and do not specify <code>LD_info</code>. Then it will run the main function with <code>L = 1</code> throughout the steps without LD matrices.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Siming Zhao, Wesley Crouse, Sheng Qian, Kaixuan Luo, Matthew Stephens, Xin He.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
