<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Data preparation • ctwas</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Data preparation">
<meta property="og:description" content="ctwas">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ctwas</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.36</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorial
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/data_preparation.html">Data preparation</a>
    </li>
    <li>
      <a href="../articles/postprocessing_regions.html">Postprocessing problematic regions</a>
    </li>
    <li>
      <a href="../articles/ctwas_summary_statistics.html">Using cTWAS single-group version with summary statistics</a>
    </li>
    <li>
      <a href="../articles/multigroup_ctwas_summary_statistics.html">Using cTWAS multi-group version with summary statistics</a>
    </li>
    <li>
      <a href="../articles/harmonization.html">Harmonization</a>
    </li>
    <li>
      <a href="../articles/regionmerging.html">Region Merging</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/xinhe-lab/ctwas" class="external-link">Source</a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="data_preparation_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Data preparation</h1>
                        <h4 data-toc-skip class="author">Kaixuan Luo</h4>
            
            <h4 data-toc-skip class="date">2024-02-13</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/xinhe-lab/ctwas/blob/main/vignettes/data_preparation.Rmd" class="external-link"><code>vignettes/data_preparation.Rmd</code></a></small>
      <div class="hidden name"><code>data_preparation.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p>This document is a tutorial for preparing and preprocessing input data for cTWAS with summary statistics. We will use the <code>multigroup</code> version of the package, but the procedure below works for using a single prediction model, or multiple prediction models.</p>
</div>
<div class="section level2">
<h2 id="getting-started">Getting started<a class="anchor" aria-label="anchor" href="#getting-started"></a>
</h2>
<p>Install the <code>multigroup_kevin</code> branch of <code>cTWAS</code> package</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/remotes/man/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"xinhe-lab/ctwas"</span>, ref <span class="op">=</span> <span class="st">"multigroup_kevin"</span><span class="op">)</span></span></code></pre></div>
<p>Load the package and set the working directory where you want to perform the analysis.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/xinhe-lab/ctwas" class="external-link">ctwas</a></span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># set the working directory interactively</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/getwd.html" class="external-link">setwd</a></span><span class="op">(</span><span class="st">"/project2/xinhe/shared_data/ctwas_tutorial"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># set the working directory when compiling this document with Knitr</span></span>
<span><span class="fu">knitr</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/knitr/man/opts_knit.html" class="external-link">opts_knit</a></span><span class="op">$</span><span class="fu">set</span><span class="op">(</span>root.dir <span class="op">=</span> <span class="st">"/project2/xinhe/shared_data/ctwas_tutorial"</span><span class="op">)</span></span></code></pre></div>
<p>The inputs for the summary statistics version of cTWAS include GWAS summary statistics for variants, prediction models for genes in PredictDB format, and LD reference.</p>
<p>We used the directories and files on the University of Chicago RCC cluster as examples. If you are at UChicago, you can load those data from RCC as below.</p>
</div>
<div class="section level2">
<h2 id="gwas-z-scores">GWAS z-scores<a class="anchor" aria-label="anchor" href="#gwas-z-scores"></a>
</h2>
<p>For this analysis, we will use summary statistics from a GWAS of LDL cholesterol in the UK Biobank. We will download the VCF from the IEU Open GWAS Project.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># set the working directory, download the summary statistics, and unzip the file.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="st">"gwas_summary_stats"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/system.html" class="external-link">system</a></span><span class="op">(</span><span class="st">"wget https://gwas.mrcieu.ac.uk/files/ukb-d-30780_irnt/ukb-d-30780_irnt.vcf.gz -P gwas_summary_stats"</span><span class="op">)</span></span>
<span><span class="fu">R.utils</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/R.utils/man/compressFile.html" class="external-link">gunzip</a></span><span class="op">(</span><span class="st">"gwas_summary_stats/ukb-d-30780_irnt.vcf.gz"</span><span class="op">)</span></span></code></pre></div>
<p>Next, we will read the summary statistics. Then, we will compute the z-scores and format the input data. We will also collect the sample size, which will be useful later. We will save this output for convenience.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># read the data using the VariantAnnotation package</span></span>
<span><span class="va">z_snp</span> <span class="op">&lt;-</span> <span class="fu">VariantAnnotation</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/VariantAnnotation/man/readVcf-methods.html" class="external-link">readVcf</a></span><span class="op">(</span><span class="st">"gwas_summary_stats/ukb-d-30780_irnt.vcf"</span><span class="op">)</span></span>
<span><span class="va">z_snp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="fu">gwasvcf</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/gwasvcf/man/vcf_to_tibble.html" class="external-link">vcf_to_tibble</a></span><span class="op">(</span><span class="va">z_snp</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># compute the z-scores</span></span>
<span><span class="va">z_snp</span><span class="op">$</span><span class="va">Z</span> <span class="op">&lt;-</span> <span class="va">z_snp</span><span class="op">$</span><span class="va">ES</span><span class="op">/</span><span class="va">z_snp</span><span class="op">$</span><span class="va">SE</span></span>
<span></span>
<span><span class="co"># collect sample size (most frequent sample size for all variants)</span></span>
<span><span class="va">gwas_n</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">z_snp</span><span class="op">$</span><span class="va">SS</span><span class="op">)</span>,decreasing<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># subset the columns and format the column names</span></span>
<span><span class="va">z_snp</span> <span class="op">&lt;-</span> <span class="va">z_snp</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"rsid"</span>, <span class="st">"ALT"</span>, <span class="st">"REF"</span>, <span class="st">"Z"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">z_snp</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"id"</span>, <span class="st">"A1"</span>, <span class="st">"A2"</span>, <span class="st">"z"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># drop multiallelic variants (id not unique)</span></span>
<span><span class="va">z_snp</span> <span class="op">&lt;-</span> <span class="va">z_snp</span><span class="op">[</span><span class="op">!</span><span class="op">(</span><span class="va">z_snp</span><span class="op">$</span><span class="va">id</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">z_snp</span><span class="op">$</span><span class="va">id</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/duplicated.html" class="external-link">duplicated</a></span><span class="op">(</span><span class="va">z_snp</span><span class="op">$</span><span class="va">id</span><span class="op">)</span><span class="op">]</span><span class="op">)</span>,<span class="op">]</span></span>
<span></span>
<span><span class="co"># save the formatted z-scores and GWAS sample size</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS</a></span><span class="op">(</span><span class="va">z_snp</span>, file<span class="op">=</span><span class="st">"gwas_summary_stats/ukb-d-30780_irnt.RDS"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS</a></span><span class="op">(</span><span class="va">gwas_n</span>, file<span class="op">=</span><span class="st">"gwas_summary_stats/gwas_n.RDS"</span><span class="op">)</span></span></code></pre></div>
<p>After the previous step, we can load the data and look at the format. <code>z_snp</code> is a data frame, and each row is a variant. <code>A1</code> is the alternate allele, and <code>A2</code> is the reference allele. The sample size for this GWAS is <code>N=343,621</code>.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">z_snp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"gwas_summary_stats/ukb-d-30780_irnt.RDS"</span><span class="op">)</span></span>
<span><span class="va">gwas_n</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">"gwas_summary_stats/gwas_n.RDS"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">z_snp</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># sample size</span></span>
<span><span class="va">gwas_n</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="prediction-models">Prediction models<a class="anchor" aria-label="anchor" href="#prediction-models"></a>
</h2>
<p>The <code>multigroup</code> version of cTWAS only accepts multiple sets of prediction models in PredictDB format. A single set of prediction models can still be specified in FUSION format, as in the <code>main</code> branch. See the section below on converting between FUSION and PredictDB weights for guidance on using multiple sets of FUSION weights.</p>
<p>To specify weights in PredictDB format, provide the path to the <code>.db</code> file. Multiple prediction models can be specified by providing multiple paths as a vector.</p>
<p>For this analysis, we will use liver and subcutaneous adipose gene expression models trained on GTEx v8 in the PredictDB format. We will download both the prediction models (<code>.db</code>) and the covariances between variants in the prediction models (<code>.txt.gz</code>). The covariances can optionally be used during harmonization to recover strand ambiguous variants.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># download the files</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/system.html" class="external-link">system</a></span><span class="op">(</span><span class="st">"wget https://zenodo.org/record/3518299/files/mashr_eqtl.tar"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># extract to ./weights folder </span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/system.html" class="external-link">system</a></span><span class="op">(</span><span class="st">"mkdir weights"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/system.html" class="external-link">system</a></span><span class="op">(</span><span class="st">"tar -xvf mashr_eqtl.tar -C weights"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/system.html" class="external-link">system</a></span><span class="op">(</span><span class="st">"rm mashr_eqtl.tar"</span><span class="op">)</span></span></code></pre></div>
<p>We also remove lncRNAs from the prediction models.</p>
<p>Process liver weights:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://rsqlite.r-dbi.org" class="external-link">RSQLite</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># specify the weight to remove lncRNA from</span></span>
<span><span class="va">weight</span> <span class="op">&lt;-</span> <span class="st">"weights/eqtl/mashr/mashr_Liver.db"</span></span>
<span></span>
<span><span class="co"># read the PredictDB weights</span></span>
<span><span class="va">sqlite</span> <span class="op">&lt;-</span> <span class="fu">dbDriver</span><span class="op">(</span><span class="st">"SQLite"</span><span class="op">)</span></span>
<span><span class="va">db</span> <span class="op">=</span> <span class="fu">dbConnect</span><span class="op">(</span><span class="va">sqlite</span>, <span class="va">weight</span><span class="op">)</span></span>
<span><span class="va">query</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">...</span><span class="op">)</span> <span class="fu">dbGetQuery</span><span class="op">(</span><span class="va">db</span>, <span class="va">...</span><span class="op">)</span></span>
<span><span class="va">weights_table</span> <span class="op">&lt;-</span> <span class="fu">query</span><span class="op">(</span><span class="st">"select * from weights"</span><span class="op">)</span></span>
<span><span class="va">extra_table</span> <span class="op">&lt;-</span> <span class="fu">query</span><span class="op">(</span><span class="st">"select * from extra"</span><span class="op">)</span></span>
<span><span class="fu">dbDisconnect</span><span class="op">(</span><span class="va">db</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># subset to protein coding genes only</span></span>
<span><span class="va">extra_table</span> <span class="op">&lt;-</span>  <span class="va">extra_table</span><span class="op">[</span><span class="va">extra_table</span><span class="op">$</span><span class="va">gene_type</span><span class="op">==</span><span class="st">"protein_coding"</span>,,drop<span class="op">=</span><span class="cn">F</span><span class="op">]</span></span>
<span><span class="va">weights_table</span> <span class="op">&lt;-</span> <span class="va">weights_table</span><span class="op">[</span><span class="va">weights_table</span><span class="op">$</span><span class="va">gene</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">extra_table</span><span class="op">$</span><span class="va">gene</span>,<span class="op">]</span></span>
<span></span>
<span><span class="co"># read and subset the covariances</span></span>
<span><span class="va">weight_info</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/connections.html" class="external-link">gzfile</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="fu">tools</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/tools/fileutils.html" class="external-link">file_path_sans_ext</a></span><span class="op">(</span><span class="va">weight</span><span class="op">)</span>, <span class="st">".txt.gz"</span><span class="op">)</span><span class="op">)</span>, header <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span><span class="va">weight_info</span> <span class="op">&lt;-</span> <span class="va">weight_info</span><span class="op">[</span><span class="va">weight_info</span><span class="op">$</span><span class="va">GENE</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">extra_table</span><span class="op">$</span><span class="va">gene</span>,<span class="op">]</span></span>
<span></span>
<span><span class="co"># write the .db file and the covariances</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="st">"weights_nolnc"</span>, showWarnings<span class="op">=</span><span class="cn">F</span><span class="op">)</span></span>
<span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.exists</a></span><span class="op">(</span><span class="st">"weights_nolnc/mashr_Liver_nolnc.db"</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">db</span> <span class="op">&lt;-</span> <span class="fu">dbConnect</span><span class="op">(</span><span class="va">sqlite</span>, <span class="st">"weights_nolnc/mashr_Liver_nolnc.db"</span><span class="op">)</span></span>
<span>  <span class="fu">dbWriteTable</span><span class="op">(</span><span class="va">db</span>, <span class="st">"extra"</span>, <span class="va">extra_table</span><span class="op">)</span></span>
<span>  <span class="fu">dbWriteTable</span><span class="op">(</span><span class="va">db</span>, <span class="st">"weights"</span>, <span class="va">weights_table</span><span class="op">)</span></span>
<span>  <span class="fu">dbDisconnect</span><span class="op">(</span><span class="va">db</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="va">weight_info_gz</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/connections.html" class="external-link">gzfile</a></span><span class="op">(</span><span class="st">"weights_nolnc/mashr_Liver_nolnc.txt.gz"</span>, <span class="st">"w"</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/write.table.html" class="external-link">write.table</a></span><span class="op">(</span><span class="va">weight_info</span>, <span class="va">weight_info_gz</span>, sep<span class="op">=</span><span class="st">" "</span>, quote<span class="op">=</span><span class="cn">F</span>, row.names<span class="op">=</span><span class="cn">F</span>, col.names<span class="op">=</span><span class="cn">T</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/connections.html" class="external-link">close</a></span><span class="op">(</span><span class="va">weight_info_gz</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>Process adipose weights:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://rsqlite.r-dbi.org" class="external-link">RSQLite</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># specify the weight to remove lncRNA from</span></span>
<span><span class="va">weight</span> <span class="op">&lt;-</span> <span class="st">"weights/eqtl/mashr/mashr_Adipose_Subcutaneous.db"</span></span>
<span></span>
<span><span class="co"># read the PredictDB weights</span></span>
<span><span class="va">sqlite</span> <span class="op">&lt;-</span> <span class="fu">dbDriver</span><span class="op">(</span><span class="st">"SQLite"</span><span class="op">)</span></span>
<span><span class="va">db</span> <span class="op">=</span> <span class="fu">dbConnect</span><span class="op">(</span><span class="va">sqlite</span>, <span class="va">weight</span><span class="op">)</span></span>
<span><span class="va">query</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">...</span><span class="op">)</span> <span class="fu">dbGetQuery</span><span class="op">(</span><span class="va">db</span>, <span class="va">...</span><span class="op">)</span></span>
<span><span class="va">weights_table</span> <span class="op">&lt;-</span> <span class="fu">query</span><span class="op">(</span><span class="st">"select * from weights"</span><span class="op">)</span></span>
<span><span class="va">extra_table</span> <span class="op">&lt;-</span> <span class="fu">query</span><span class="op">(</span><span class="st">"select * from extra"</span><span class="op">)</span></span>
<span><span class="fu">dbDisconnect</span><span class="op">(</span><span class="va">db</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># subset to protein coding genes only</span></span>
<span><span class="va">extra_table</span> <span class="op">&lt;-</span>  <span class="va">extra_table</span><span class="op">[</span><span class="va">extra_table</span><span class="op">$</span><span class="va">gene_type</span><span class="op">==</span><span class="st">"protein_coding"</span>,,drop<span class="op">=</span><span class="cn">F</span><span class="op">]</span></span>
<span><span class="va">weights_table</span> <span class="op">&lt;-</span> <span class="va">weights_table</span><span class="op">[</span><span class="va">weights_table</span><span class="op">$</span><span class="va">gene</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">extra_table</span><span class="op">$</span><span class="va">gene</span>,<span class="op">]</span></span>
<span></span>
<span><span class="co"># read and subset the covariances</span></span>
<span><span class="va">weight_info</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/connections.html" class="external-link">gzfile</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="fu">tools</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/tools/fileutils.html" class="external-link">file_path_sans_ext</a></span><span class="op">(</span><span class="va">weight</span><span class="op">)</span>, <span class="st">".txt.gz"</span><span class="op">)</span><span class="op">)</span>, header <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span><span class="va">weight_info</span> <span class="op">&lt;-</span> <span class="va">weight_info</span><span class="op">[</span><span class="va">weight_info</span><span class="op">$</span><span class="va">GENE</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">extra_table</span><span class="op">$</span><span class="va">gene</span>,<span class="op">]</span></span>
<span></span>
<span><span class="co"># write the .db file and the covariances</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="st">"weights_nolnc"</span>, showWarnings<span class="op">=</span><span class="cn">F</span><span class="op">)</span></span>
<span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.exists</a></span><span class="op">(</span><span class="st">"weights_nolnc/mashr_Adipose_Subcutaneous_nolnc.db"</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">db</span> <span class="op">&lt;-</span> <span class="fu">dbConnect</span><span class="op">(</span><span class="va">sqlite</span>, <span class="st">"weights_nolnc/mashr_Adipose_Subcutaneous_nolnc.db"</span><span class="op">)</span></span>
<span>  <span class="fu">dbWriteTable</span><span class="op">(</span><span class="va">db</span>, <span class="st">"extra"</span>, <span class="va">extra_table</span><span class="op">)</span></span>
<span>  <span class="fu">dbWriteTable</span><span class="op">(</span><span class="va">db</span>, <span class="st">"weights"</span>, <span class="va">weights_table</span><span class="op">)</span></span>
<span>  <span class="fu">dbDisconnect</span><span class="op">(</span><span class="va">db</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="va">weight_info_gz</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/connections.html" class="external-link">gzfile</a></span><span class="op">(</span><span class="st">"weights_nolnc/mashr_Adipose_Subcutaneous_nolnc.txt.gz"</span>, <span class="st">"w"</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/write.table.html" class="external-link">write.table</a></span><span class="op">(</span><span class="va">weight_info</span>, <span class="va">weight_info_gz</span>, sep<span class="op">=</span><span class="st">" "</span>, quote<span class="op">=</span><span class="cn">F</span>, row.names<span class="op">=</span><span class="cn">F</span>, col.names<span class="op">=</span><span class="cn">T</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/connections.html" class="external-link">close</a></span><span class="op">(</span><span class="va">weight_info_gz</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/rm.html" class="external-link">rm</a></span><span class="op">(</span><span class="va">extra_table</span>, <span class="va">weight_info</span>, <span class="va">weights_table</span>, <span class="va">weight_info_gz</span>, <span class="va">sqlite</span>, <span class="va">db</span><span class="op">)</span></span></code></pre></div>
<p>We specify these processed prediction models as a vector of paths:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">weight</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"weights_nolnc/mashr_Liver_nolnc.db"</span>, <span class="st">"weights_nolnc/mashr_Adipose_Subcutaneous_nolnc.db"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/stopifnot.html" class="external-link">stopifnot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.exists</a></span><span class="op">(</span><span class="va">weight</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="converting-between-fusion-and-predictdb-format">Converting between FUSION and PredictDB format<a class="anchor" aria-label="anchor" href="#converting-between-fusion-and-predictdb-format"></a>
</h2>
<p>Because multiple sets of prediction models can only be specified in PredictDB format, it may be necessary to convert FUSION format weights to PredictDB format. See Jing’s code for converting between FUSION and PredictDB weights.</p>
<p>PredictDB weights assume that variant genotypes are not standardized before imputation, but our implementation assumes standardized variant genotypes. For this reason, PredictDB weights are be scaled by genotype variance before imputing gene expression by default. If converting from FUSION to PredictDB format, the weights are already on the standardized scale, and scaling should be turned off using the option <code>scale_by_ld_variance=F</code> in the <code>impute_expr_z</code> function.</p>
</div>
<div class="section level2">
<h2 id="ld-reference-and-regions">LD reference and regions<a class="anchor" aria-label="anchor" href="#ld-reference-and-regions"></a>
</h2>
<p>LD reference information can be provided to cTWAS as either individual-level genotype data (in PLINK format), or as genetic correlation matrices (termed “R matrices”) for regions that are approximately LD-independent. These regions must also be specified, regardless of how the LD reference information is provided.</p>
<p>We will use the regions and LD matrices specified in the user guide for the <code>main</code> branch of cTWAS.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ld_R_dir</span> <span class="op">&lt;-</span> <span class="st">"/project2/mstephens/wcrouse/UKB_LDR_0.1/"</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="data-harmonization-and-preprocessing">Data harmonization and preprocessing<a class="anchor" aria-label="anchor" href="#data-harmonization-and-preprocessing"></a>
</h2>
<p>These inputs should be harmonized prior to analysis (i.e. the reference and alternative alleles for each variant should match across all three data sources).</p>
<div class="section level3">
<h3 id="harmonizing-gwas-z-scores-and-ld-reference-and-ld-mismatch-diagnoisis">Harmonizing GWAS z-scores and LD reference and LD mismatch diagnoisis<a class="anchor" aria-label="anchor" href="#harmonizing-gwas-z-scores-and-ld-reference-and-ld-mismatch-diagnoisis"></a>
</h3>
<p>This step will harmonize GWAS z-scores and LD reference, then perform LD mismatch diagnosis using SuSiE RSS, and flip z-scores with allele flipping issues.</p>
<p>We use SuSiE RSS to perform LD mismatch diagnosis and flip z-scores for allele flipped variants. This procedure can be time consuming. So we can split this task by chromosomes and combine the results.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">outputdir</span> <span class="op">&lt;-</span> <span class="st">"multigroup/processed_data/"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="va">outputdir</span>, showWarnings<span class="op">=</span><span class="cn">F</span>, recursive<span class="op">=</span><span class="cn">T</span><span class="op">)</span></span>
<span><span class="va">outname</span> <span class="op">&lt;-</span> <span class="st">"example"</span></span>
<span></span>
<span><span class="co"># number of cores to parallelize regions for LD mismatch diagnosis</span></span>
<span><span class="va">ncore</span> <span class="op">&lt;-</span> <span class="fl">6</span></span>
<span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fl">22</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.exists</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">outputdir</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">outname</span>, <span class="st">"_chr"</span>, <span class="va">i</span>, <span class="st">".harmonized.z_snp.RDS"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/preprocess_z_ld.html">preprocess_z_ld</a></span><span class="op">(</span>z_snp <span class="op">=</span> <span class="va">z_snp</span>,</span>
<span>                           ld_R_dir <span class="op">=</span> <span class="va">ld_R_dir</span>,</span>
<span>                           chrom <span class="op">=</span> <span class="va">i</span>,</span>
<span>                           gwas_n <span class="op">=</span> <span class="va">gwas_n</span>,</span>
<span>                           detect_ld_mismatch <span class="op">=</span> <span class="cn">T</span>, <span class="co"># If TRUE, detect SNPs with LD mismatches</span></span>
<span>                           flip_allele <span class="op">=</span> <span class="cn">T</span>, <span class="co"># If TRUE, flip z-scores with allele flipping issues</span></span>
<span>                           filter_ld_mismatch <span class="op">=</span> <span class="cn">F</span>, <span class="co"># do not filter problematic SNPs</span></span>
<span>                           outputdir <span class="op">=</span> <span class="va">outputdir</span>,</span>
<span>                           outname <span class="op">=</span> <span class="va">outname</span>,</span>
<span>                           ncore <span class="op">=</span> <span class="va">ncore</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">z_snp</span>, <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">outputdir</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">outname</span>, <span class="st">"_chr"</span>, <span class="va">i</span>, <span class="st">".harmonized.z_snp.RDS"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">ld_mismatch_res</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span>, <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">outputdir</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">outname</span>, <span class="st">"_chr"</span>, <span class="va">i</span>, <span class="st">".ld_mismatch_res.RDS"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co"># combine the harmonized z-scores</span></span>
<span><span class="va">z_snp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fl">22</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">z_snp</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">outputdir</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">outname</span>, <span class="st">"_chr"</span>, <span class="va">i</span>, <span class="st">".harmonized.z_snp.RDS"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="va">z_snp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="va">rbind</span>, <span class="va">z_snp</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">z_snp</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="cn">NULL</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/save.html" class="external-link">save</a></span><span class="op">(</span><span class="va">z_snp</span>, file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">outputdir</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">outname</span>, <span class="st">".harmonized.z_snp.Rd"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># extract and save the problematic variants</span></span>
<span><span class="va">problematic_snps</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fl">22</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">ld_mismatch_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">outputdir</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">outname</span>, <span class="st">"_chr"</span>, <span class="va">i</span>, <span class="st">".ld_mismatch_res.RDS"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="va">problematic_snps</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">problematic_snps</span>, <span class="va">ld_mismatch_res</span><span class="op">$</span><span class="va">problematic_snps</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/write.table.html" class="external-link">write.table</a></span><span class="op">(</span><span class="va">problematic_snps</span>, <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">outputdir</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">outname</span>, <span class="st">".problematic_snps.txt"</span><span class="op">)</span><span class="op">)</span>, </span>
<span>            col.names <span class="op">=</span> <span class="cn">F</span>, row.names <span class="op">=</span> <span class="cn">F</span>, quote <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span></code></pre></div>
<p>If you do not want to detect LD mismatches (and allele flipping) using use SuSiE RSS, you can turn off <code>detect_ld_mismatch</code>. It runs much faster without detecting LD mismatches, so we can simply run all the chromosomes together.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">outputdir</span> <span class="op">&lt;-</span> <span class="st">"multigroup/processed_data/"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="va">outputdir</span>, showWarnings<span class="op">=</span><span class="cn">F</span>, recursive<span class="op">=</span><span class="cn">T</span><span class="op">)</span></span>
<span><span class="va">outname</span> <span class="op">&lt;-</span> <span class="st">"example"</span></span>
<span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/preprocess_z_ld.html">preprocess_z_ld</a></span><span class="op">(</span>z_snp <span class="op">=</span> <span class="va">z_snp</span>,</span>
<span>                       ld_R_dir <span class="op">=</span> <span class="va">ld_R_dir</span>,</span>
<span>                       chrom <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">22</span>,</span>
<span>                       ld_regions <span class="op">=</span> <span class="st">"EUR"</span>,</span>
<span>                       ld_regions_version <span class="op">=</span> <span class="st">"b38"</span>,</span>
<span>                       filestem <span class="op">=</span> <span class="st">"ukb_b38_0.1"</span>,</span>
<span>                       gwas_n <span class="op">=</span> <span class="va">gwas_n</span>,</span>
<span>                       outputdir <span class="op">=</span> <span class="va">outputdir</span>,</span>
<span>                       outname <span class="op">=</span> <span class="va">outname</span>,</span>
<span>                       detect_ld_mismatch <span class="op">=</span> <span class="cn">F</span>,</span>
<span>                       save_result <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span>
<span></span>
<span><span class="va">z_snp</span> <span class="op">&lt;-</span> <span class="va">res</span><span class="op">$</span><span class="va">z_snp</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/save.html" class="external-link">save</a></span><span class="op">(</span><span class="va">z_snp</span>, file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">outputdir</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">outname</span>, <span class="st">".harmonized.z_snp.Rd"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="harmonizing-prediction-models-and-ld-reference">Harmonizing prediction models and LD reference<a class="anchor" aria-label="anchor" href="#harmonizing-prediction-models-and-ld-reference"></a>
</h3>
<p>Here we harmonize the PredictDB prediction models and LD reference before the cTWAS analysis. Harmonization only needs to be done once per combination of prediction models and LD reference. The output is a harmonized <code>.db</code> file in the output directory.</p>
<p><code>strand_ambig_action</code> gives options to harmonize strand ambiguous variants (A/T, G/C) between the prediction models and LD reference. “drop” (default) removes the ambiguous variant from the prediction models. “none” treats the variant as unambiguous, flipping the weights to match the LD reference and then taking no additional action. “recover” uses a procedure to recover strand ambiguous variants. This procedure compares correlations between variants in the LD reference and prediction models, and it can only be used with PredictDB format prediction models, which include this information.</p>
<p><em>We do not return the harmonized covariances between variants in the prediction models (<code>.txt.gz</code>).</em></p>
<p>Harmonize the liver prediction model:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">weight_dir</span> <span class="op">&lt;-</span> <span class="st">"multigroup/processed_data/weights_nolnc/"</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/preprocess_wgt_ld.html">preprocess_wgt_ld</a></span><span class="op">(</span>weight <span class="op">=</span> <span class="st">"weights_nolnc/mashr_Liver_nolnc.db"</span>,</span>
<span>                         ld_R_dir <span class="op">=</span> <span class="va">ld_R_dir</span>,</span>
<span>                         outputdir <span class="op">=</span> <span class="va">weight_dir</span>,</span>
<span>                         outname <span class="op">=</span> <span class="st">"mashr_Liver_nolnc_harmonized"</span>,</span>
<span>                         strand_ambig_action <span class="op">=</span> <span class="st">"drop"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/rm.html" class="external-link">rm</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span></code></pre></div>
<p>Harmonize the PredictDB adipose prediction model:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/preprocess_wgt_ld.html">preprocess_wgt_ld</a></span><span class="op">(</span>weight <span class="op">=</span> <span class="st">"weights_nolnc/mashr_Adipose_Subcutaneous_nolnc.db"</span>,</span>
<span>                         ld_R_dir <span class="op">=</span> <span class="va">ld_R_dir</span>,</span>
<span>                         outputdir <span class="op">=</span> <span class="va">weight_dir</span>,</span>
<span>                         outname <span class="op">=</span> <span class="st">"mashr_Adipose_Subcutaneous_nolnc_harmonized"</span>,</span>
<span>                         strand_ambig_action <span class="op">=</span> <span class="st">"drop"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/rm.html" class="external-link">rm</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="imputing-gene-z-scores">Imputing gene z-scores<a class="anchor" aria-label="anchor" href="#imputing-gene-z-scores"></a>
</h2>
<p>After we have done data harmonization and preprocessing, we specify the harmonized weights as a vector and impute the gene z-scores.</p>
<p>To impute gene z-scores genome-wide, we specify the GWAS summary statistics, both the PredictDB liver and adipose weights, and the LD reference matrices.</p>
<p>If specifying only a single <code>.db</code> file (or a single set of FUSION weights), the output of this function using the <code>multigroup</code> branch is identical to the output using the <code>main</code> branch. If specifying more than one <code>.db</code> file, the gene names will have the name of the corresponding <code>.db</code> file appended after the gene name, e.g. “gene1|mashr_Liver_nolnc”, but otherwise the output is the same.</p>
<p>In this version of cTWAS, we’ve included two options to parallelize and divide up the imputation. First, we’ve added the option <code>ncore</code> to specify the number of cores to use during imputation. Second, we’ve added the option to impute over a subset of chromosomes, allowing the imputation to be divided into multiple jobs.</p>
<p>Here, we impute gene z-scores for individual chromosomes using multiple cores, and then combining the results:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">outputdir</span> <span class="op">&lt;-</span> <span class="st">"multigroup/processed_data/"</span></span>
<span><span class="va">weight_dir</span> <span class="op">&lt;-</span> <span class="st">"multigroup/processed_data/weights_nolnc/"</span></span>
<span><span class="va">outname</span> <span class="op">&lt;-</span> <span class="st">"example"</span></span>
<span><span class="va">ld_R_dir</span> <span class="op">&lt;-</span> <span class="st">"/project2/mstephens/wcrouse/UKB_LDR_0.1/"</span></span>
<span></span>
<span><span class="co"># load processed z_snp file</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span>file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">outputdir</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">outname</span>, <span class="st">".harmonized.z_snp.Rd"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># processed prediction models</span></span>
<span><span class="va">weight</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">weight_dir</span>, <span class="st">"mashr_Liver_nolnc_harmonized.db"</span><span class="op">)</span>, </span>
<span>            <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">weight_dir</span>, <span class="st">"mashr_Adipose_Subcutaneous_nolnc_harmonized.db"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/stopifnot.html" class="external-link">stopifnot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.exists</a></span><span class="op">(</span><span class="va">weight</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># impute gene z-scores by chromosome</span></span>
<span><span class="va">ncore</span> <span class="op">&lt;-</span> <span class="fl">6</span></span>
<span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fl">22</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.exists</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">outputdir</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">outname</span>, <span class="st">"_chr"</span>, <span class="va">i</span>, <span class="st">".exprvar"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="st">"Imputing gene z-scores in chr"</span>, <span class="va">i</span>, <span class="st">"\n"</span><span class="op">)</span></span>
<span>    <span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/impute_expr_z.html">impute_expr_z</a></span><span class="op">(</span>z_snp <span class="op">=</span> <span class="va">z_snp</span>,</span>
<span>                         weight <span class="op">=</span> <span class="va">weight</span>,</span>
<span>                         ld_R_dir <span class="op">=</span> <span class="va">ld_R_dir</span>,</span>
<span>                         outputdir <span class="op">=</span> <span class="va">outputdir</span>,</span>
<span>                         outname <span class="op">=</span> <span class="va">outname</span>,</span>
<span>                         chrom<span class="op">=</span><span class="va">i</span>,</span>
<span>                         ncore<span class="op">=</span><span class="va">ncore</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co"># combine the imputed gene z-scores</span></span>
<span><span class="va">z_gene</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fl">22</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">outputdir</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">outname</span>, <span class="st">"_chr"</span>, <span class="va">i</span>, <span class="st">".exprqc.Rd"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="va">z_gene</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">z_gene_chr</span></span>
<span><span class="op">}</span></span>
<span><span class="va">z_gene</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="va">rbind</span>, <span class="va">z_gene</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">z_gene</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="cn">NULL</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/save.html" class="external-link">save</a></span><span class="op">(</span><span class="va">z_gene</span>, file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">outputdir</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">outname</span>, <span class="st">"_z_gene.Rd"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">z_gene</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/rm.html" class="external-link">rm</a></span><span class="op">(</span><span class="va">qclist</span>, <span class="va">wgtlist</span>, <span class="va">z_gene_chr</span><span class="op">)</span></span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Siming Zhao, Wesley Crouse, Sheng Qian, Kaixuan Luo, Matthew Stephens, Xin He.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
