<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Running cTWAS modules separately • ctwas</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Running cTWAS modules separately">
<meta property="og:description" content="ctwas">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ctwas</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.2.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorial
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/multigroup_ctwas_pipeline.html">Using multigroup cTWAS with summary statistics</a>
    </li>
    <li>
      <a href="../articles/multigroup_ctwas_modules.html">Running cTWAS modules separately</a>
    </li>
    <li>
      <a href="../articles/FAQ.html">FAQ</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/xinhe-lab/ctwas/tree/multigroup_test" class="external-link">Source</a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="multigroup_ctwas_modules_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Running cTWAS modules separately</h1>
                        <h4 data-toc-skip class="author">Kaixuan Luo, Sheng Qian</h4>
            
            <h4 data-toc-skip class="date">2024-04-19</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/xinhe-lab/ctwas/blob/multigroup_test/vignettes/multigroup_ctwas_modules.Rmd" class="external-link"><code>vignettes/multigroup_ctwas_modules.Rmd</code></a></small>
      <div class="hidden name"><code>multigroup_ctwas_modules.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="running-ctwas-modules-separately">Running cTWAS modules separately<a class="anchor" aria-label="anchor" href="#running-ctwas-modules-separately"></a>
</h2>
<p>The <code>ctwas_sumstats</code> function contains several cTWAS modules. We can run each of the cTWAS modules separately. This would allow more flexible controls of the cTWAS modules with different settings (memory, parallelization, etc.). We can save the results from each module for future access.</p>
<div class="section level3">
<h3 id="computing-z-scores-of-molecular-traits-">Computing z-scores of molecular traits.<a class="anchor" aria-label="anchor" href="#computing-z-scores-of-molecular-traits-"></a>
</h3>
<p>After we have done data preprocessing, we compute z-scores of molecular traits.</p>
<p>This function computes z-scores for the molecular traits using preprocessed SNP z-scores (<code>z_snp</code>) and the preprocessed weights.</p>
<p><em>Note: the software use the term “gene”, but they could be any molecular trait.</em></p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">z_gene</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/compute_gene_z.html">compute_gene_z</a></span><span class="op">(</span><span class="va">z_snp</span>, <span class="va">weights</span>, ncore<span class="op">=</span><span class="va">ncore</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/save.html" class="external-link">save</a></span><span class="op">(</span><span class="va">z_gene</span>, file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">outputdir</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">outname</span>, <span class="st">".gene_z.Rd"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="assemble-input-data-for-the-regions">Assemble input data for the regions<a class="anchor" aria-label="anchor" href="#assemble-input-data-for-the-regions"></a>
</h3>
<p>After we have done data preprocessing and imputed gene z-scores, we assemble the input data for all the regions.<br>
The <code>assemble_region_data</code> function will assign genes, SNPs and their z-score data for each region. It will identify the genes that cross region boundaries, and reassign to the region with the highest r2. It will down-sample the SNPs by the value of <code>thin</code> to reduce running time of parameter estimation and region screening. If <code>thin = 1</code>, it will use all the SNPs. It will trim (randomly by default) the SNPs if the number of SNPs in a region is more than <code>max_snp_region</code>. This functions returns a list with <code>region_data</code>, a list of regions with assembled input data, as well as a data frame of cross-boundary genes.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">thin</span> <span class="op">&lt;-</span> <span class="fl">0.1</span></span>
<span><span class="va">max_snp_region</span> <span class="op">&lt;-</span> <span class="fl">20000</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/assemble_region_data.html">assemble_region_data</a></span><span class="op">(</span><span class="va">region_info</span>, <span class="va">z_snp</span>, <span class="va">z_gene</span>, <span class="va">weights</span>,</span>
<span>                            maxSNP <span class="op">=</span> <span class="va">max_snp_region</span>,</span>
<span>                            trim_by <span class="op">=</span> <span class="st">"random"</span>,</span>
<span>                            thin <span class="op">=</span> <span class="va">thin</span>,</span>
<span>                            minvar <span class="op">=</span> <span class="fl">2</span>,</span>
<span>                            adjust_boundary_genes <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                            ncore <span class="op">=</span> <span class="va">ncore</span><span class="op">)</span></span>
<span><span class="va">region_data</span> <span class="op">&lt;-</span> <span class="va">res</span><span class="op">$</span><span class="va">region_data</span></span>
<span><span class="va">boundary_genes</span> <span class="op">&lt;-</span> <span class="va">res</span><span class="op">$</span><span class="va">boundary_genes</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS</a></span><span class="op">(</span><span class="va">res</span>, <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">outputdir</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">outname</span>, <span class="st">".region_data.thin"</span>, <span class="va">thin</span>, <span class="st">".RDS"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="estimating-parameters">Estimating parameters<a class="anchor" aria-label="anchor" href="#estimating-parameters"></a>
</h3>
<p>We use <code>est_param</code> function to estimate parameters. It will take the assembled <code>region_data</code> as input.</p>
<p>There are several options for how to handle the group prior variance parameters by specifying the <code>group_prior_var_stucture</code>:</p>
<ul>
<li>“independent” is the default and allows all groups to have their own separate variance parameters.</li>
<li>“shared_all” allows all groups to share the same variance parameter.</li>
<li>“shared_type” allows all groups in one molecular QTL type to share the same variance parameter.</li>
</ul>
<p>This step will run EM algorithm to estimate parameters and return estimate parameters (<code>group_prior</code>, <code>group_prior_var</code>, etc.). It will run two rounds of EM algorithm, first round uses fewer iterations (3 iterations by default) to get rough parameter estimates, and using the estimated priors in the first round, we select regions with single effect to run the second round. We use more iterations in the second round (30 iterations by default) to get accurate parameter estimates.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">param</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/est_param.html">est_param</a></span><span class="op">(</span><span class="va">region_data</span>, </span>
<span>                   group_prior_var_structure <span class="op">=</span> <span class="st">"shared_type"</span>, </span>
<span>                   niter1 <span class="op">=</span> <span class="fl">3</span>, niter2 <span class="op">=</span> <span class="fl">30</span>, </span>
<span>                   ncore <span class="op">=</span> <span class="va">ncore</span><span class="op">)</span></span>
<span><span class="va">group_prior</span> <span class="op">&lt;-</span> <span class="va">param</span><span class="op">$</span><span class="va">group_prior</span></span>
<span><span class="va">group_prior_var</span> <span class="op">&lt;-</span> <span class="va">param</span><span class="op">$</span><span class="va">group_prior_var</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS</a></span><span class="op">(</span><span class="va">param</span>, <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">outputdir</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">outname</span>, <span class="st">".param.RDS"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>We could use <code><a href="../reference/summarize_param.html">summarize_param()</a></code> to assess the convergence of the estimated parameters and to compute the proportion of variance explained (PVE) by variants and genes.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ctwas_parameters</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/summarize_param.html">summarize_param</a></span><span class="op">(</span><span class="va">param</span>, <span class="va">gwas_n</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS</a></span><span class="op">(</span><span class="va">ctwas_parameters</span>, <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">outputdir</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">outname</span>, <span class="st">".ctwas_parameters.RDS"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="screening-regions">Screening regions<a class="anchor" aria-label="anchor" href="#screening-regions"></a>
</h3>
<p>This step will run a screening process to fine-map all regions with down-sampled SNPs, and select regions with strong signals (<code>min_nonSNP_PIP = 0.5</code> by default). It returns a data frame with region IDs and nonSNP PIPs for each region. We could further narrow down the list of regions by using more stringent cutoffs on the nonSNP PIPs.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">region_nonSNP_PIP_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/screen_regions.html">screen_regions</a></span><span class="op">(</span><span class="va">region_data</span>,</span>
<span>                                       <span class="va">region_info</span>,</span>
<span>                                       <span class="va">weights</span>,</span>
<span>                                       L <span class="op">=</span> <span class="fl">5</span>,</span>
<span>                                       group_prior <span class="op">=</span> <span class="va">group_prior</span>,</span>
<span>                                       group_prior_var <span class="op">=</span> <span class="va">group_prior_var</span>,</span>
<span>                                       min_nonSNP_PIP <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>                                       ncore <span class="op">=</span> <span class="va">ncore</span>,</span>
<span>                                       verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">screened_region_ids</span> <span class="op">&lt;-</span> <span class="va">region_nonSNP_PIP_df</span><span class="op">$</span><span class="va">region_id</span></span>
<span><span class="va">screened_region_data</span> <span class="op">&lt;-</span> <span class="va">region_data</span><span class="op">[</span><span class="va">screened_region_ids</span><span class="op">]</span></span></code></pre></div>
<p>After selecting regions with strong signals, we expand the screened regions with full sets of SNPs for finemapping. This step is not needed if <code>thin = 1</code>.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="va">thin</span> <span class="op">&lt;</span> <span class="fl">1</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">screened_region_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/expand_region_data.html">expand_region_data</a></span><span class="op">(</span><span class="va">screened_region_data</span>,</span>
<span>                                             <span class="va">region_info</span>,</span>
<span>                                             <span class="va">z_snp</span>,</span>
<span>                                             <span class="va">z_gene</span>,</span>
<span>                                             trim_by <span class="op">=</span> <span class="st">"z"</span>,</span>
<span>                                             maxSNP <span class="op">=</span> <span class="va">max_snp_region</span>,</span>
<span>                                             ncore <span class="op">=</span> <span class="va">ncore</span><span class="op">)</span></span>
<span>  <span class="co"># replace data in screened regions with screened_region_data</span></span>
<span>  <span class="va">region_data</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">screened_region_data</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">screened_region_data</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS</a></span><span class="op">(</span><span class="va">region_data</span>, <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">outputdir</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">outname</span>, <span class="st">".region_data.RDS"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="fine-mapping-screened-regions">Fine-mapping screened regions<a class="anchor" aria-label="anchor" href="#fine-mapping-screened-regions"></a>
</h3>
<p>We now perform the fine-mapping step with the estimated parameters for each of screened regions.</p>
<p>The <code>finemap_regions</code> function performs finemapping for the screened regions in <code>screened_region_data</code>.</p>
<p>It first computes correlation matrices (SNP-SNP, SNP-gene, and gene-gene correlation matrices), and then runs fine-mapping with L = 5 with estimated parameters (<code>group_prior</code> and <code>group_prior_var</code>), and returns a data frame with annotated finemapping results for the regions.</p>
<p>If <code>save_cor = TRUE</code>, it will save the computed correlation matrices to <code>cor_dir</code> for future access.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">finemap_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/finemap_regions.html">finemap_regions</a></span><span class="op">(</span><span class="va">screened_region_data</span>,</span>
<span>                               <span class="va">region_info</span>,</span>
<span>                               <span class="va">weights</span>,</span>
<span>                               group_prior <span class="op">=</span> <span class="va">group_prior</span>,</span>
<span>                               group_prior_var <span class="op">=</span> <span class="va">group_prior_var</span>,</span>
<span>                               L <span class="op">=</span> <span class="fl">5</span>,</span>
<span>                               save_cor <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                               cor_dir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">outputdir</span>, <span class="st">"/cor_matrix/"</span><span class="op">)</span>,</span>
<span>                               ncore <span class="op">=</span> <span class="va">ncore</span>,</span>
<span>                               verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS</a></span><span class="op">(</span><span class="va">finemap_res</span>, <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">outputdir</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">outname</span>, <span class="st">".ctwas.res.RDS"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="fine-mapping-a-single-region">Fine-mapping a single region<a class="anchor" aria-label="anchor" href="#fine-mapping-a-single-region"></a>
</h3>
<p>If we already have assembled <code>region_data</code> for all the regions, we can expand the region of interest with full SNPs.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">region_id</span> <span class="op">&lt;-</span> <span class="st">"16:71020125-72901251"</span></span>
<span><span class="va">selected_region_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/expand_region_data.html">expand_region_data</a></span><span class="op">(</span><span class="va">region_data</span><span class="op">[</span><span class="va">region_id</span><span class="op">]</span>,</span>
<span>                                           <span class="va">region_info</span>,</span>
<span>                                           <span class="va">z_snp</span>,</span>
<span>                                           <span class="va">z_gene</span>,</span>
<span>                                           trim_by <span class="op">=</span> <span class="st">"z"</span><span class="op">)</span></span></code></pre></div>
<p>Otherwise, we can assemble <code>region_data</code> for this region with <code>thin=1</code>. But note, because we only have one region, this would not adjust boundary genes.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">regioninfo</span> <span class="op">&lt;-</span> <span class="va">region_info</span><span class="op">[</span><span class="va">region_info</span><span class="op">$</span><span class="va">region_id</span> <span class="op">==</span> <span class="va">region_id</span>, <span class="op">]</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/assemble_region_data.html">assemble_region_data</a></span><span class="op">(</span><span class="va">regioninfo</span>,</span>
<span>                            <span class="va">z_snp</span>,</span>
<span>                            <span class="va">z_gene</span>,</span>
<span>                            <span class="va">weights</span>,</span>
<span>                            trim_by <span class="op">=</span> <span class="st">"z"</span>,</span>
<span>                            thin <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                            adjust_boundary_genes <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">selected_region_data</span> <span class="op">&lt;-</span> <span class="va">res</span><span class="op">$</span><span class="va">region_data</span></span></code></pre></div>
<p>We can perform finemapping for this region of interest:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">finemap_region_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/finemap_regions.html">finemap_regions</a></span><span class="op">(</span><span class="va">selected_region_data</span>,</span>
<span>                                      <span class="va">region_info</span>,</span>
<span>                                      <span class="va">weights</span>,</span>
<span>                                      group_prior <span class="op">=</span> <span class="va">group_prior</span>,</span>
<span>                                      group_prior_var <span class="op">=</span> <span class="va">group_prior_var</span>,</span>
<span>                                      L <span class="op">=</span> <span class="fl">5</span>,</span>
<span>                                      save_cor <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                                      cor_dir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">outputdir</span>, <span class="st">"cor_matrix"</span><span class="op">)</span>,</span>
<span>                                      verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Siming Zhao, Wesley Crouse, Sheng Qian, Kaixuan Luo, Matthew Stephens, Xin He.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
