<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Using multigroup cTWAS with summary statistics • ctwas</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Using multigroup cTWAS with summary statistics">
<meta property="og:description" content="ctwas">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ctwas</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.2.0.9000</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorial
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/multigroup_ctwas_pipeline.html">Using multigroup cTWAS with summary statistics</a>
    </li>
    <li>
      <a href="../articles/multigroup_ctwas_modules.html">Running cTWAS modules separately</a>
    </li>
    <li>
      <a href="../articles/FAQ.html">FAQ</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/xinhe-lab/ctwas/tree/multigroup_test" class="external-link">Source</a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="multigroup_ctwas_pipeline_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Using multigroup cTWAS with summary statistics</h1>
                        <h4 data-toc-skip class="author">Kaixuan Luo, Sheng Qian</h4>
            
            <h4 data-toc-skip class="date">2024-04-18</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/xinhe-lab/ctwas/blob/multigroup_test/vignettes/multigroup_ctwas_pipeline.Rmd" class="external-link"><code>vignettes/multigroup_ctwas_pipeline.Rmd</code></a></small>
      <div class="hidden name"><code>multigroup_ctwas_pipeline.Rmd</code></div>

    </div>

    
    
<p>This document shows the updated pipeline for running cTWAS with summary statistics using the <code>multi-group</code> version of the R package, which extends our earlier cTWAS method (single group version) to integrate multiple groups of prediction models, e.g. multiple tissues, cell types or conditions, or multiple molecular traits.</p>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>Install the <code>multigroup</code> branch of <code>cTWAS</code> package</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/remotes/man/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"xinhe-lab/ctwas"</span>, ref <span class="op">=</span> <span class="st">"multigroup_test"</span><span class="op">)</span></span></code></pre></div>
<p>Load the package</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/xinhe-lab/ctwas" class="external-link">ctwas</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="preparing-input-data">Preparing input data<a class="anchor" aria-label="anchor" href="#preparing-input-data"></a>
</h2>
<p>In this version of <code>cTWAS</code> package, we require first running the data preprocessing and harmonization steps before running parameter estimation and finemapping steps of cTWAS analysis.</p>
<p>The inputs for the summary statistics version of cTWAS include GWAS summary statistics, prediction models in PredictDB format or FUSION format, and LD reference.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># set output directory and output name</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/getwd.html" class="external-link">setwd</a></span><span class="op">(</span><span class="st">"/project2/xinhe/shared_data/multigroup_ctwas/LDL_multigroup_tutorial/"</span><span class="op">)</span></span>
<span><span class="va">outputdir</span> <span class="op">&lt;-</span> <span class="st">"/project2/xinhe/shared_data/multigroup_ctwas/LDL_multigroup_tutorial/output"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="va">outputdir</span>, showWarnings<span class="op">=</span><span class="cn">F</span>, recursive<span class="op">=</span><span class="cn">T</span><span class="op">)</span></span>
<span></span>
<span><span class="va">outname</span> <span class="op">&lt;-</span> <span class="st">"LDL_Multiomics"</span></span>
<span><span class="va">gwas_name</span> <span class="op">&lt;-</span> <span class="st">"ukb-d-30780_irnt"</span></span></code></pre></div>
<div class="section level3">
<h3 id="gwas-z-scores">GWAS z-scores<a class="anchor" aria-label="anchor" href="#gwas-z-scores"></a>
</h3>
<p>We read the GWAS summary statistics as a data frame <code>z_snp</code>, with columns “id”, “A1”, “A2”, “z”, and each row is a variant. <code>A1</code> is the alternate allele, and <code>A2</code> is the reference allele.</p>
<p>For this analysis, we will use summary statistics from a GWAS of LDL cholesterol in the UK Biobank. We will download the VCF from the IEU Open GWAS Project.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># download the summary statistics, and unzip the file.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="st">"gwas_summary_stats"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/system.html" class="external-link">system</a></span><span class="op">(</span><span class="st">"wget https://gwas.mrcieu.ac.uk/files/ukb-d-30780_irnt/ukb-d-30780_irnt.vcf.gz -P gwas_summary_stats"</span><span class="op">)</span></span>
<span><span class="co">#R.utils::gunzip("gwas_summary_stats/ukb-d-30780_irnt.vcf.gz")</span></span></code></pre></div>
<p>Next, we will read the summary statistics. Then, we will compute the z-scores and format the input data. We will also collect the sample size, which will be useful later. We will save this output for convenience.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># read the data using the VariantAnnotation package</span></span>
<span><span class="va">z_snp</span> <span class="op">&lt;-</span> <span class="fu">VariantAnnotation</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/VariantAnnotation/man/readVcf-methods.html" class="external-link">readVcf</a></span><span class="op">(</span><span class="st">"gwas_summary_stats/ukb-d-30780_irnt.vcf.gz"</span><span class="op">)</span></span>
<span><span class="va">z_snp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="fu">gwasvcf</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/gwasvcf/man/vcf_to_tibble.html" class="external-link">vcf_to_tibble</a></span><span class="op">(</span><span class="va">z_snp</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># compute the z-scores</span></span>
<span><span class="va">z_snp</span><span class="op">$</span><span class="va">Z</span> <span class="op">&lt;-</span> <span class="va">z_snp</span><span class="op">$</span><span class="va">ES</span><span class="op">/</span><span class="va">z_snp</span><span class="op">$</span><span class="va">SE</span></span>
<span></span>
<span><span class="co"># collect sample size (most frequent sample size for all variants)</span></span>
<span><span class="va">gwas_n</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">z_snp</span><span class="op">$</span><span class="va">SS</span><span class="op">)</span>,decreasing<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="st">"gwas_n="</span>, <span class="va">gwas_n</span>, <span class="st">"\n"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># subset the columns and format the column names</span></span>
<span><span class="va">z_snp</span> <span class="op">&lt;-</span> <span class="va">z_snp</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"rsid"</span>, <span class="st">"ALT"</span>, <span class="st">"REF"</span>, <span class="st">"Z"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">z_snp</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"id"</span>, <span class="st">"A1"</span>, <span class="st">"A2"</span>, <span class="st">"z"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">z_snp_outfile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">outputdir</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">gwas_name</span>, <span class="st">".z_snp.Rd"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/save.html" class="external-link">save</a></span><span class="op">(</span><span class="va">z_snp</span>, <span class="va">gwas_n</span>, file<span class="op">=</span><span class="va">z_snp_outfile</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="prediction-models">Prediction models<a class="anchor" aria-label="anchor" href="#prediction-models"></a>
</h3>
<p>The <code>multigroup</code> version of cTWAS accepts multiple sets of prediction models in PredictDB or FUSION format.</p>
<p>Please check <a href="http://predictdb.org/" class="external-link">PredictDB</a> for the format of PredictDB weights. To specify weights in PredictDB format, provide the path to the <code>.db</code> file.</p>
<p>Please check <a href="http://gusevlab.org/projects/fusion/#compute-your-own-predictive-models" class="external-link">Fusion/TWAS</a> for the format of FUSION weights. To specify weights in FUSION format, provide the path to the folder that contains the <code>.wgt.RDat</code> files. The <code>.pos</code> file which points to the individual *.RDat weight files is assumed to be outside the folder. For more format details, check the FUSION website.</p>
</div>
<div class="section level3">
<h3 id="ld-reference-and-region-info">LD reference and region info<a class="anchor" aria-label="anchor" href="#ld-reference-and-region-info"></a>
</h3>
<p>LD reference information can be provided as genetic correlation matrices (termed “R matrices”) for regions that are approximately LD-independent.</p>
<p>cTWAS performs its analysis region-by-region. The preferred way to run cTWAS is to provide pre-computed LD matrices for each region.</p>
<p><em>It is critical that the genome build (e.g. hg38) of the LD reference matches the genome build used to train the prediction models.</em></p>
<p>The genome build of the GWAS summary statistics does not matter because variant positions are determined by the LD reference.</p>
<p>The choice of LD reference population is important for fine-mapping. Best practice for fine-mapping is to use in-sample LD reference (LD computed using the subjects in the GWAS sample). If in-sample LD reference is not an option, the LD reference should be as representative of the population in the GWAS sample as possible. Given that cTWAS is an extended fine-mapping algorithm, and that gene z-scores are computed using the observed GWAS z-scores, which reflect patterns of LD in the GWAS population, our recommendation is to match the LD reference to the GWAS population, not the population used to build the prediction models.</p>
<div class="section level4">
<h4 id="defining-regions">Defining regions<a class="anchor" aria-label="anchor" href="#defining-regions"></a>
</h4>
<p>cTWAS includes pre-defined regions based on European (EUR), Asian (ASN), or African (AFR) populations, using either genome build b38 or b37. These regions were previously generated using <a href="https://github.com/endrebak/ldetect" class="external-link">LDetect</a>.</p>
</div>
<div class="section level4">
<h4 id="ld-matrices">LD matrices<a class="anchor" aria-label="anchor" href="#ld-matrices"></a>
</h4>
<p>To use LD matrices for the LD reference, provide a directory containing all of the <code>.RDS</code> matrix files and matching <code>.Rvar</code> variant information files.</p>
<p>We have precomputed reference LD matrices and the variant information tables accompanying the LD matrices for both UKB and 1000G European Phase3 references. The complete LD matrices of European individuals from UK Biobank can be downloaded <a href="https://uchicago.box.com/s/jqocacd2fulskmhoqnasrknbt59x3xkn" class="external-link">here</a>. On the University of Chicago RCC cluster, the b38 reference is available at <code>/project2/mstephens/wcrouse/UKB_LDR_0.1/</code> and the b37 reference is available at <code>/project2/mstephens/wcrouse/UKB_LDR_0.1_b37/</code>.</p>
</div>
<div class="section level4">
<h4 id="region-info">Region info<a class="anchor" aria-label="anchor" href="#region-info"></a>
</h4>
<p>In this version of the package, we also require a data frame <code>region_info</code> containing the following columns: “chrom”, “start”, “stop”, “region_id”, “LD_matrix”, “SNP_info”.</p>
<p>“chrom”, “start”, “stop” are the genomic coordinates of the regions, “region_id” contains the IDs of the region, and we use “chr:start-stop” format in region IDs by default. “LD_matrix” stores the paths to the precomputed LD (R) matrices (<code>.RDS</code> files in our precomputed reference LD files), “SNP_info” stores the paths to the variant information corresponding to the LD matrices (<code>.Rvar</code> files in our precomputed reference LD files).</p>
<p>The <code>.RDS</code> file is <a href="https://www.rdocumentation.org/packages/base/versions/3.3.2/topics/readRDS?tap_a=5644-dce66f&amp;tap_s=10907-287229" class="external-link">R .RDS format</a>. It stores the LD correlation matrix for a region (a <span class="math inline">\(p \times p\)</span> matrix, <span class="math inline">\(p\)</span> is the number of variants in the region). We require that for each <code>.RDS</code> file, in the same directory, there is a corresponding file with the same stem but ending with the suffix <code>.Rvar</code>. This <code>.Rvar</code> files includes variant information for the region, and the order of its rows must match the order of rows and columns in the <code>.RDS</code> file.</p>
<p>Here we use b38 European region file, which is included in the package.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">genome_version</span> <span class="op">&lt;-</span> <span class="st">"b38"</span></span>
<span><span class="va">ld_R_dir</span> <span class="op">&lt;-</span> <span class="st">"/project2/mstephens/wcrouse/UKB_LDR_0.1/"</span></span>
<span></span>
<span><span class="va">region_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/ldetect"</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"EUR."</span>, <span class="va">genome_version</span>, <span class="st">".bed"</span><span class="op">)</span>, package <span class="op">=</span> <span class="st">"ctwas"</span><span class="op">)</span></span>
<span><span class="va">region_info</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="va">region_file</span>, header <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">region_info</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"chrom"</span>, <span class="st">"start"</span>, <span class="st">"stop"</span><span class="op">)</span></span>
<span><span class="va">region_info</span><span class="op">$</span><span class="va">chrom</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">"chr"</span>, <span class="st">""</span>, <span class="va">region_info</span><span class="op">$</span><span class="va">chrom</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">region_info</span><span class="op">$</span><span class="va">region_id</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">region_info</span><span class="op">$</span><span class="va">chr</span>, <span class="st">":"</span>, <span class="va">region_info</span><span class="op">$</span><span class="va">start</span>, <span class="st">"-"</span>, <span class="va">region_info</span><span class="op">$</span><span class="va">stop</span><span class="op">)</span></span>
<span></span>
<span><span class="va">filestem</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"ukb_"</span>, <span class="va">genome_version</span>, <span class="st">"_0.1"</span><span class="op">)</span></span>
<span><span class="va">ld_filestem</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sprintf.html" class="external-link">sprintf</a></span><span class="op">(</span><span class="st">"%s_chr%s.R_snp.%s_%s"</span>, <span class="va">filestem</span>, <span class="va">region_info</span><span class="op">$</span><span class="va">chrom</span>, <span class="va">region_info</span><span class="op">$</span><span class="va">start</span>, <span class="va">region_info</span><span class="op">$</span><span class="va">stop</span><span class="op">)</span></span>
<span><span class="va">region_info</span><span class="op">$</span><span class="va">LD_matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">ld_R_dir</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">ld_filestem</span>, <span class="st">".RDS"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">region_info</span><span class="op">$</span><span class="va">SNP_info</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">ld_R_dir</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">ld_filestem</span>, <span class="st">".Rvar"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># Check to make sure all the LD matrices and SNP info files are available.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/stopifnot.html" class="external-link">stopifnot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/all.html" class="external-link">all</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.exists</a></span><span class="op">(</span><span class="va">region_info</span><span class="op">$</span><span class="va">LD_matrix</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/stopifnot.html" class="external-link">stopifnot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/all.html" class="external-link">all</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.exists</a></span><span class="op">(</span><span class="va">region_info</span><span class="op">$</span><span class="va">SNP_info</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/write.table.html" class="external-link">write.table</a></span><span class="op">(</span><span class="va">region_info</span>, file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">outputdir</span>, <span class="st">"region_info.txt"</span><span class="op">)</span>, sep <span class="op">=</span> <span class="st">"\t"</span>, col.names <span class="op">=</span> <span class="cn">T</span>, row.names <span class="op">=</span> <span class="cn">F</span>, quote <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span></code></pre></div>
<p>The columns of the <code>.Rvar</code> file include information on chromosome, variant name, position in base pairs, and the alternative and reference alleles. The <code>variance</code> column is the variance of each variant prior to standardization; this is required for PredictDB weights but not FUSION weights. PredictDB weights should be scaled by the variance before imputing gene expression. This is because PredictDB weights assume that variant genotypes are not standardized before imputation, but our implementation assumes standardized variant genotypes. If variance information is missing, or if weights are in PredictDB format but are already on the standardized scale (e.g. if they were converted from FUSION to PredictDB format), this scaling can be turned off using the option <code>scale_by_ld_variance=FALSE</code> using the multigroup version of cTWAS. We’ve also include information on allele frequency in the variant info, but this is optional.</p>
<p>The naming convention for the LD matrices is <code>[filestem]_chr[#].R_snp.[start]_[end].RDS</code>. cTWAS expects that all <code>.RDS</code> and <code>.Rvar</code> files in the directory contain LD information, so no other files with these suffixes should be in the directory. Each variant should be uniquely assigned to a region, and the regions should be left closed and right open, i.e. [start, stop). The positions of the LD matrices must match exactly the positions specified by the region file. Do not include invariant or multiallelic variants in the LD reference.</p>
</div>
<div class="section level4">
<h4 id="convert-genotype-data-to-ld-matrices">Convert genotype data to LD matrices<a class="anchor" aria-label="anchor" href="#convert-genotype-data-to-ld-matrices"></a>
</h4>
<p>We could use the <code>convert_geno_to_LD_matrix</code> function to convert genotype files (and LD regions) to the LD matrices and corresponding variant information.</p>
<p>Below is an example of generating LD matrices in hg38 using UKB genotype data.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># specify LD reference</span></span>
<span><span class="va">ldref_dir</span> <span class="op">&lt;-</span> <span class="st">"/gpfs/data/xhe-lab/ukb_LDR/genotype_data_0.1"</span></span>
<span><span class="va">genotype_files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">ldref_dir</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"ukb_chr"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">22</span>, <span class="st">".pgen"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># the output Rvar files use the positions and allele information in varinfo_files</span></span>
<span><span class="va">varinfo_files</span> <span class="op">&lt;-</span> <span class="st">"/gpfs/data/xhe-lab/ukb_LDR/neale_lab/neale_variants_hg38.bim"</span></span>
<span><span class="co"># prepare a data frame region_info for LD regions with columns "chr", "start", and "stop"</span></span>
<span><span class="co"># the positions should match those in varinfo_files</span></span>
<span><span class="va">region_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/ldetect"</span>, <span class="st">"EUR.b38.bed"</span>, package <span class="op">=</span> <span class="st">"ctwas"</span><span class="op">)</span></span>
<span><span class="va">region_info</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="va">region_file</span>, header <span class="op">=</span> <span class="cn">TRUE</span>, stringsAsFactors <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="co"># specify output</span></span>
<span><span class="va">outputdir</span> <span class="op">&lt;-</span> <span class="st">"/gpfs/data/xhe-lab/ctwas/LDR/UKB_b38/"</span></span>
<span><span class="va">outname</span> <span class="op">&lt;-</span> <span class="st">"ukb_b38_0.1"</span></span>
<span></span>
<span><span class="co"># generate LD matrices (.RDS) and variant info (.Rvar)</span></span>
<span><span class="co"># update region info with paths of LD matrices and variant info files</span></span>
<span><span class="co"># we could run this for one chromosome or multiple chromosomes</span></span>
<span><span class="va">updated_region_info</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/convert_geno_to_LD_matrix.html">convert_geno_to_LD_matrix</a></span><span class="op">(</span><span class="va">region_info</span>, </span>
<span>                                                 <span class="va">genotype_files</span>, </span>
<span>                                                 <span class="va">varinfo_files</span>,</span>
<span>                                                 chrom <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">22</span>,</span>
<span>                                                 outputdir <span class="op">=</span> <span class="va">outputdir</span>, </span>
<span>                                                 outname <span class="op">=</span> <span class="va">outname</span><span class="op">)</span></span></code></pre></div>
<p>We provide example R scripts to generate the LD matrices for UKB or 1KG European genotype files.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># path to UKB b38 script</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/scripts"</span>, <span class="st">"convert_geno_to_LD_matrix_UKB_b38.R"</span>, package <span class="op">=</span> <span class="st">"ctwas"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># path to UKB b37 script</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/scripts"</span>, <span class="st">"convert_geno_to_LD_matrix_UKB_b37.R"</span>, package <span class="op">=</span> <span class="st">"ctwas"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># path to 1KG European b37 script</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/scripts"</span>, <span class="st">"convert_geno_to_LD_matrix_1KG_EUR_b37.R"</span>, package <span class="op">=</span> <span class="st">"ctwas"</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="data-harmonization">Data harmonization<a class="anchor" aria-label="anchor" href="#data-harmonization"></a>
</h3>
<p>These inputs should be harmonized prior to cTWAS analysis (i.e. the reference and alternative alleles for each variant should match across all three data sources).</p>
<div class="section level4">
<h4 id="harmonizing-gwas-z-scores-and-ld-reference">Harmonizing GWAS z-scores and LD reference<a class="anchor" aria-label="anchor" href="#harmonizing-gwas-z-scores-and-ld-reference"></a>
</h4>
<p>The <code><a href="../reference/preprocess_z_snp.html">preprocess_z_snp()</a></code> function harmonizes GWAS z-scores and LD reference based on the included allele information. If <code>drop_multiallelic = TRUE</code>, it will drop multiallelic variants. If <code>drop_strand_ambig = TRUE</code>, it will drop strand ambiguous variants.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># load the GWAS sumstats data</span></span>
<span><span class="va">z_snp_outfile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">outputdir</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">gwas_name</span>, <span class="st">".z_snp.Rd"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="va">z_snp_outfile</span><span class="op">)</span></span>
<span></span>
<span><span class="va">z_snp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/preprocess_z_snp.html">preprocess_z_snp</a></span><span class="op">(</span><span class="va">z_snp</span>, <span class="va">region_info</span>, <span class="va">gwas_n</span>, drop_multiallelic <span class="op">=</span> <span class="cn">TRUE</span>, drop_strand_ambig <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="harmonizing-prediction-models-and-ld-reference">Harmonizing prediction models and LD reference<a class="anchor" aria-label="anchor" href="#harmonizing-prediction-models-and-ld-reference"></a>
</h4>
<p>The <code>preprocess_weight()</code> function harmonizes the PredictDB/FUSION prediction models and LD reference. This only needs to be done once per combination of prediction models and LD reference.</p>
<p>In this example, we use three weights, one eQTL model, one sQTL model and one 3‘aQTL model from liver. eQTL and sQTL model are downloaded from <a href="https://predictdb.org/post/2021/07/21/gtex-v8-models-on-eqtl-and-sqtl/" class="external-link">PredictDB website</a>. 3‘aQTL model (Alternative polyadenylation) is in FUSION format and downloaded from <a href="http://bioinfo.szbl.ac.cn/TCGD/download.php" class="external-link">TCGD database</a>. We preprocess each weight file separately and combine them in the end. We specify PredictDB or FUSION format in <code>weight_format</code> and the specific method in FUSION by <code>method_Fusion</code>. We specify the <code>type</code> as the molecular QTL types, and <code>context</code> as the tissues, cell types, or conditions. For the expression models, we limit to protein coding genes, by setting <code>filter_protein_coding_genes = TRUE</code>. We set <code>scale_by_ld_variance=TRUE</code> for PredictDB weights because PredictDB assumes the unstandardized genotypes. For PredictDB models, we set <code>load_predictdb_LD=TRUE</code> to load pre-computed correlations (<code>.txt.gz</code> file) between weight variants. For FUSION models, we set <code>load_predictdb_LD=FALSE</code> to calculate correlations between weight variants with LD reference.</p>
<p>We use 6 cores to parallelize the computation over the weights.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">weight_files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"/project2/xinhe/shared_data/multigroup_ctwas/weights/expression_models/mashr_Liver.db"</span>,</span>
<span>                  <span class="st">"/project2/xinhe/shared_data/multigroup_ctwas/weights/splicing_models/mashr_Liver.db"</span>,</span>
<span>                  <span class="st">"/project2/xinhe/shared_data/multigroup_ctwas/weights/apa_models/Liver/Liver/"</span><span class="op">)</span></span>
<span><span class="va">ncore</span> <span class="op">&lt;-</span> <span class="fl">6</span></span>
<span></span>
<span><span class="va">weights_liver_expression</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/preprocess_weights.html">preprocess_weights</a></span><span class="op">(</span>weight_file <span class="op">=</span> <span class="va">weight_files</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>,</span>
<span>                                               <span class="va">region_info</span>,</span>
<span>                                               <span class="va">z_snp</span><span class="op">$</span><span class="va">id</span>,</span>
<span>                                               type <span class="op">=</span> <span class="st">"eQTL"</span>,</span>
<span>                                               context <span class="op">=</span> <span class="st">"liver"</span>,</span>
<span>                                               weight_format <span class="op">=</span> <span class="st">"PredictDB"</span>,</span>
<span>                                               ncore <span class="op">=</span> <span class="va">ncore</span>,</span>
<span>                                               drop_strand_ambig <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                                               scale_by_ld_variance <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                                               load_predictdb_LD <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                                               filter_protein_coding_genes <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">weights_liver_splicing</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/preprocess_weights.html">preprocess_weights</a></span><span class="op">(</span>weight_file <span class="op">=</span> <span class="va">weight_files</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span>,</span>
<span>                                             <span class="va">region_info</span>,</span>
<span>                                             <span class="va">z_snp</span><span class="op">$</span><span class="va">id</span>,</span>
<span>                                             type <span class="op">=</span> <span class="st">"sQTL"</span>,</span>
<span>                                             context <span class="op">=</span> <span class="st">"liver"</span>,</span>
<span>                                             weight_format <span class="op">=</span> <span class="st">"PredictDB"</span>,</span>
<span>                                             ncore <span class="op">=</span> <span class="va">ncore</span>,</span>
<span>                                             drop_strand_ambig <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                                             scale_by_ld_variance <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                                             load_predictdb_LD <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                                             filter_protein_coding_genes <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">weights_liver_apa</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/preprocess_weights.html">preprocess_weights</a></span><span class="op">(</span>weight_file <span class="op">=</span> <span class="va">weight_files</span><span class="op">[</span><span class="fl">3</span><span class="op">]</span>,</span>
<span>                                        <span class="va">region_info</span>,</span>
<span>                                        <span class="va">z_snp</span><span class="op">$</span><span class="va">id</span>,</span>
<span>                                        type <span class="op">=</span> <span class="st">"aQTL"</span>,</span>
<span>                                        context <span class="op">=</span> <span class="st">"Liver"</span>,</span>
<span>                                        weight_format <span class="op">=</span> <span class="st">"Fusion"</span>,</span>
<span>                                        ncore <span class="op">=</span> <span class="va">ncore</span>,</span>
<span>                                        scale_by_ld_variance <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                                        drop_strand_ambig <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                                        load_predictdb_LD <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                                        filter_protein_coding_genes <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                                        method_Fusion <span class="op">=</span> <span class="st">"enet"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">weights</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">weights_liver_expression</span>, <span class="va">weights_liver_splicing</span>, <span class="va">weights_liver_apa</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
</div>
<div class="section level2">
<h2 id="running-ctwas-analysis">Running cTWAS analysis<a class="anchor" aria-label="anchor" href="#running-ctwas-analysis"></a>
</h2>
<p>We can run the entire process of cTWAS analysis from computing z-scores of molecular traits, estimating parameters to finemapping using the main <code>ctwas_sumstats</code> function.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ctwas_sumstats.html">ctwas_sumstats</a></span><span class="op">(</span><span class="va">z_snp</span>, <span class="va">weights</span>, <span class="va">region_info</span>, </span>
<span>                      thin <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>                      niter1 <span class="op">=</span> <span class="fl">3</span>, niter2 <span class="op">=</span> <span class="fl">30</span>, </span>
<span>                      L <span class="op">=</span> <span class="fl">5</span>, </span>
<span>                      group_prior_var_structure <span class="op">=</span> <span class="st">"independent"</span>, </span>
<span>                      max_snp_region <span class="op">=</span> <span class="fl">20000</span>, </span>
<span>                      min_nonSNP_PIP <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>                      ncore <span class="op">=</span> <span class="va">ncore</span>, </span>
<span>                      save_cor <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                      cor_dir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">outputdir</span>, <span class="st">"cor_matrix"</span><span class="op">)</span>,</span>
<span>                      logfile <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">outputdir</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">outname</span>, <span class="st">".ctwas_sumstats.log"</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                      verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">param</span> <span class="op">&lt;-</span> <span class="va">res</span><span class="op">$</span><span class="va">param</span></span>
<span><span class="va">finemap_res</span> <span class="op">&lt;-</span> <span class="va">res</span><span class="op">$</span><span class="va">finemap_res</span></span>
<span><span class="va">boundary_genes</span> <span class="op">&lt;-</span> <span class="va">res</span><span class="op">$</span><span class="va">boundary_genes</span></span>
<span><span class="va">region_data</span> <span class="op">&lt;-</span> <span class="va">res</span><span class="op">$</span><span class="va">region_data</span></span></code></pre></div>
<p>We can also run each of the cTWAS modules separately. This would allow more flexible controls of the cTWAS modules with different settings (memory, parallelization, etc.) See [this tutorial][] for more details.</p>
</div>
<div class="section level2">
<h2 id="postprocessing">Postprocessing<a class="anchor" aria-label="anchor" href="#postprocessing"></a>
</h2>
<div class="section level3">
<h3 id="merging-regions">Merging regions<a class="anchor" aria-label="anchor" href="#merging-regions"></a>
</h3>
<p>For cross-boundary genes, we will merge the regions for high PIP (&gt;0.5) genes.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">high_PIP_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">finemap_res</span><span class="op">[</span><span class="va">finemap_res</span><span class="op">$</span><span class="va">type</span> <span class="op">!=</span> <span class="st">"SNP"</span> <span class="op">&amp;</span> <span class="va">finemap_res</span><span class="op">$</span><span class="va">susie_pip</span> <span class="op">&gt;=</span> <span class="fl">0.5</span>, <span class="st">"id"</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">high_PIP_boundary_genes</span> <span class="op">&lt;-</span> <span class="va">boundary_genes</span><span class="op">[</span><span class="va">boundary_genes</span><span class="op">$</span><span class="va">id</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">high_PIP_genes</span>, , drop<span class="op">=</span><span class="cn">FALSE</span><span class="op">]</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">high_PIP_boundary_genes</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/merge_region_data.html">merge_region_data</a></span><span class="op">(</span><span class="va">region_data</span>,</span>
<span>                           <span class="va">high_PIP_boundary_genes</span>,</span>
<span>                           <span class="va">region_info</span>,</span>
<span>                           <span class="va">z_snp</span>,</span>
<span>                           <span class="va">z_gene</span>,</span>
<span>                           expand <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                           maxSNP <span class="op">=</span> <span class="va">max_snp_region</span><span class="op">)</span></span>
<span>  <span class="va">merged_region_data</span> <span class="op">&lt;-</span> <span class="va">res</span><span class="op">$</span><span class="va">merged_region_data</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>For those merged regions, we will expand those regions with full SNPs. We rerun finemapping for the merged regions, and then update the finemapping results for those merged regions.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Expand those merged regions with full SNPs</span></span>
<span><span class="va">merged_region_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/expand_region_data.html">expand_region_data</a></span><span class="op">(</span><span class="va">merged_region_data</span>,</span>
<span>                                         <span class="va">region_info</span>,</span>
<span>                                         <span class="va">z_snp</span>,</span>
<span>                                         <span class="va">z_gene</span>,</span>
<span>                                         trim_by <span class="op">=</span> <span class="st">"z"</span>,</span>
<span>                                         maxSNP <span class="op">=</span> <span class="va">max_snp_region</span>,</span>
<span>                                         ncore <span class="op">=</span> <span class="va">ncore</span><span class="op">)</span></span>
<span></span>
<span><span class="va">finemap_merged_regions_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/finemap_regions.html">finemap_regions</a></span><span class="op">(</span><span class="va">merged_region_data</span>,</span>
<span>                                              <span class="va">region_info</span>,</span>
<span>                                              <span class="va">weights</span>,</span>
<span>                                              group_prior <span class="op">=</span> <span class="va">group_prior</span>,</span>
<span>                                              group_prior_var <span class="op">=</span> <span class="va">group_prior_var</span>,</span>
<span>                                              L <span class="op">=</span> <span class="fl">5</span>,</span>
<span>                                              save_cor <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                                              cor_dir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">outputdir</span>, <span class="st">"/merged_cor_matrix/"</span><span class="op">)</span>,</span>
<span>                                              ncore <span class="op">=</span> <span class="va">ncore</span>,</span>
<span>                                              verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="rerun-finemapping-with-l-1-for-regions-with-problematic-snps">Rerun finemapping with L = 1 for regions with problematic SNPs<a class="anchor" aria-label="anchor" href="#rerun-finemapping-with-l-1-for-regions-with-problematic-snps"></a>
</h3>
<p>Select regions with high PIP (&gt;0.5) SNPs and molecular traits.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">highPIP_finemap_res</span> <span class="op">&lt;-</span> <span class="va">finemap_res</span><span class="op">[</span><span class="va">finemap_res</span><span class="op">$</span><span class="va">susie_pip</span> <span class="op">&gt;</span> <span class="fl">0.5</span>, <span class="op">]</span></span>
<span><span class="va">highPIP_region_ids</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">highPIP_finemap_res</span><span class="op">$</span><span class="va">region_id</span><span class="op">)</span></span></code></pre></div>
<p>We use SuSiE RSS to perform LD mismatch diagnosis for these high PIP regions.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">high_PIP_region_info</span> <span class="op">&lt;-</span> <span class="va">region_info</span><span class="op">[</span><span class="va">region_info</span><span class="op">$</span><span class="va">region_id</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">highPIP_region_ids</span>, , drop<span class="op">=</span><span class="cn">FALSE</span><span class="op">]</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/detect_ld_mismatch_susie.html">detect_ld_mismatch_susie</a></span><span class="op">(</span><span class="va">z_snp</span>, <span class="va">high_PIP_region_info</span>, <span class="va">gwas_n</span>, ncore <span class="op">=</span> <span class="va">ncore</span>, p_diff_thresh <span class="op">=</span> <span class="fl">5e-8</span><span class="op">)</span></span>
<span><span class="va">problematic_snps</span> <span class="op">&lt;-</span> <span class="va">res</span><span class="op">$</span><span class="va">problematic_snps</span></span></code></pre></div>
<p>We will rerun the fine-mapping with L = 1 for the regions containing high PIP (&gt;0.5) SNPs detected with LD mismatches, or high PIP molecular traits with problematic SNPs in their weights.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># select regions containing high PIP (&gt;0.5) SNPs detected with LD mismatches, </span></span>
<span><span class="co"># or high PIP molecular traits with problematic SNPs in their weights</span></span>
<span><span class="va">problematic_region_ids</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/select_problematic_regions.html">select_problematic_regions</a></span><span class="op">(</span><span class="va">problematic_snps</span>, <span class="va">highPIP_finemap_res</span>, <span class="va">weights</span><span class="op">)</span></span>
<span><span class="va">rerun_region_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/expand_region_data.html">expand_region_data</a></span><span class="op">(</span><span class="va">region_data</span><span class="op">[</span><span class="va">problematic_region_ids</span><span class="op">]</span>,</span>
<span>                                        <span class="va">region_info</span>,</span>
<span>                                        <span class="va">z_snp</span>,</span>
<span>                                        <span class="va">z_gene</span>,</span>
<span>                                        trim_by <span class="op">=</span> <span class="st">"z"</span>,</span>
<span>                                        maxSNP <span class="op">=</span> <span class="va">max_snp_region</span>,</span>
<span>                                        ncore <span class="op">=</span> <span class="va">ncore</span><span class="op">)</span></span>
<span></span>
<span><span class="va">finemap_merged_regions_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/finemap_regions.html">finemap_regions</a></span><span class="op">(</span><span class="va">rerun_region_data</span>,</span>
<span>                                              <span class="va">region_info</span>,</span>
<span>                                              <span class="va">weights</span>,</span>
<span>                                              group_prior <span class="op">=</span> <span class="va">group_prior</span>,</span>
<span>                                              group_prior_var <span class="op">=</span> <span class="va">group_prior_var</span>,</span>
<span>                                              L <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                                              cor_dir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">outputdir</span>, <span class="st">"/cor_matrix/"</span><span class="op">)</span>,</span>
<span>                                              ncore <span class="op">=</span> <span class="va">ncore</span>,</span>
<span>                                              verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="ctwas-output">cTWAS output<a class="anchor" aria-label="anchor" href="#ctwas-output"></a>
</h2>
<div class="section level3">
<h3 id="interpreting-ctwas-output">Interpreting cTWAS output<a class="anchor" aria-label="anchor" href="#interpreting-ctwas-output"></a>
</h3>
<p>We will add the gene names to the results (the PredictDB weights use Ensembl IDs as the primary identifier), as well as the z-scores for each SNP and gene. We then show all molecular traits with PIP &gt; 0.8, which is the threshold we used in the paper.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># load information for all genes</span></span>
<span><span class="va">gene_info</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_gene_info.html">get_gene_info</a></span><span class="op">(</span><span class="va">weights</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># add gene names to cTWAS results</span></span>
<span><span class="va">gene_idx</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span><span class="va">finemap_res</span><span class="op">$</span><span class="va">id</span><span class="op">[</span><span class="va">finemap_res</span><span class="op">$</span><span class="va">type</span><span class="op">!=</span><span class="st">"SNP"</span><span class="op">]</span>, <span class="va">gene_info</span><span class="op">$</span><span class="va">id</span><span class="op">)</span></span>
<span><span class="va">finemap_res</span><span class="op">$</span><span class="va">genename</span><span class="op">[</span><span class="va">finemap_res</span><span class="op">$</span><span class="va">type</span><span class="op">!=</span><span class="st">"SNP"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">gene_info</span><span class="op">$</span><span class="va">genename</span><span class="op">[</span><span class="va">gene_idx</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># add z-scores to cTWAS results</span></span>
<span><span class="va">s_idx</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span><span class="va">finemap_res</span><span class="op">$</span><span class="va">id</span><span class="op">[</span><span class="va">finemap_res</span><span class="op">$</span><span class="va">type</span><span class="op">==</span><span class="st">"SNP"</span><span class="op">]</span>, <span class="va">z_snp</span><span class="op">$</span><span class="va">id</span><span class="op">)</span></span>
<span><span class="va">finemap_res</span><span class="op">$</span><span class="va">z</span><span class="op">[</span><span class="va">finemap_res</span><span class="op">$</span><span class="va">type</span><span class="op">==</span><span class="st">"SNP"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">z_snp</span><span class="op">$</span><span class="va">z</span><span class="op">[</span><span class="va">s_idx</span><span class="op">]</span></span>
<span></span>
<span><span class="va">g_idx</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span><span class="va">finemap_res</span><span class="op">$</span><span class="va">id</span><span class="op">[</span><span class="va">finemap_res</span><span class="op">$</span><span class="va">type</span><span class="op">!=</span><span class="st">"SNP"</span><span class="op">]</span>, <span class="va">z_gene</span><span class="op">$</span><span class="va">id</span><span class="op">)</span></span>
<span><span class="va">finemap_res</span><span class="op">$</span><span class="va">z</span><span class="op">[</span><span class="va">finemap_res</span><span class="op">$</span><span class="va">type</span><span class="op">!=</span><span class="st">"SNP"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">z_gene</span><span class="op">$</span><span class="va">z</span><span class="op">[</span><span class="va">g_idx</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># select PIP &gt; 0.8</span></span>
<span><span class="va">finemap_res</span> <span class="op">&lt;-</span> <span class="va">finemap_res</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="op">-</span><span class="va">finemap_res</span><span class="op">$</span><span class="va">susie_pip</span><span class="op">)</span>,<span class="op">]</span></span>
<span><span class="va">ctwas_res</span> <span class="op">&lt;-</span> <span class="va">finemap_res</span><span class="op">[</span><span class="va">finemap_res</span><span class="op">$</span><span class="va">type</span><span class="op">!=</span><span class="st">"SNP"</span> <span class="op">&amp;</span> <span class="va">finemap_res</span><span class="op">$</span><span class="va">susie_pip</span> <span class="op">&gt;</span> <span class="fl">0.8</span>,<span class="op">]</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="combining-pips-across-tissues-or-molecular-traits">Combining PIPs across tissues or molecular traits<a class="anchor" aria-label="anchor" href="#combining-pips-across-tissues-or-molecular-traits"></a>
</h3>
</div>
</div>
<div class="section level2">
<h2 id="visualzing-ctwas-results">Visualzing cTWAS results<a class="anchor" aria-label="anchor" href="#visualzing-ctwas-results"></a>
</h2>
<p>We could make locus plot for regions of interest.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Siming Zhao, Wesley Crouse, Sheng Qian, Kaixuan Luo, Matthew Stephens, Xin He.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
