<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Using cTWAS multi-group version with summary statistics • ctwas</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Using cTWAS multi-group version with summary statistics">
<meta property="og:description" content="ctwas">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ctwas</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.36</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorial
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/data_preparation.html">Data preparation</a>
    </li>
    <li>
      <a href="../articles/postprocessing_regions.html">Postprocessing problematic regions</a>
    </li>
    <li>
      <a href="../articles/ctwas_summary_statistics.html">Using cTWAS single-group version with summary statistics</a>
    </li>
    <li>
      <a href="../articles/multigroup_ctwas_summary_statistics.html">Using cTWAS multi-group version with summary statistics</a>
    </li>
    <li>
      <a href="../articles/harmonization.html">Harmonization</a>
    </li>
    <li>
      <a href="../articles/regionmerging.html">Region Merging</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/xinhe-lab/ctwas" class="external-link">Source</a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="multigroup_ctwas_summary_statistics_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Using cTWAS multi-group version with summary statistics</h1>
                        <h4 data-toc-skip class="author">Kaixuan Luo, Sheng Qian, Wesley Crouse, Xin He</h4>
            
            <h4 data-toc-skip class="date">2024-02-13</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/xinhe-lab/ctwas/blob/main/vignettes/multigroup_ctwas_summary_statistics.Rmd" class="external-link"><code>vignettes/multigroup_ctwas_summary_statistics.Rmd</code></a></small>
      <div class="hidden name"><code>multigroup_ctwas_summary_statistics.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p>This document is a tutorial for running cTWAS with summary statistics using the <code>multi-group</code> version of the R package, which extends our earlier cTWAS method (single group version) to integrate multiple groups of prediction models. This can be used to jointly analyze gene expression from multiple tissues, and/or to jointly analyze prediction models from different types of -omics data.</p>
<p>In this tutorial, we will show one example to jointly model gene expression from multiple tissues, but the same framework could also be used to jointly modeling multiple molecular QTL data types (with slight change of the parameter settings).</p>
<p>This document assumes that you have already read the main tutorial for cTWAS (single group version), available <a href="https://xinhe-lab.github.io/ctwas/articles/ctwas_summary_statistics.html" class="external-link">here</a>.</p>
</div>
<div class="section level2">
<h2 id="getting-started">Getting started<a class="anchor" aria-label="anchor" href="#getting-started"></a>
</h2>
<p>Install the <code>multigroup_kevin</code> branch of <code>cTWAS</code> package</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/remotes/man/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"xinhe-lab/ctwas"</span>, ref <span class="op">=</span> <span class="st">"multigroup_kevin"</span><span class="op">)</span></span></code></pre></div>
<p>Load the package and set the working directory where you want to perform the analysis.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/xinhe-lab/ctwas" class="external-link">ctwas</a></span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#set the working directory interactively</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/getwd.html" class="external-link">setwd</a></span><span class="op">(</span><span class="st">"/project2/xinhe/shared_data/ctwas_tutorial/multigroup"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#set the working directory when compiling this document with Knitr</span></span>
<span><span class="fu">knitr</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/knitr/man/opts_knit.html" class="external-link">opts_knit</a></span><span class="op">$</span><span class="fu">set</span><span class="op">(</span>root.dir <span class="op">=</span> <span class="st">"/project2/xinhe/shared_data/ctwas_tutorial/multigroup"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="preparing-input-data">Preparing input data<a class="anchor" aria-label="anchor" href="#preparing-input-data"></a>
</h2>
<p>In this version of <code>cTWAS</code> package, we require running the data preparation steps before running parameter estimation and finemapping steps of cTWAS analysis. Please see the [data preparation tutorial][here](<a href="https://xinhe-lab.github.io/ctwas/articles/data_preparation.html" class="external-link uri">https://xinhe-lab.github.io/ctwas/articles/data_preparation.html</a>) on preparing and harmonizing GWAS summary statistics, prediction models, and LD reference.</p>
</div>
<div class="section level2">
<h2 id="running-ctwas-fine-mapping-with-fixed-parameters-at-a-single-locus">Running cTWAS fine-mapping with fixed parameters at a single locus<a class="anchor" aria-label="anchor" href="#running-ctwas-fine-mapping-with-fixed-parameters-at-a-single-locus"></a>
</h2>
<p>After preparing input data, we are ready to run the cTWAS analysis.</p>
<p>The full analysis involves first estimating parameters from the data, and then fine-mapping the genes and variants using these estimated parameters. This can be computationally intensive.</p>
<p>In this example, we will only run the fine-mapping step at an example locus, using the parameters we estimated in the paper.</p>
<p>Let’s run the cTWAS fine-mapping step at the HPR locus using the estimated parameters.</p>
<p>Load the processed example data (also used in the <code>main</code> branch tutorial):</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">outputdir</span> <span class="op">&lt;-</span> <span class="st">"/project2/xinhe/shared_data/ctwas_tutorial/multigroup/results/single_locus/"</span></span>
<span><span class="va">outname</span> <span class="op">&lt;-</span> <span class="st">"example_locus"</span></span>
<span></span>
<span><span class="co"># load imputed gene z-scores, the harmonized GWAS z-scores</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">outputdir</span>, <span class="st">"/"</span>, <span class="va">outname</span>, <span class="st">"_z_gene.Rd"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">outputdir</span>, <span class="st">"/"</span>, <span class="va">outname</span>, <span class="st">"_z_snp.Rd"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">ld_exprvarfs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">outputdir</span>, <span class="st">"/"</span>, <span class="va">outname</span>, <span class="st">"_chr"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">22</span>, <span class="st">".exprvar"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/stopifnot.html" class="external-link">stopifnot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.exists</a></span><span class="op">(</span><span class="va">ld_exprvarfs</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">ld_R_dir</span> <span class="op">&lt;-</span> <span class="st">"/project2/mstephens/wcrouse/UKB_LDR_0.1/"</span></span></code></pre></div>
<p>Here we use estimated parameters:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># the estimated prior inclusion probabilities for genes and variants from the paper</span></span>
<span><span class="va">group_prior</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.0107220302</span>, <span class="fl">0.0001715896</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># the estimated effect sizes for genes and variants from the paper</span></span>
<span><span class="va">group_prior_var</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">41.327666</span>, <span class="fl">9.977841</span><span class="op">)</span></span></code></pre></div>
<p>If we want to rerun a region using precomputed regionlist, we can simply specify the <code>region_tags</code> to rerun, and set <code>reuse_regionlist = TRUE</code>.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ctwas_finemap_rss</span><span class="op">(</span>z_gene <span class="op">=</span> <span class="va">z_gene_subset</span>,</span>
<span>                  z_snp <span class="op">=</span> <span class="va">z_snp_subset</span>,</span>
<span>                  ld_exprvarfs <span class="op">=</span> <span class="va">ld_exprvarfs</span>,</span>
<span>                  ld_R_dir <span class="op">=</span> <span class="va">ld_R_dir</span>,</span>
<span>                  reuse_regionlist <span class="op">=</span> <span class="cn">T</span>,</span>
<span>                  region_tags <span class="op">=</span> <span class="st">"16_1"</span>, <span class="co"># rerun on region 1 in chr16</span></span>
<span>                  outputdir <span class="op">=</span> <span class="va">outputdir</span>,</span>
<span>                  outname <span class="op">=</span> <span class="va">outname</span>,</span>
<span>                  group_prior <span class="op">=</span> <span class="va">group_prior</span>,</span>
<span>                  group_prior_var <span class="op">=</span> <span class="va">group_prior_var</span><span class="op">)</span></span></code></pre></div>
<p>or we could run finemapping on regions listed in a custom region file:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">regions_file</span> <span class="op">&lt;-</span> <span class="st">"/project2/xinhe/shared_data/ctwas_tutorial/regions/regions_subset.bed"</span></span>
<span><span class="va">regions_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="va">regions_file</span>, header <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">regions_df</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">ctwas_finemap_regions_rss</span><span class="op">(</span>z_gene <span class="op">=</span> <span class="va">z_gene_subset</span>,</span>
<span>                          z_snp <span class="op">=</span> <span class="va">z_snp_subset</span>,</span>
<span>                          ld_exprvarfs <span class="op">=</span> <span class="va">ld_exprvarfs</span>,</span>
<span>                          ld_R_dir <span class="op">=</span> <span class="va">ld_R_dir</span>,</span>
<span>                          ld_regions_custom <span class="op">=</span> <span class="va">regions_file</span>,</span>
<span>                          outputdir <span class="op">=</span> <span class="va">outputdir</span>,</span>
<span>                          outname <span class="op">=</span> <span class="va">outname</span>,</span>
<span>                          group_prior <span class="op">=</span> <span class="va">group_prior</span>,</span>
<span>                          group_prior_var <span class="op">=</span> <span class="va">group_prior_var</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="running-ctwas-genome-wide-with-multiple-weights">Running cTWAS genome-wide with multiple weights<a class="anchor" aria-label="anchor" href="#running-ctwas-genome-wide-with-multiple-weights"></a>
</h2>
<p><span style="color: red;">TODO: This section needs to be updated. </span></p>
<p>We now provide a workflow for the genome-wide analysis using cTWAS and multiple groups of weights. This can be computationally intensive. We performed these analyses with 10 cores and 56GB of RAM, although the exact resource requirements will vary with the numbers of genes and variants provided.</p>
<div class="section level3">
<h3 id="estimating-parameters">Estimating parameters<a class="anchor" aria-label="anchor" href="#estimating-parameters"></a>
</h3>
<p>In the <code>multigroup</code> branch, we’ve included an option (<code>fine_map=F</code>) to run only the parameter estimation step in <code>ctwas_rss</code>. This makes it easier to divide larger analyses into shorter individual jobs.</p>
<p>To specify that our gene z-scores come from multiple groups, we simply add a <code>type</code> column specifying the group name for each gene. We use the following code to extract the tissue label from the concatenated gene names. Note that this code parses the specific prediction model names in our analysis and will need to be adjusted when using other data. For example, this code would give the same group name (“Adipose”) to both “Adipose_Subcutaneous” and “Adipose_Visceral_Omentum” if using those tissues.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># set the type for each gene (i.e. tissue)</span></span>
<span><span class="va">z_gene</span><span class="op">$</span><span class="va">type</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">z_gene</span><span class="op">$</span><span class="va">id</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">{</span><span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/strsplit.html" class="external-link">strsplit</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/strsplit.html" class="external-link">strsplit</a></span><span class="op">(</span><span class="va">x</span>, <span class="st">"[|]"</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span>,<span class="st">"_"</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">}</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">z_gene</span><span class="op">)</span></span></code></pre></div>
<p>In addition to the options we specified in the <code>main</code> branch cTWAS tutorial, we’ve added an option (<code>ncore_LDR</code>) to parallelize the creation of the SNP-by-gene and gene-by-gene correlation matrices:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># options used in main branch tutorial</span></span>
<span><span class="va">thin</span> <span class="op">&lt;-</span> <span class="fl">0.1</span></span>
<span><span class="va">max_snp_region</span> <span class="op">&lt;-</span> <span class="fl">20000</span></span>
<span><span class="va">ncore</span> <span class="op">&lt;-</span> <span class="fl">6</span></span>
<span></span>
<span><span class="co"># additional options for the multigroup branch</span></span>
<span><span class="va">fine_map</span> <span class="op">&lt;-</span> <span class="cn">F</span></span>
<span><span class="va">ncore_LDR</span> <span class="op">&lt;-</span> <span class="fl">6</span></span></code></pre></div>
<p>In the <code>multigroup</code> branch, there are several options for how to handle the effect size prior. By default, each group has its own independent effect size parameter (<code>group_prior_var_stucture = "shared"</code>). This allows for group-specific priors but may also be too flexible if these parameters are not well informed. In this example, we specify that the non-SNP groups should share a single effect size parameter (<code>group_prior_var_structure = "shared"</code>). There is also the option to have the non-SNP groups share an effect size with the variants as well (<code>group_prior_var_structure = "shared+snps"</code>). Finally, there is an option to place an inverse-Gamma prior on independent variances for each group (<code>group_prior_var_structure = "inv_gamma"</code>); the prior parameters are controlled using the <code>inv_gamma_shape</code> and <code>inv_gamma_rate</code> options.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># prior variance structure</span></span>
<span><span class="va">group_prior_var_structure</span> <span class="op">&lt;-</span> <span class="st">"shared"</span></span></code></pre></div>
<p>We are now ready to run the parameter estimation step using cTWAS. As specified, this step took approximately 9 hours, using 6 cores.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.exists</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">outputdir</span>, <span class="va">outname</span>, <span class="st">".s2.susieIrssres.Rd"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">outputdir</span>, <span class="va">outname</span>, <span class="st">".s2.susieIrssres.Rd"</span><span class="op">)</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="va">group_prior</span> <span class="op">&lt;-</span> <span class="va">group_prior_rec</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">group_prior_rec</span><span class="op">)</span><span class="op">]</span></span>
<span>  <span class="va">group_prior</span><span class="op">[</span><span class="st">"SNP"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">group_prior</span><span class="op">[</span><span class="st">"SNP"</span><span class="op">]</span><span class="op">*</span><span class="va">thin</span> <span class="co">#adjust for thin argument</span></span>
<span>  </span>
<span>  <span class="va">group_prior_var</span> <span class="op">&lt;-</span> <span class="va">group_prior_var_rec</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">group_prior_var_rec</span><span class="op">)</span><span class="op">]</span></span>
<span>  </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/rm.html" class="external-link">rm</a></span><span class="op">(</span><span class="va">group_prior_rec</span>, <span class="va">group_prior_var_rec</span><span class="op">)</span></span>
<span><span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>  <span class="va">parameters</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ctwas_rss.html">ctwas_rss</a></span><span class="op">(</span>z_gene <span class="op">=</span> <span class="va">z_gene</span>, </span>
<span>                          z_snp <span class="op">=</span> <span class="va">z_snp</span>, </span>
<span>                          ld_exprfs <span class="op">=</span> <span class="va">ld_exprfs</span>, </span>
<span>                          ld_R_dir <span class="op">=</span> <span class="va">ld_R_dir</span>, </span>
<span>                          ld_regions <span class="op">=</span> <span class="st">"EUR"</span>,</span>
<span>                          ld_regions_version <span class="op">=</span> <span class="st">"b38"</span>,</span>
<span>                          outputdir <span class="op">=</span> <span class="va">outputdir</span>, </span>
<span>                          outname <span class="op">=</span> <span class="va">outname</span>,</span>
<span>                          thin <span class="op">=</span> <span class="va">thin</span>,</span>
<span>                          max_snp_region <span class="op">=</span> <span class="va">max_snp_region</span>,</span>
<span>                          ncore <span class="op">=</span> <span class="va">ncore</span>,</span>
<span>                          ncore_LDR <span class="op">=</span> <span class="va">ncore_LDR</span>,</span>
<span>                          group_prior_var_structure <span class="op">=</span> <span class="va">group_prior_var_structure</span>,</span>
<span>                          fine_map <span class="op">=</span> <span class="va">fine_map</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="va">group_prior</span> <span class="op">&lt;-</span> <span class="va">parameters</span><span class="op">$</span><span class="va">group_prior</span></span>
<span>  <span class="va">group_prior_var</span> <span class="op">&lt;-</span> <span class="va">parameters</span><span class="op">$</span><span class="va">group_prior_var</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="va">group_prior</span></span>
<span><span class="va">group_prior_var</span></span></code></pre></div>
<p>Both the <code>group_prior</code> and <code>group_prior_var</code> objects are named vectors containing the parameter estimates. In the <code>main</code> branch, these objects were not named. Note that the effect size estimates for Liver and Adipose are identical because we specified that they would share a single parameter.</p>
<p><span style="color: red;">NOTE: This parameter-estimation-only step still computes the SNP-by-gene and gene-by-gene matrices even though it doesn’t need to. Of the 9 hours that this ran, approximately 1.5 hours were spent computing these matrices (and indexing genes/variants to regions, not sure what proportion of time is just building the matrices). These matrices are computed again during the fine-mapping-only step although they could be recycled. They are only computed once when running the combined parameter estimation and fine-mapping. The simplest option to avoid this computation would be to set the SNP-by-gene matrices to contain all zeros and the gene-by-gene matrices to be identity matrices when <code>fine_map=F</code>. This option that would need to be passed to the <code>index_regions</code> function, which computes the LDR matrices. This would avoid some computation (not sure exactly how much) but the downside is that this would still save a lot of (effectively empty) files in the LDR folder, so it isn’t the cleanest fix. </span></p>
<p>In this example, regions are merged (<code>merge=T</code>) and analyzed jointly when a gene spans a region boundary, which is the default behavior. However, if many prediction models are specified, this could lead to excessive region merging, even if prediction models are sparse. If more than a handful of regions are merged, there many be computational issues in subsequent steps. In my experience running this analysis using prediction models from all 49 GTEx tissues, region merging was not tractable. Turning region merging off (<code>merge=F</code>) discards all genes that span region boundaries. This will resolve computational issues around excessive region merging but will also prevent the analysis of boundary-spanning genes. For larger analyses, turning off region merging will likely be necessary. I think that this issue is currently the biggest limitation to using the <code>multigroup</code> branch.</p>
</div>
<div class="section level3">
<h3 id="assessing-parameter-estimates">Assessing parameter estimates<a class="anchor" aria-label="anchor" href="#assessing-parameter-estimates"></a>
</h3>
<p>We provide code to assess the convergence of the estimated parameters and to compute the proportion of variance explained (PVE) by variants and genes. Note that this function has been modified from the <code>main</code> branch in order to accommodate multiple groups and the named objects returned by the <code>multigroup</code> branch.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ctwas_parameters</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ctwas_summarize_parameters.html">ctwas_summarize_parameters</a></span><span class="op">(</span>outputdir <span class="op">=</span> <span class="va">outputdir</span>, </span>
<span>                                               outname <span class="op">=</span> <span class="va">outname</span>, </span>
<span>                                               gwas_n <span class="op">=</span> <span class="va">gwas_n</span>, </span>
<span>                                               thin <span class="op">=</span> <span class="va">thin</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#estimated prior inclusion probability</span></span>
<span><span class="va">ctwas_parameters</span><span class="op">$</span><span class="va">group_prior</span></span>
<span></span>
<span><span class="co">#estimated prior effect size</span></span>
<span><span class="va">ctwas_parameters</span><span class="op">$</span><span class="va">group_prior_var</span></span>
<span></span>
<span><span class="co">#estimated enrichment of genes over variants</span></span>
<span><span class="va">ctwas_parameters</span><span class="op">$</span><span class="va">enrichment</span></span>
<span></span>
<span><span class="co">#PVE explained by genes and variants</span></span>
<span><span class="va">ctwas_parameters</span><span class="op">$</span><span class="va">group_pve</span></span>
<span></span>
<span><span class="co">#total heritability (sum of PVE)</span></span>
<span><span class="va">ctwas_parameters</span><span class="op">$</span><span class="va">total_pve</span></span>
<span></span>
<span><span class="co">#attributable heritability</span></span>
<span><span class="va">ctwas_parameters</span><span class="op">$</span><span class="va">attributable_pve</span></span>
<span></span>
<span><span class="co">#plot convergence</span></span>
<span><span class="va">ctwas_parameters</span><span class="op">$</span><span class="va">convergence_plot</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="fine-mapping">Fine-mapping<a class="anchor" aria-label="anchor" href="#fine-mapping"></a>
</h3>
<p>We now perform the fine-mapping step with the parameters that we estimated in the previous step. Note that the options used in the fine-mapping step should be the same as in parameter estimation step (e.g. if setting <code>merge=F</code>, make sure to specify it in both functions). The <code>group_prior_var_strucutre</code> option is only necessary during parameter estimation.</p>
<p>As specified, this step took approximately 3.5 hours, using 6 cores.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ctwas_rss.html">ctwas_rss</a></span><span class="op">(</span>z_gene <span class="op">=</span> <span class="va">z_gene</span>, </span>
<span>          z_snp <span class="op">=</span> <span class="va">z_snp</span>, </span>
<span>          ld_exprfs <span class="op">=</span> <span class="va">ld_exprfs</span>, </span>
<span>          ld_R_dir <span class="op">=</span> <span class="va">ld_R_dir</span>, </span>
<span>          ld_regions <span class="op">=</span> <span class="st">"EUR"</span>,</span>
<span>          ld_regions_version <span class="op">=</span> <span class="st">"b38"</span>,</span>
<span>          outputdir <span class="op">=</span> <span class="va">outputdir</span>, </span>
<span>          outname <span class="op">=</span> <span class="va">outname</span>,</span>
<span>          thin <span class="op">=</span> <span class="va">thin</span>,</span>
<span>          max_snp_region <span class="op">=</span> <span class="va">max_snp_region</span>,</span>
<span>          ncore <span class="op">=</span> <span class="va">ncore</span>,</span>
<span>          ncore_LDR <span class="op">=</span> <span class="va">ncore_LDR</span>,</span>
<span>          group_prior <span class="op">=</span> <span class="va">group_prior</span>, </span>
<span>          group_prior_var <span class="op">=</span> <span class="va">group_prior_var</span>, </span>
<span>          estimate_group_prior <span class="op">=</span> <span class="cn">F</span>, </span>
<span>          estimate_group_prior_var <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span></code></pre></div>
<p><span style="color: red;">NOTE: If <code>thin</code> is specified, then for regions with a gene having PIP greater than <code>rerun_gene_PIP</code>, cTWAS makes a final pass, analyzing these regions using the full set of variants. However, this behavior may not be appropriate for the multigroup version. In the context of prediction models with multiple tissues, cTWAS will only be re-run if a single gene/tissue pair is over the threshold. If prediction models across tissues are correlated, there could be a high probability of a gene effect in a region, but with high uncertainty on which tissue, and thus a relatively low PIP on any particular gene/tissue pair. cTWAS would not be re-run in this case, and this region would only contain results using thinned variants. </span></p>
</div>
<div class="section level3">
<h3 id="viewing-the-results">Viewing the results<a class="anchor" aria-label="anchor" href="#viewing-the-results"></a>
</h3>
<p>As before, we will add the gene names to the results (the PredictDB weights use Ensembl IDs as the primary identifier), as well as the z-scores for each SNP and gene. We then show all genes with PIP greater than 0.8, which is the threshold we used in the paper.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#load cTWAS results</span></span>
<span><span class="va">ctwas_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">outputdir</span>, <span class="va">outname</span>, <span class="st">".susieIrss.txt"</span><span class="op">)</span>, header<span class="op">=</span><span class="cn">T</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#load information for all genes</span></span>
<span><span class="va">gene_info</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>gene<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="op">)</span>, genename<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="op">)</span>, gene_type<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="op">)</span>, weight<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">weight</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">sqlite</span> <span class="op">&lt;-</span> <span class="fu">RSQLite</span><span class="fu">::</span><span class="fu">dbDriver</span><span class="op">(</span><span class="st">"SQLite"</span><span class="op">)</span></span>
<span>  <span class="va">db</span> <span class="op">=</span> <span class="fu">RSQLite</span><span class="fu">::</span><span class="fu">dbConnect</span><span class="op">(</span><span class="va">sqlite</span>, <span class="va">weight</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">)</span></span>
<span>  <span class="va">query</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">...</span><span class="op">)</span> <span class="fu">RSQLite</span><span class="fu">::</span><span class="fu">dbGetQuery</span><span class="op">(</span><span class="va">db</span>, <span class="va">...</span><span class="op">)</span></span>
<span>  <span class="va">gene_info_current</span> <span class="op">&lt;-</span> <span class="fu">query</span><span class="op">(</span><span class="st">"select gene, genename, gene_type from extra"</span><span class="op">)</span></span>
<span>  <span class="fu">RSQLite</span><span class="fu">::</span><span class="fu">dbDisconnect</span><span class="op">(</span><span class="va">db</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="va">gene_info_current</span><span class="op">$</span><span class="va">weight</span> <span class="op">&lt;-</span> <span class="va">weight</span><span class="op">[</span><span class="va">i</span><span class="op">]</span></span>
<span>  </span>
<span>  <span class="va">gene_info</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">rbind</a></span><span class="op">(</span><span class="va">gene_info</span>, <span class="va">gene_info_current</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="va">gene_info</span><span class="op">$</span><span class="va">weight</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">gene_info</span><span class="op">$</span><span class="va">weight</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">{</span><span class="fu"><a href="https://rdrr.io/r/base/rev.html" class="external-link">rev</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/strsplit.html" class="external-link">strsplit</a></span><span class="op">(</span><span class="fu">tools</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/tools/fileutils.html" class="external-link">file_path_sans_ext</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span>, <span class="st">"/"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">}</span><span class="op">)</span></span>
<span><span class="va">gene_info</span><span class="op">$</span><span class="va">id</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">gene_info</span><span class="op">$</span><span class="va">gene</span>, <span class="va">gene_info</span><span class="op">$</span><span class="va">weight</span>, sep<span class="op">=</span><span class="st">"|"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#add gene names to cTWAS results</span></span>
<span><span class="va">ctwas_res</span><span class="op">$</span><span class="va">genename</span><span class="op">[</span><span class="va">ctwas_res</span><span class="op">$</span><span class="va">type</span><span class="op">!=</span><span class="st">"SNP"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">gene_info</span><span class="op">$</span><span class="va">genename</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span><span class="va">ctwas_res</span><span class="op">$</span><span class="va">id</span><span class="op">[</span><span class="va">ctwas_res</span><span class="op">$</span><span class="va">type</span><span class="op">!=</span><span class="st">"SNP"</span><span class="op">]</span>, <span class="va">gene_info</span><span class="op">$</span><span class="va">id</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/rm.html" class="external-link">rm</a></span><span class="op">(</span><span class="va">gene_info</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#add z-scores to cTWAS results</span></span>
<span><span class="va">ctwas_res</span><span class="op">$</span><span class="va">z</span><span class="op">[</span><span class="va">ctwas_res</span><span class="op">$</span><span class="va">type</span><span class="op">==</span><span class="st">"SNP"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">z_snp</span><span class="op">$</span><span class="va">z</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span><span class="va">ctwas_res</span><span class="op">$</span><span class="va">id</span><span class="op">[</span><span class="va">ctwas_res</span><span class="op">$</span><span class="va">type</span><span class="op">==</span><span class="st">"SNP"</span><span class="op">]</span>, <span class="va">z_snp</span><span class="op">$</span><span class="va">id</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">ctwas_res</span><span class="op">$</span><span class="va">z</span><span class="op">[</span><span class="va">ctwas_res</span><span class="op">$</span><span class="va">type</span><span class="op">!=</span><span class="st">"SNP"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">z_gene</span><span class="op">$</span><span class="va">z</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/match.html" class="external-link">match</a></span><span class="op">(</span><span class="va">ctwas_res</span><span class="op">$</span><span class="va">id</span><span class="op">[</span><span class="va">ctwas_res</span><span class="op">$</span><span class="va">type</span><span class="op">!=</span><span class="st">"SNP"</span><span class="op">]</span>, <span class="va">z_gene</span><span class="op">$</span><span class="va">id</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span><span class="co">#display the genes with PIP &gt; 0.8</span></span>
<span><span class="va">ctwas_res</span> <span class="op">&lt;-</span> <span class="va">ctwas_res</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="op">-</span><span class="va">ctwas_res</span><span class="op">$</span><span class="va">susie_pip</span><span class="op">)</span>,<span class="op">]</span></span>
<span><span class="va">ctwas_res</span><span class="op">[</span><span class="va">ctwas_res</span><span class="op">$</span><span class="va">type</span><span class="op">!=</span><span class="st">"SNP"</span> <span class="op">&amp;</span> <span class="va">ctwas_res</span><span class="op">$</span><span class="va">susie_pip</span> <span class="op">&gt;</span> <span class="fl">0.8</span>,<span class="op">]</span></span></code></pre></div>
<p>In the context of genes in multiple tissues, it is also useful to evaluate the PIP for each gene combined over tissues:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># aggregate by gene name</span></span>
<span><span class="va">df_gene</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/aggregate.html" class="external-link">aggregate</a></span><span class="op">(</span><span class="va">ctwas_res</span><span class="op">$</span><span class="va">susie_pip</span><span class="op">[</span><span class="va">ctwas_res</span><span class="op">$</span><span class="va">type</span><span class="op">!=</span><span class="st">"SNP"</span><span class="op">]</span>, by<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">ctwas_res</span><span class="op">$</span><span class="va">genename</span><span class="op">[</span><span class="va">ctwas_res</span><span class="op">$</span><span class="va">type</span><span class="op">!=</span><span class="st">"SNP"</span><span class="op">]</span><span class="op">)</span>, FUN<span class="op">=</span><span class="va">sum</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">df_gene</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"genename"</span>, <span class="st">"combined_pip"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># drop duplicated gene names</span></span>
<span><span class="va">df_gene</span> <span class="op">&lt;-</span> <span class="va">df_gene</span><span class="op">[</span><span class="op">!</span><span class="op">(</span><span class="va">df_gene</span><span class="op">$</span><span class="va">genename</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">ctwas_res</span><span class="op">$</span><span class="va">genename</span><span class="op">)</span><span class="op">&gt;</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">weight</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>,<span class="op">]</span></span>
<span></span>
<span><span class="co"># collect tissue-level results</span></span>
<span><span class="va">all_tissue_names</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">ctwas_res</span><span class="op">$</span><span class="va">type</span><span class="op">[</span><span class="va">ctwas_res</span><span class="op">$</span><span class="va">type</span><span class="op">!=</span><span class="st">"SNP"</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="va">df_gene_pips</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="cn">NA</span>, nrow<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">df_gene</span><span class="op">)</span>, ncol<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">all_tissue_names</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">df_gene_pips</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">all_tissue_names</span></span>
<span></span>
<span><span class="va">ctwas_gene_res</span> <span class="op">&lt;-</span> <span class="va">ctwas_res</span><span class="op">[</span><span class="va">ctwas_res</span><span class="op">$</span><span class="va">type</span><span class="op">!=</span><span class="st">"SNP"</span>,<span class="op">]</span></span>
<span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">df_gene</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span>  <span class="va">gene</span> <span class="op">&lt;-</span> <span class="va">df_gene</span><span class="op">$</span><span class="va">genename</span><span class="op">[</span><span class="va">i</span><span class="op">]</span></span>
<span>  <span class="va">ctwas_gene_res_subset</span> <span class="op">&lt;-</span> <span class="va">ctwas_gene_res</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">ctwas_gene_res</span><span class="op">$</span><span class="va">genename</span><span class="op">==</span><span class="va">gene</span><span class="op">)</span>,<span class="op">]</span></span>
<span>  <span class="va">df_gene_pips</span><span class="op">[</span><span class="va">i</span>,<span class="va">ctwas_gene_res_subset</span><span class="op">$</span><span class="va">type</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">ctwas_gene_res_subset</span><span class="op">$</span><span class="va">susie_pip</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="va">df_gene</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">df_gene</span>, <span class="va">df_gene_pips</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/rm.html" class="external-link">rm</a></span><span class="op">(</span><span class="va">df_gene_pips</span>, <span class="va">ctwas_gene_res</span>, <span class="va">ctwas_gene_res_subset</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#sort by combined PIP</span></span>
<span><span class="va">df_gene</span> <span class="op">&lt;-</span> <span class="va">df_gene</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="op">-</span><span class="va">df_gene</span><span class="op">$</span><span class="va">combined_pip</span><span class="op">)</span>,<span class="op">]</span></span>
<span><span class="va">df_gene</span> <span class="op">&lt;-</span> <span class="va">df_gene</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">df_gene</span>, <span class="fl">2</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">{</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/all.html" class="external-link">all</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">)</span><span class="op">}</span><span class="op">)</span><span class="op">]</span> <span class="co">#drop genes that weren't imputed in any tissue</span></span>
<span></span>
<span><span class="co">#genes with PIP&gt;0.8 or 20 highest PIPs</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">df_gene</span>, <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">df_gene</span><span class="op">$</span><span class="va">combined_pip</span><span class="op">&gt;</span><span class="fl">0.8</span><span class="op">)</span>, <span class="fl">20</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><span style="color: red;">NOTE: To visualize individual loci, the <code>ctwas_locus_plot</code> function will need to be updated to accommodate multiple groups. </span></p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Siming Zhao, Wesley Crouse, Sheng Qian, Kaixuan Luo, Matthew Stephens, Xin He.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
