<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Postprocessing problematic regions • ctwas</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Postprocessing problematic regions">
<meta property="og:description" content="ctwas">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ctwas</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.36</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorial
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/data_preparation.html">Data preparation</a>
    </li>
    <li>
      <a href="../articles/postprocessing_regions.html">Postprocessing problematic regions</a>
    </li>
    <li>
      <a href="../articles/ctwas_summary_statistics.html">Using cTWAS single-group version with summary statistics</a>
    </li>
    <li>
      <a href="../articles/multigroup_ctwas_summary_statistics.html">Using cTWAS multi-group version with summary statistics</a>
    </li>
    <li>
      <a href="../articles/harmonization.html">Harmonization</a>
    </li>
    <li>
      <a href="../articles/regionmerging.html">Region Merging</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/xinhe-lab/ctwas" class="external-link">Source</a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="postprocessing_regions_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Postprocessing problematic regions</h1>
                        <h4 data-toc-skip class="author">Kaixuan Luo</h4>
            
            <h4 data-toc-skip class="date">2024-02-13</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/xinhe-lab/ctwas/blob/main/vignettes/postprocessing_regions.Rmd" class="external-link"><code>vignettes/postprocessing_regions.Rmd</code></a></small>
      <div class="hidden name"><code>postprocessing_regions.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p>This document is a tutorial for fine-mapping problematic regions with LD mismatches. We will use the <code>multigroup</code> version of the package, but the procedure below works for using a single prediction model, or multiple prediction models.</p>
</div>
<div class="section level2">
<h2 id="getting-started">Getting started<a class="anchor" aria-label="anchor" href="#getting-started"></a>
</h2>
<p>Install the <code>multigroup_kevin</code> branch of <code>cTWAS</code> package</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/remotes/man/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"xinhe-lab/ctwas"</span>, ref <span class="op">=</span> <span class="st">"multigroup_kevin"</span><span class="op">)</span></span></code></pre></div>
<p>Load the package and set the working directory where you want to perform the analysis.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/xinhe-lab/ctwas" class="external-link">ctwas</a></span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># set the working directory interactively</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/getwd.html" class="external-link">setwd</a></span><span class="op">(</span><span class="st">"/project2/xinhe/shared_data/ctwas_tutorial"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># set the working directory when compiling this document with Knitr</span></span>
<span><span class="fu">knitr</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/knitr/man/opts_knit.html" class="external-link">opts_knit</a></span><span class="op">$</span><span class="fu">set</span><span class="op">(</span>root.dir <span class="op">=</span> <span class="st">"/project2/xinhe/shared_data/ctwas_tutorial"</span><span class="op">)</span></span></code></pre></div>
<p>Follow the section on data preparation about harmonizing GWAS z-scores and LD reference and LD mismatch diagnosis.</p>
<p>First, we load the LD mismatch diagnosis result from running the <code>preprocess_z_ld</code> step, and gets a list of problematic SNPs with rsIDs.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">outputdir</span> <span class="op">&lt;-</span> <span class="st">"multigroup/processed_data/"</span></span>
<span><span class="va">outname</span> <span class="op">&lt;-</span> <span class="st">"example"</span></span>
<span></span>
<span><span class="va">ld_mismatch_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">outputdir</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">outname</span>, <span class="st">".ld_mismatch_res.RDS"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">problematic_snps</span> <span class="op">&lt;-</span> <span class="va">ld_mismatch_res</span><span class="op">$</span><span class="va">problematic_snps</span></span></code></pre></div>
<p>You can also use those problematic SNPs (rsIDs) obtained by other tools, e.g. DENTIST.</p>
<p>Then, let’s load the <code>cTWAS</code> full result, and harmonized <code>z_snp</code> and <code>z_gene</code> data.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ctwas_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">outputdir</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">outname</span>, <span class="st">".susieIrss.txt"</span><span class="op">)</span><span class="op">)</span>, header<span class="op">=</span><span class="cn">T</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># load processed z_snp file</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span>file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">outputdir</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">outname</span>, <span class="st">".harmonized.z_snp.Rd"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># load processed z_snp file</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span>file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">outputdir</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">outname</span>, <span class="st">"_z_gene.Rd"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># harmonized prediction models</span></span>
<span><span class="va">weight_dir</span> <span class="op">&lt;-</span> <span class="st">"multigroup/processed_data/weights_nolnc/"</span></span>
<span><span class="va">weight</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">weight_dir</span>, <span class="st">"mashr_Liver_nolnc_harmonized.db"</span><span class="op">)</span>, </span>
<span>            <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">weight_dir</span>, <span class="st">"mashr_Adipose_Subcutaneous_nolnc_harmonized.db"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>We will rerun the fine-mapping step with L = 1 for the regions containing high PIP (&gt;0.5) problematic SNPs or genes problematic SNPs in their weights.</p>
<p>The following function runs fine-mapping for the regions in parallel (with 6 cores) and saves results to <code>[outname].L1.susieIrss.txt</code>.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/rerun_ctwas_finemap_regions_L1_rss.html">rerun_ctwas_finemap_regions_L1_rss</a></span><span class="op">(</span><span class="va">z_snp</span>,</span>
<span>                                   <span class="va">z_gene</span>,</span>
<span>                                   problematic_snps <span class="op">=</span> <span class="va">problematic_snps</span>,</span>
<span>                                   weight <span class="op">=</span> <span class="va">weight</span>,</span>
<span>                                   pip_thresh <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>                                   ctwas_outputdir <span class="op">=</span> <span class="va">outputdir</span>,</span>
<span>                                   ctwas_outname <span class="op">=</span> <span class="va">outname</span>,</span>
<span>                                   ncore <span class="op">=</span> <span class="fl">6</span><span class="op">)</span></span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Siming Zhao, Wesley Crouse, Sheng Qian, Kaixuan Luo, Matthew Stephens, Xin He.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
