<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Simple cTWAS tutorial • ctwas</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Simple cTWAS tutorial">
<meta property="og:description" content="ctwas">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ctwas</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.2.46</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorial
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/simple_ctwas_tutorial.html">Simple cTWAS tutorial</a>
    </li>
    <li>
      <a href="../articles/preparing_ctwas_input_data.html">Preparing cTWAS input data</a>
    </li>
    <li>
      <a href="../articles/ctwas_main_function.html">Using cTWAS main function</a>
    </li>
    <li>
      <a href="../articles/ctwas_modules.html">Using cTWAS modules</a>
    </li>
    <li>
      <a href="../articles/summarizing_ctwas_results.html">Summarizing cTWAS results</a>
    </li>
    <li>
      <a href="../articles/postprocessing_ctwas_results.html">Post-processing cTWAS results</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/xinhe-lab/ctwas/tree/multigroup_test" class="external-link">Source</a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="simple_ctwas_tutorial_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Simple cTWAS tutorial</h1>
                        <h4 data-toc-skip class="author">Kaixuan Luo, Sheng Qian</h4>
            
            <h4 data-toc-skip class="date">2024-06-28</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/xinhe-lab/ctwas/blob/multigroup_test/vignettes/simple_ctwas_tutorial.Rmd" class="external-link"><code>vignettes/simple_ctwas_tutorial.Rmd</code></a></small>
      <div class="hidden name"><code>simple_ctwas_tutorial.Rmd</code></div>

    </div>

    
    
<p>This document demonstrates how to run a simple analysis using cTWAS with GWAS summary statistics. Running cTWAS involves several steps: preparing input data, computing z-scores of molecular traits, assembling the input data for the regions, estimating parameters, screening regions with strong signals, and fine-mapping the variants and molecular traits.</p>
<p>In this tutorial, we will show how to perform cTWAS main analysis with summary statistics without using LD matrices.</p>
<p>Load the <code>ctwas</code> package.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/xinhe-lab/ctwas/" class="external-link">ctwas</a></span><span class="op">)</span></span></code></pre></div>
<p>We require first running the data preprocessing and harmonization steps before running cTWAS analysis. Please read the “Preparing cTWAS input data” tutorial to prepare cTWAS input data.</p>
<p>In this tutorial, we use sample data for chromosome 16 provided in the package.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># load z_snp sample data</span></span>
<span><span class="va">z_snp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/sample_data"</span>, <span class="st">"LDL_example.z_snp.RDS"</span>, package <span class="op">=</span> <span class="st">"ctwas"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>We also need the GWAS sample size for summarizing the cTWAS result.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># GWAS sample size</span></span>
<span><span class="va">gwas_n</span> <span class="op">&lt;-</span> <span class="fl">343621</span></span></code></pre></div>
<div class="section level2">
<h2 id="reference-data">Reference data<a class="anchor" aria-label="anchor" href="#reference-data"></a>
</h2>
<p>cTWAS assumes that the genome is partitioned into approximately independent LD regions. LD reference information can be provided as genetic correlation matrices (termed “R matrices”) of these regions.</p>
<p>Note: when running cTWAS without LD, you can skip “LD matrices”. But you will still need to have a list of SNPs <code>snp_info</code> as a “reference”, with the information of these SNPs. See the section “SNP info”.</p>
<p>It is critical that the genome build (e.g. hg38) of the LD reference matches the genome build used to train the prediction models. The genome build of the GWAS summary statistics does not matter because variant positions are determined by the LD reference.</p>
<p>The choice of LD reference population is important for fine-mapping. Best practice for fine-mapping is to use an in-sample LD reference (LD computed using the subjects in the GWAS sample). If in-sample LD reference is not an option, the LD reference should be as representative of the population in the GWAS sample as possible.</p>
<div class="section level3">
<h3 id="region-info-and-snp-info">Region info and SNP info<a class="anchor" aria-label="anchor" href="#region-info-and-snp-info"></a>
</h3>
<p>We require a data frame <code>region_info</code> containing the region definitions, with the following columns: “chrom”, “start”, “stop”, for the genomic coordinates of the regions, and “region_id” for the IDs of the regions (by default, we use [chrom_start_stop] as region IDs).</p>
<p>Here we use the b38 European LDetect blocks, included in the package.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">region_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/ldetect"</span>, <span class="st">"EUR.b38.bed"</span>, package <span class="op">=</span> <span class="st">"ctwas"</span><span class="op">)</span></span>
<span><span class="va">region_info</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="va">region_file</span>, header <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">region_info</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"chrom"</span>, <span class="st">"start"</span>, <span class="st">"stop"</span><span class="op">)</span></span>
<span><span class="va">region_info</span><span class="op">$</span><span class="va">chrom</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">"chr"</span>, <span class="st">""</span>, <span class="va">region_info</span><span class="op">$</span><span class="va">chrom</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">region_info</span><span class="op">$</span><span class="va">region_id</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">region_info</span><span class="op">$</span><span class="va">chr</span>, <span class="st">"_"</span>, <span class="va">region_info</span><span class="op">$</span><span class="va">start</span>, <span class="st">"_"</span>, <span class="va">region_info</span><span class="op">$</span><span class="va">stop</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="ld-and-snp-info">LD and SNP info<a class="anchor" aria-label="anchor" href="#ld-and-snp-info"></a>
</h3>
<p>We need a list of variants as a “reference”, with the positions and allele information of these variants. We provide a function, <code><a href="../reference/preprocess_region_LD_snp_info.html">preprocess_region_LD_snp_info()</a></code> to preprocess region definitions and map variants from the LD reference to regions, which takes input of <code>region_info</code> and a data frame <code>ref_snp_info</code> with variant information from the entire genome, and returns processed <code>region_info</code>, <code>snp_info</code>.</p>
<p>We have the lists of reference variant information from all the LD matrices in the genome in <a href="https://uchicago.box.com/s/t089or92dkovv0epkrjvxq8r9db9ys99" class="external-link">hg38</a> and <a href="https://uchicago.box.com/s/ufko2gjagcb693dob4khccqubuztb9pz" class="external-link">hg19</a>.</p>
<p>In this example, we use variant information for chr16 in hg38.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">example_chrom</span> <span class="op">&lt;-</span> <span class="fl">16</span></span>
<span><span class="va">ref_snp_info_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/sample_data"</span>, <span class="st">"ukb_b38_0.1_chr16_var_info.Rvar.gz"</span>, package <span class="op">=</span> <span class="st">"ctwas"</span><span class="op">)</span></span>
<span><span class="va">ref_snp_info</span> <span class="op">&lt;-</span> <span class="fu">data.table</span><span class="fu">::</span><span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="va">ref_snp_info_file</span>, sep <span class="op">=</span> <span class="st">"\t"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/preprocess_region_LD_snp_info.html">preprocess_region_LD_snp_info</a></span><span class="op">(</span><span class="va">region_info</span>, </span>
<span>                                     ref_snp_info <span class="op">=</span> <span class="va">ref_snp_info</span>, </span>
<span>                                     chrom <span class="op">=</span> <span class="va">example_chrom</span>, </span>
<span>                                     use_LD <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">region_info</span> <span class="op">&lt;-</span> <span class="va">res</span><span class="op">$</span><span class="va">region_info</span></span>
<span><span class="va">snp_info</span> <span class="op">&lt;-</span> <span class="va">res</span><span class="op">$</span><span class="va">snp_info</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">region_info</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##   chrom   start    stop          region_id</span></span>
<span><span class="co">## 1    16   10054 1157206   16_10054_1157206</span></span>
<span><span class="co">## 2    16 1157206 2714828 16_1157206_2714828</span></span>
<span><span class="co">## 3    16 2714828 3951195 16_2714828_3951195</span></span>
<span><span class="co">## 4    16 3951195 5068344 16_3951195_5068344</span></span>
<span><span class="co">## 5    16 5068344 5841579 16_5068344_5841579</span></span>
<span><span class="co">## 6    16 5841579 6843550 16_5841579_6843550</span></span></code></pre>
<p>Note: in this simple tutorial, we use the “no-LD” version, so we use <code>use_LD = FALSE</code>. Please see the other tutorials for running full cTWAS analysis using LD.</p>
</div>
<div class="section level3">
<h3 id="harmonizing-gwas-z-scores-and-the-reference-data">Harmonizing GWAS z-scores and the reference data<a class="anchor" aria-label="anchor" href="#harmonizing-gwas-z-scores-and-the-reference-data"></a>
</h3>
<p>The <code><a href="../reference/preprocess_z_snp.html">preprocess_z_snp()</a></code> function harmonizes GWAS z-scores and the reference data in <code>snp_info</code>. We drop multiallelic variants and drop strand ambiguous variants, by default.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># preprocess z_snp</span></span>
<span><span class="va">z_snp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/preprocess_z_snp.html">preprocess_z_snp</a></span><span class="op">(</span><span class="va">z_snp</span>, </span>
<span>                          <span class="va">snp_info</span>, </span>
<span>                          drop_multiallelic <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                          drop_strand_ambig <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 2024-06-28 18:59:17 INFO::Preprocessing z_snp...</span></span>
<span><span class="co">## 2024-06-28 18:59:17 INFO::z_snp has 407828 variants in total</span></span>
<span><span class="co">## 2024-06-28 18:59:17 INFO::267361 variants left after filtering by snp_info</span></span>
<span><span class="co">## 2024-06-28 18:59:17 INFO::Drop 30 multiallelic variants</span></span>
<span><span class="co">## 2024-06-28 18:59:17 INFO::Harmonize 267331 variants in both GWAS and LD reference</span></span>
<span><span class="co">## 2024-06-28 18:59:18 INFO::Flip signs for 267289 (99.98%) variants</span></span>
<span><span class="co">## 2024-06-28 18:59:18 INFO::Remove 45738 strand ambiguous variants</span></span>
<span><span class="co">## 2024-06-28 18:59:18 INFO::221593 variants left after preprocessing and harmonization.</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="harmonizing-prediction-models-and-the-reference-data">Harmonizing prediction models and the reference data<a class="anchor" aria-label="anchor" href="#harmonizing-prediction-models-and-the-reference-data"></a>
</h3>
<p>The <code>preprocess_weight()</code> function harmonizes the PredictDB/FUSION prediction models and LD reference.</p>
<p>In this version of cTWAS, we allow the joint analysis of multiple groups of molecular traits. This could be: eQTL of multiple tissues; or eQTLs, splicing QTLs and other types of QTLs in a single tissue. In a more complex setting, multiple types of QTL data from multiple tissues/cell types. Each group is defined by its “type” (kind of molecular traits), and “context” (tissue, cell type, condition, etc.). So, we specify the <code>type</code> and <code>context</code> arguments for each weight file.</p>
<p>In this example, we use liver and subcutaneous adipose gene expression models. We specify the <code>type</code> and <code>context</code> arguments for each weight file. We preprocess each of the weight files, and then simply concatenate them to get a list of all processed weights from different types or contexts.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># preprocess weights</span></span>
<span><span class="va">weight_liver_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/sample_data"</span>, <span class="st">"expression_Liver.db"</span>, package <span class="op">=</span> <span class="st">"ctwas"</span><span class="op">)</span></span>
<span><span class="va">weights_liver</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/preprocess_weights.html">preprocess_weights</a></span><span class="op">(</span><span class="va">weight_liver_file</span>,</span>
<span>                                    <span class="va">region_info</span>,</span>
<span>                                    <span class="va">z_snp</span><span class="op">$</span><span class="va">id</span>,</span>
<span>                                    <span class="va">snp_info</span>,</span>
<span>                                    type <span class="op">=</span> <span class="st">"expression"</span>,</span>
<span>                                    context <span class="op">=</span> <span class="st">"liver"</span>,</span>
<span>                                    weight_format <span class="op">=</span> <span class="st">"PredictDB"</span>,</span>
<span>                                    drop_strand_ambig <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                                    scale_predictdb_weights <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                                    load_predictdb_LD <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                                    filter_protein_coding_genes <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 2024-06-28 18:59:18 INFO::Load weight: /tmp/jobs/40635171/RtmpFaF4kh/temp_libpath664a63bd4155/ctwas/extdata/sample_data/expression_Liver.db</span></span>
<span><span class="co">## 2024-06-28 18:59:18 INFO::type: expression</span></span>
<span><span class="co">## 2024-06-28 18:59:18 INFO::context: liver</span></span>
<span><span class="co">## 2024-06-28 18:59:18 INFO::Loading PredictDB weights ...</span></span>
<span><span class="co">## 2024-06-28 18:59:19 INFO::Keep protein coding genes only</span></span>
<span><span class="co">## 2024-06-28 18:59:19 INFO::weight_name: expression_Liver</span></span>
<span><span class="co">## 2024-06-28 18:59:19 INFO::Remove 1 genes without predictdb LD</span></span>
<span><span class="co">## 2024-06-28 18:59:19 INFO::Number of genes with weights: 11479</span></span>
<span><span class="co">## 2024-06-28 18:59:19 INFO::Number of variants in weights: 16929</span></span>
<span><span class="co">## 2024-06-28 18:59:19 INFO::529 genes and 410 variants left after filtering by GWAS and LD reference</span></span>
<span><span class="co">## 2024-06-28 18:59:19 INFO::Harmonizing and processing weights ...</span></span></code></pre>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">weight_adipose_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/sample_data"</span>, <span class="st">"expression_Adipose_Subcutaneous.db"</span>, package <span class="op">=</span> <span class="st">"ctwas"</span><span class="op">)</span></span>
<span><span class="va">weights_adipose</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/preprocess_weights.html">preprocess_weights</a></span><span class="op">(</span><span class="va">weight_adipose_file</span>,</span>
<span>                                      <span class="va">region_info</span>,</span>
<span>                                      <span class="va">z_snp</span><span class="op">$</span><span class="va">id</span>,</span>
<span>                                      <span class="va">snp_info</span>,</span>
<span>                                      type <span class="op">=</span> <span class="st">"expression"</span>,</span>
<span>                                      context <span class="op">=</span> <span class="st">"adipose"</span>,</span>
<span>                                      weight_format <span class="op">=</span> <span class="st">"PredictDB"</span>,</span>
<span>                                      drop_strand_ambig <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                                      scale_predictdb_weights <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                                      load_predictdb_LD <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                                      filter_protein_coding_genes <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 2024-06-28 18:59:20 INFO::Load weight: /tmp/jobs/40635171/RtmpFaF4kh/temp_libpath664a63bd4155/ctwas/extdata/sample_data/expression_Adipose_Subcutaneous.db</span></span>
<span><span class="co">## 2024-06-28 18:59:20 INFO::type: expression</span></span>
<span><span class="co">## 2024-06-28 18:59:20 INFO::context: adipose</span></span>
<span><span class="co">## 2024-06-28 18:59:20 INFO::Loading PredictDB weights ...</span></span>
<span><span class="co">## 2024-06-28 18:59:20 INFO::Keep protein coding genes only</span></span>
<span><span class="co">## 2024-06-28 18:59:20 INFO::weight_name: expression_Adipose_Subcutaneous</span></span>
<span><span class="co">## 2024-06-28 18:59:20 INFO::Remove 0 genes without predictdb LD</span></span>
<span><span class="co">## 2024-06-28 18:59:20 INFO::Number of genes with weights: 13102</span></span>
<span><span class="co">## 2024-06-28 18:59:20 INFO::Number of variants in weights: 23737</span></span>
<span><span class="co">## 2024-06-28 18:59:20 INFO::720 genes and 484 variants left after filtering by GWAS and LD reference</span></span>
<span><span class="co">## 2024-06-28 18:59:20 INFO::Harmonizing and processing weights ...</span></span></code></pre>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># concatenate weights from different types or contexts</span></span>
<span><span class="va">weights</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">weights_liver</span>, <span class="va">weights_adipose</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="running-ctwas-analysis-without-ld">Running cTWAS analysis without LD<a class="anchor" aria-label="anchor" href="#running-ctwas-analysis-without-ld"></a>
</h2>
<p>We use the function <code><a href="../reference/ctwas_sumstats_noLD.html">ctwas_sumstats_noLD()</a></code> to run cTWAS analysis without LD. It takes preprocessed GWAS z-scores (<code>z_snp</code>), <code>weights</code>, <code>region_info</code>, <code>snp_info</code> as input, and will perform the main steps: computing z_scores of molecular traits, assembling input <code>region_data</code>, estimating parameters, screening regions, and finemapping.</p>
<p>If <code>z_gene</code> is already computed when preparing cTWAS input data, we could specify the <code>z_gene</code> argument in the function, then it will skip computing z-scores of molecular traits.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ctwas_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ctwas_sumstats_noLD.html">ctwas_sumstats_noLD</a></span><span class="op">(</span><span class="va">z_snp</span>,</span>
<span>                                 <span class="va">weights</span>,</span>
<span>                                 <span class="va">region_info</span>,</span>
<span>                                 <span class="va">snp_info</span>,</span>
<span>                                 thin <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>                                 niter_prefit <span class="op">=</span> <span class="fl">3</span>,</span>
<span>                                 niter <span class="op">=</span> <span class="fl">30</span>,</span>
<span>                                 group_prior_var_structure <span class="op">=</span> <span class="st">"shared_type"</span>,</span>
<span>                                 maxSNP <span class="op">=</span> <span class="fl">20000</span>,</span>
<span>                                 min_nonSNP_PIP <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>                                 ncore <span class="op">=</span> <span class="fl">6</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># saveRDS(ctwas_res, file.path(outputdir, paste0(outname, ".ctwas_sumstats_noLD_res.RDS")))</span></span></code></pre></div>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">param</span> <span class="op">&lt;-</span> <span class="va">ctwas_res</span><span class="op">$</span><span class="va">param</span></span>
<span><span class="va">finemap_res</span> <span class="op">&lt;-</span> <span class="va">ctwas_res</span><span class="op">$</span><span class="va">finemap_res</span></span>
<span><span class="va">boundary_genes</span> <span class="op">&lt;-</span> <span class="va">ctwas_res</span><span class="op">$</span><span class="va">boundary_genes</span></span>
<span><span class="va">z_gene</span> <span class="op">&lt;-</span> <span class="va">ctwas_res</span><span class="op">$</span><span class="va">z_gene</span></span>
<span><span class="va">region_data</span> <span class="op">&lt;-</span> <span class="va">ctwas_res</span><span class="op">$</span><span class="va">region_data</span></span>
<span><span class="va">screened_region_data</span> <span class="op">&lt;-</span> <span class="va">ctwas_res</span><span class="op">$</span><span class="va">screened_region_data</span></span></code></pre></div>
<p>Major argument settings:</p>
<ul>
<li>
<code>z_snp</code> is a data frame with preprocessed GWAS z-scores.</li>
<li>
<code>weights</code> is a list of preprocessed weights from PredictDB or FUSION prediction models.</li>
<li>
<code>region_info</code> is a data frame with genome coordinates, LD matrix and SNP info of all the regions</li>
<li>
<code>snp_info</code> is a list of reference SNP info in each of the regions.</li>
<li>The <code>thin</code> argument randomly selects a subset of variants (10% when thin = 0.1) to use during the parameter estimation and screening regions, reducing computation.</li>
<li>The <code>niter_prefit</code> argument and <code>niter</code> argument sets the number of EM iterations for estimating parameters. We set <code>niter_prefit = 3</code>, and <code>niter = 30</code> by default.</li>
<li>There are several options for how to handle the prior effect size parameters by specifying the <code>group_prior_var_stucture</code>. “shared_type” (default option) allows all groups in one molecular QTL type to share the same variance parameter. “shared_context” allows all groups in one context (tissue, cell type, condition) to share the same variance parameter. “shared_nonSNP” allows all non-SNP groups to share the same variance parameter. “shared_all” allows all groups to share the same variance parameter. “independent” allows all groups to have their own separate variance parameters.</li>
<li>The <code>maxSNP</code> sets a maximum on the number of variants that can be in a single region to prevent memory issues during fine-mapping.</li>
<li>
<code>min_nonSNP_PIP</code> is a threshold for selecting regions with strong non-SNP signals for finemapping. We set <code>min_nonSNP_PIP = 0.5</code> by default, so regions with total non-SNP signals greater than 0.5 will be selected for finemapping.</li>
<li>The <code>ncore</code> argument specifies the number of cores to use when parallelizing over regions.</li>
</ul>
<p>The function returns a list including: estimated parameters, fine-mapping results, boundary genes, z-scores of molecular traits, assembled region data for all the regions and region data with full SNPs for selected regions.</p>
</div>
<div class="section level2">
<h2 id="summarizing-ctwas-result">Summarizing cTWAS result<a class="anchor" aria-label="anchor" href="#summarizing-ctwas-result"></a>
</h2>
<div class="section level3">
<h3 id="assessing-parameters-and-computing-pve">Assessing parameters and computing PVE<a class="anchor" aria-label="anchor" href="#assessing-parameters-and-computing-pve"></a>
</h3>
<p>We could use the function <code><a href="../reference/summarize_param.html">summarize_param()</a></code> to assess the convergence of the estimated parameters and to compute the proportion of variance explained (PVE) by variants and genes.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ctwas_parameters</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/summarize_param.html">summarize_param</a></span><span class="op">(</span><span class="va">param</span>, <span class="va">gwas_n</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Estimated prior inclusion probability</span></span>
<span><span class="va">ctwas_parameters</span><span class="op">$</span><span class="va">group_prior</span></span></code></pre></div>
<pre><code><span><span class="co">##   expression|liver expression|adipose                SNP </span></span>
<span><span class="co">##       0.0112258728       0.0032677115       0.0002253614</span></span></code></pre>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Estimated prior effect size:</span></span>
<span><span class="va">ctwas_parameters</span><span class="op">$</span><span class="va">group_prior_var</span></span></code></pre></div>
<pre><code><span><span class="co">##   expression|liver expression|adipose                SNP </span></span>
<span><span class="co">##         159.366028         159.366028           4.889058</span></span></code></pre>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Estimated enrichment of molecular traits over variants:</span></span>
<span><span class="va">ctwas_parameters</span><span class="op">$</span><span class="va">enrichment</span></span></code></pre></div>
<pre><code><span><span class="co">##   expression|liver expression|adipose </span></span>
<span><span class="co">##           49.81275           14.49987</span></span></code></pre>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># PVE explained by molecular traits and variants:</span></span>
<span><span class="va">ctwas_parameters</span><span class="op">$</span><span class="va">group_pve</span></span></code></pre></div>
<pre><code><span><span class="co">##   expression|liver expression|adipose                SNP </span></span>
<span><span class="co">##       0.0021346173       0.0007335085       0.0005839915</span></span></code></pre>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Total heritability (sum of PVE)</span></span>
<span><span class="va">ctwas_parameters</span><span class="op">$</span><span class="va">total_pve</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 0.003452117</span></span></code></pre>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Attributable heritability</span></span>
<span><span class="va">ctwas_parameters</span><span class="op">$</span><span class="va">attributable_pve</span></span></code></pre></div>
<pre><code><span><span class="co">##   expression|liver expression|adipose                SNP </span></span>
<span><span class="co">##          0.6183502          0.2124807          0.1691691</span></span></code></pre>
<p>Make convergence plots of estimated parameters</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/make_convergence_plots.html">make_convergence_plots</a></span><span class="op">(</span><span class="va">param</span>, <span class="va">gwas_n</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="simple_ctwas_tutorial_files/figure-html/convergence_plot-1.png" alt="&amp;nbsp;" width="840"><p class="caption">
 
</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="interpreting-ctwas-results">Interpreting cTWAS results<a class="anchor" aria-label="anchor" href="#interpreting-ctwas-results"></a>
</h2>
<p>To interpret and visualize the results, we first need to add gene annotations (gene names and gene types) to the finemapping result.</p>
<p>We need a user defined data frame <code>gene_annot</code> that should contain the following columns: ‘chrom’, ‘start’, ‘end’, ‘gene_id’, ‘gene_name’ and ‘gene_type’ for each gene or molecular trait.</p>
<p>If you only used gene expression data, you can use the following function to obtain the gene annotations from the Ensembl database. We use <code>EnsDb.Hsapiens.v86</code> for the example data, please choose the Ensembl database for your specific data.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">EnsDb.Hsapiens.v86</span><span class="op">)</span></span>
<span><span class="va">ens_db</span> <span class="op">&lt;-</span> <span class="va">EnsDb.Hsapiens.v86</span></span>
<span><span class="va">finemap_gene_res</span> <span class="op">&lt;-</span> <span class="va">finemap_res</span><span class="op">[</span><span class="va">finemap_res</span><span class="op">$</span><span class="va">type</span><span class="op">!=</span><span class="st">"SNP"</span>,<span class="op">]</span></span>
<span><span class="va">gene_ids</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/strsplit.html" class="external-link">strsplit</a></span><span class="op">(</span><span class="va">finemap_gene_res</span><span class="op">$</span><span class="va">id</span>, split <span class="op">=</span> <span class="st">"[|]"</span><span class="op">)</span>, <span class="st">"[["</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">gene_annot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_gene_annot_from_ens_db.html">get_gene_annot_from_ens_db</a></span><span class="op">(</span><span class="va">ens_db</span>, <span class="va">gene_ids</span><span class="op">)</span></span></code></pre></div>
<p>We then use <code><a href="../reference/anno_finemap_res.html">anno_finemap_res()</a></code> function to add gene annotations to the finemapping result, and use gene mid-points to represent gene positions.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># add gene annotations and use gene mid-points to represent gene positions</span></span>
<span><span class="va">finemap_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/anno_finemap_res.html">anno_finemap_res</a></span><span class="op">(</span><span class="va">finemap_res</span>, </span>
<span>                                <span class="va">snp_info</span>,</span>
<span>                                <span class="va">gene_annot</span>,</span>
<span>                                use_gene_pos <span class="op">=</span> <span class="st">"mid"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 2024-06-28 18:59:25 INFO::Annotating ctwas finemapping result ...</span></span>
<span><span class="co">## 2024-06-28 18:59:25 INFO::add gene_name and gene_type</span></span>
<span><span class="co">## 2024-06-28 18:59:25 INFO::use gene mid positions</span></span>
<span><span class="co">## 2024-06-28 18:59:25 INFO::add SNP positions from snp_info</span></span></code></pre>
<p>We list all molecular traits with PIP &gt; 0.8, which is the threshold we used in the paper.</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># select PIP &gt; 0.8</span></span>
<span><span class="va">sig_finemap_res</span> <span class="op">&lt;-</span> <span class="va">finemap_res</span><span class="op">[</span><span class="va">finemap_res</span><span class="op">$</span><span class="va">type</span><span class="op">!=</span><span class="st">"SNP"</span> <span class="op">&amp;</span> <span class="va">finemap_res</span><span class="op">$</span><span class="va">susie_pip</span> <span class="op">&gt;</span> <span class="fl">0.8</span>,<span class="op">]</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">sig_finemap_res</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="op">-</span><span class="va">sig_finemap_res</span><span class="op">$</span><span class="va">susie_pip</span><span class="op">)</span>,<span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##    chrom      pos            gene_id gene_name      gene_type</span></span>
<span><span class="co">## 56    16 72070235  ENSG00000261701.6       HPR protein_coding</span></span>
<span><span class="co">## 36    16 56972848 ENSG00000087237.10      CETP protein_coding</span></span>
<span><span class="co">##                                                    id       type context</span></span>
<span><span class="co">## 56                 ENSG00000261701.6|expression_Liver expression   liver</span></span>
<span><span class="co">## 36 ENSG00000087237.10|expression_Adipose_Subcutaneous expression adipose</span></span>
<span><span class="co">##                 group         z            region_id susie_pip      mu2</span></span>
<span><span class="co">## 56   expression|liver -18.40315 16_71020125_72901251         1 335.4591</span></span>
<span><span class="co">## 36 expression|adipose  13.74984 16_55869862_57630418         1 187.7014</span></span></code></pre>
<p>When we have multiple contexts (tissues), it is useful to evaluate the combined PIP for each molecular trait across contexts:</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># combine gene PIPs by context</span></span>
<span><span class="va">combined_pip_by_context</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/combine_gene_pips.html">combine_gene_pips</a></span><span class="op">(</span><span class="va">finemap_res</span>, </span>
<span>                                             by <span class="op">=</span> <span class="st">"context"</span>,</span>
<span>                                             filter_protein_coding_genes <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">combined_pip_by_context</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##    gene_name combined_pip liver_pip adipose_pip</span></span>
<span><span class="co">## 22       HPR            1         1           0</span></span>
<span><span class="co">## 9       CETP            1         0           1</span></span>
<span><span class="co">## 33     NLRC5            0         0           0</span></span>
<span><span class="co">## 1     ADGRG1            0        NA           0</span></span>
<span><span class="co">## 2       AMFR            0         0          NA</span></span>
<span><span class="co">## 3     ATXN1L            0         0           0</span></span></code></pre>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># select combined PIP &gt; 0.8</span></span>
<span><span class="va">sig_gene_pips_df</span> <span class="op">&lt;-</span> <span class="va">combined_pip_by_context</span><span class="op">[</span><span class="va">combined_pip_by_context</span><span class="op">$</span><span class="va">combined_pip</span> <span class="op">&gt;</span> <span class="fl">0.8</span>, <span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">sig_gene_pips_df</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##    gene_name combined_pip liver_pip adipose_pip</span></span>
<span><span class="co">## 22       HPR            1         1           0</span></span>
<span><span class="co">## 9       CETP            1         0           1</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="visualizing-ctwas-results">Visualizing cTWAS results<a class="anchor" aria-label="anchor" href="#visualizing-ctwas-results"></a>
</h2>
<p>We could make locus plots for regions of interest. We could zoom in the region by specifying the <code>locus_range</code>.</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/make_locusplot.html">make_locusplot</a></span><span class="op">(</span><span class="va">finemap_res</span>,</span>
<span>               region_id <span class="op">=</span> <span class="st">"16_71020125_72901251"</span>,</span>
<span>               weights <span class="op">=</span> <span class="va">weights</span>,</span>
<span>               ens_db <span class="op">=</span> <span class="va">ens_db</span>,</span>
<span>               locus_range <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">71.6e6</span>,<span class="fl">72.4e6</span><span class="op">)</span>,</span>
<span>               highlight_pip <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span>               legend.position <span class="op">=</span> <span class="st">"top"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 2024-06-28 18:59:25 INFO::focus gene: HPR</span></span>
<span><span class="co">## 2024-06-28 18:59:25 INFO::focus id: ENSG00000261701.6|expression_Liver</span></span>
<span><span class="co">## 2024-06-28 18:59:25 INFO::plot locus range: chr16 71600000,72400000</span></span>
<span><span class="co">## 2024-06-28 18:59:25 INFO::HPR liver expression QTLs</span></span>
<span><span class="co">## 2024-06-28 18:59:25 INFO::QTL positions: 72063820,72063928</span></span></code></pre>
<div class="figure" style="text-align: center">
<img src="simple_ctwas_tutorial_files/figure-html/locus_plot-1.png" alt="&amp;nbsp;" width="1200"><p class="caption">
 
</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Siming Zhao, Wesley Crouse, Sheng Qian, Kaixuan Luo, Matthew Stephens, Xin He.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
