<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Summarizing cTWAS results â€¢ ctwas</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Summarizing cTWAS results">
<meta property="og:description" content="ctwas">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ctwas</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.2.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorial
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/simple_ctwas_tutorial.html">Simple cTWAS tutorial</a>
    </li>
    <li>
      <a href="../articles/preparing_ctwas_input_data.html">Preparing cTWAS input data</a>
    </li>
    <li>
      <a href="../articles/ctwas_main_function.html">Using cTWAS main function</a>
    </li>
    <li>
      <a href="../articles/ctwas_modules.html">Using cTWAS modules</a>
    </li>
    <li>
      <a href="../articles/summarizing_ctwas_results.html">Summarizing cTWAS results</a>
    </li>
    <li>
      <a href="../articles/postprocessing_ctwas_results.html">Post-processing cTWAS results</a>
    </li>
    <li>
      <a href="../articles/FAQ.html">FAQ</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/xinhe-lab/ctwas/tree/multigroup_test" class="external-link">Source</a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="summarizing_ctwas_results_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Summarizing cTWAS results</h1>
                        <h4 data-toc-skip class="author">Kaixuan Luo, Sheng Qian</h4>
            
            <h4 data-toc-skip class="date">2024-05-30</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/xinhe-lab/ctwas/blob/multigroup_test/vignettes/summarizing_ctwas_results.Rmd" class="external-link"><code>vignettes/summarizing_ctwas_results.Rmd</code></a></small>
      <div class="hidden name"><code>summarizing_ctwas_results.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="summarizing-and-visualizing-ctwas-results">Summarizing and visualizing cTWAS results<a class="anchor" aria-label="anchor" href="#summarizing-and-visualizing-ctwas-results"></a>
</h2>
<p>Load the <code>ctwas</code> package. We also need packages <code>tidyverse</code>, <code>locuszoomr</code> for summarizing and visualizing cTWAS result</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/xinhe-lab/ctwas" class="external-link">ctwas</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/myles-lewis/locuszoomr" class="external-link">locuszoomr</a></span><span class="op">)</span></span></code></pre></div>
<p>We also need gene annotations from the Ensembl database. We use <code>EnsDb.Hsapiens.v86</code> for the example data, please choose the Ensembl database for your specific data.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">EnsDb.Hsapiens.v86</span><span class="op">)</span></span></code></pre></div>
<p>Load the cTWAS results from the <code><a href="../reference/ctwas_sumstats.html">ctwas_sumstats()</a></code> function</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># directory with computed correlation matrices</span></span>
<span><span class="va">cor_dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/sample_data"</span>, <span class="st">"cor_matrix"</span>, package <span class="op">=</span> <span class="st">"ctwas"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># GWAS sample size</span></span>
<span><span class="va">gwas_n</span> <span class="op">&lt;-</span> <span class="fl">343621</span></span>
<span></span>
<span><span class="co"># Load region_info, LD_info and snp_info</span></span>
<span><span class="va">region_info</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/sample_data"</span>, <span class="st">"LDL_example.region_info.RDS"</span>, package <span class="op">=</span> <span class="st">"ctwas"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">LD_info</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/sample_data"</span>, <span class="st">"LDL_example.LD_info.RDS"</span>, package <span class="op">=</span> <span class="st">"ctwas"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">snp_info</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/sample_data"</span>, <span class="st">"LDL_example.snp_info.RDS"</span>, package <span class="op">=</span> <span class="st">"ctwas"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Load z_snp</span></span>
<span><span class="va">z_snp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/sample_data"</span>, <span class="st">"LDL_example.preprocessed.z_snp.RDS"</span>, package <span class="op">=</span> <span class="st">"ctwas"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Load weights</span></span>
<span><span class="va">weights</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/sample_data"</span>, <span class="st">"LDL_example.preprocessed.weights.RDS"</span>, package <span class="op">=</span> <span class="st">"ctwas"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Load the cTWAS results</span></span>
<span><span class="va">param</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/sample_data"</span>, <span class="st">"LDL_example.param.RDS"</span>, package <span class="op">=</span> <span class="st">"ctwas"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">finemap_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/sample_data"</span>, <span class="st">"LDL_example.finemap_res.RDS"</span>, package <span class="op">=</span> <span class="st">"ctwas"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">screened_region_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/sample_data"</span>, <span class="st">"LDL_example.screened_region_data.RDS"</span>, package <span class="op">=</span> <span class="st">"ctwas"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="assessing-parameters-and-computing-pve">Assessing parameters and computing PVE<a class="anchor" aria-label="anchor" href="#assessing-parameters-and-computing-pve"></a>
</h3>
<p>We could use the <code><a href="../reference/summarize_param.html">summarize_param()</a></code> function to assess the convergence of the estimated parameters and to compute the proportion of variance explained (PVE) by variants and genes.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ctwas_parameters</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/summarize_param.html">summarize_param</a></span><span class="op">(</span><span class="va">param</span>, <span class="va">gwas_n</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Estimated prior inclusion probability</span></span>
<span><span class="va">ctwas_parameters</span><span class="op">$</span><span class="va">group_prior</span></span></code></pre></div>
<pre><code><span><span class="co">##   expression|liver expression|adipose                SNP </span></span>
<span><span class="co">##       0.0112258728       0.0032677115       0.0002253614</span></span></code></pre>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Estimated prior effect size:</span></span>
<span><span class="va">ctwas_parameters</span><span class="op">$</span><span class="va">group_prior_var</span></span></code></pre></div>
<pre><code><span><span class="co">##   expression|liver expression|adipose                SNP </span></span>
<span><span class="co">##         159.366028         159.366028           4.889058</span></span></code></pre>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Estimated enrichment of molecular traits over variants:</span></span>
<span><span class="va">ctwas_parameters</span><span class="op">$</span><span class="va">enrichment</span></span></code></pre></div>
<pre><code><span><span class="co">##   expression|liver expression|adipose </span></span>
<span><span class="co">##           49.81275           14.49987</span></span></code></pre>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># PVE explained by molecular traits and variants:</span></span>
<span><span class="va">ctwas_parameters</span><span class="op">$</span><span class="va">group_pve</span></span></code></pre></div>
<pre><code><span><span class="co">##   expression|liver expression|adipose                SNP </span></span>
<span><span class="co">##       0.0021346173       0.0007335085       0.0005839915</span></span></code></pre>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Total heritability (sum of PVE)</span></span>
<span><span class="va">ctwas_parameters</span><span class="op">$</span><span class="va">total_pve</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 0.003452117</span></span></code></pre>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Attributable heritability</span></span>
<span><span class="va">ctwas_parameters</span><span class="op">$</span><span class="va">attributable_pve</span></span></code></pre></div>
<pre><code><span><span class="co">##   expression|liver expression|adipose                SNP </span></span>
<span><span class="co">##          0.6183502          0.2124807          0.1691691</span></span></code></pre>
<p>Plot convergence</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ctwas_parameters</span><span class="op">$</span><span class="va">convergence_plot</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="summarizing_ctwas_results_files/figure-html/convergence_plot-1.png" alt="&amp;nbsp;" width="840"><p class="caption">
Â 
</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="summarizing-ctwas-results">Summarizing cTWAS results<a class="anchor" aria-label="anchor" href="#summarizing-ctwas-results"></a>
</h2>
<p>To interpret and visualize the results, we need to add gene names to the finemapping result, and use mid-points to represent gene positions.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># load gene info database</span></span>
<span><span class="va">ens_db</span> <span class="op">&lt;-</span> <span class="va">EnsDb.Hsapiens.v86</span></span>
<span></span>
<span><span class="co"># add gene names and use gene mid-points to represent gene positions</span></span>
<span><span class="va">finemap_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/anno_finemap_res.html">anno_finemap_res</a></span><span class="op">(</span><span class="va">finemap_res</span>, <span class="va">snp_info</span>, <span class="va">ens_db</span>, use_gene_pos <span class="op">=</span> <span class="st">"mid"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="interpreting-ctwas-results">Interpreting cTWAS results<a class="anchor" aria-label="anchor" href="#interpreting-ctwas-results"></a>
</h2>
<p>We list all molecular traits with PIP &gt; 0.8, which is the threshold we used in the paper.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># select PIP &gt; 0.8</span></span>
<span><span class="va">sig_finemap_res</span> <span class="op">&lt;-</span> <span class="va">finemap_res</span><span class="op">[</span><span class="va">finemap_res</span><span class="op">$</span><span class="va">type</span><span class="op">!=</span><span class="st">"SNP"</span> <span class="op">&amp;</span> <span class="va">finemap_res</span><span class="op">$</span><span class="va">susie_pip</span> <span class="op">&gt;</span> <span class="fl">0.8</span>,<span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">sig_finemap_res</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html" class="external-link">order</a></span><span class="op">(</span><span class="op">-</span><span class="va">sig_finemap_res</span><span class="op">$</span><span class="va">susie_pip</span><span class="op">)</span>,<span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##       chrom      pos                                                 id</span></span>
<span><span class="co">## 13052    16 72070235                 ENSG00000261701.6|expression_Liver</span></span>
<span><span class="co">## 13065    16 72057824  ENSG00000257017.8|expression_Adipose_Subcutaneous</span></span>
<span><span class="co">## 7750     16 56972848 ENSG00000087237.10|expression_Adipose_Subcutaneous</span></span>
<span><span class="co">##             type context              group          z            region_id</span></span>
<span><span class="co">## 13052 expression   liver   expression|liver -18.403150 16:71020125-72901251</span></span>
<span><span class="co">## 13065 expression adipose expression|adipose  -4.810203 16:71020125-72901251</span></span>
<span><span class="co">## 7750  expression adipose expression|adipose  13.749840 16:55869862-57630418</span></span>
<span><span class="co">##       susie_pip      mu2 cs_index         gene_id gene_name</span></span>
<span><span class="co">## 13052 1.0000000 474.3442        1 ENSG00000261701       HPR</span></span>
<span><span class="co">## 13065 1.0000000 420.6018        3 ENSG00000257017        HP</span></span>
<span><span class="co">## 7750  0.9999615 149.3997        1 ENSG00000087237      CETP</span></span></code></pre>
<p>When we have multiple contexts (tissues, conditions), it is useful to evaluate the combined PIP for each molecular trait over different contexts:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># combine PIPs across contexts (tissues)</span></span>
<span><span class="va">combined_gene_pips</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sum_pip_across_contexts.html">sum_pip_across_contexts</a></span><span class="op">(</span><span class="va">finemap_res</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># select combined PIP &gt; 0.8</span></span>
<span><span class="va">sig_gene_pips_df</span> <span class="op">&lt;-</span> <span class="va">combined_gene_pips</span><span class="op">[</span><span class="va">combined_gene_pips</span><span class="op">$</span><span class="va">combined_pip</span> <span class="op">&gt;</span> <span class="fl">0.8</span>, <span class="op">]</span></span>
<span></span>
<span><span class="co"># list genes with combined PIP &gt; 0.8 or top 20 genes with highest combined PIPs</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">combined_gene_pips</span>, <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">sig_gene_pips_df</span><span class="op">)</span>, <span class="fl">20</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##             gene_id gene_name combined_pip    liver_pip  adipose_pip</span></span>
<span><span class="co">## 12  ENSG00000087237      CETP   1.02577519 2.581369e-02 9.999615e-01</span></span>
<span><span class="co">## 151 ENSG00000261701       HPR   1.00000126 1.000000e+00 1.262840e-06</span></span>
<span><span class="co">## 150 ENSG00000257017        HP   0.99999999           NA 1.000000e+00</span></span>
<span><span class="co">## 77  ENSG00000153786    ZDHHC7   0.36848818 6.262351e-03 3.622258e-01</span></span>
<span><span class="co">## 78  ENSG00000156853    ZNF689   0.10617886 1.061789e-01           NA</span></span>
<span><span class="co">## 45  ENSG00000103507     BCKDK   0.09254458 9.254458e-02           NA</span></span>
<span><span class="co">## 62  ENSG00000140832  MARVELD3   0.08883893 6.523017e-05 8.877370e-02</span></span>
<span><span class="co">## 23  ENSG00000099381    SETD1A   0.07480232 5.771762e-02 1.708469e-02</span></span>
<span><span class="co">## 69  ENSG00000140990   NDUFB10   0.07115304 7.115304e-02           NA</span></span>
<span><span class="co">## 79  ENSG00000156873     PHKG2   0.06487829 6.487829e-02           NA</span></span>
<span><span class="co">## 99  ENSG00000167969      ECI1   0.06052240 5.710587e-02 3.416526e-03</span></span>
<span><span class="co">## 119 ENSG00000178226    PRSS36   0.05383430 5.208019e-02 1.754102e-03</span></span>
<span><span class="co">## 110 ENSG00000169957    ZNF768   0.05084518 2.191066e-02 2.893452e-02</span></span>
<span><span class="co">## 93  ENSG00000167397    VKORC1   0.05009503           NA 5.009503e-02</span></span>
<span><span class="co">## 96  ENSG00000167965     MLST8   0.04862082 1.752138e-02 3.109944e-02</span></span>
<span><span class="co">## 22  ENSG00000099365     STX1B   0.04736620 3.668845e-02 1.067775e-02</span></span>
<span><span class="co">## 73  ENSG00000149927     DOC2A   0.04235676 4.060820e-02 1.748560e-03</span></span>
<span><span class="co">## 94  ENSG00000167962    ZNF598   0.04092684 3.899191e-02 1.934930e-03</span></span>
<span><span class="co">## 70  ENSG00000149922      TBX6   0.03967361 3.566991e-02 4.003701e-03</span></span>
<span><span class="co">## 92  ENSG00000167395    ZNF646   0.03097746 2.328356e-02 7.693900e-03</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="visualizing-ctwas-results">Visualizing cTWAS results<a class="anchor" aria-label="anchor" href="#visualizing-ctwas-results"></a>
</h2>
<p>We could make locus plots for regions of interest. We could zoom in the region by specifying the <code>locus_range</code>.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/make_locusplot.html">make_locusplot</a></span><span class="op">(</span><span class="va">finemap_res</span>,</span>
<span>               region_id <span class="op">=</span> <span class="st">"16:71020125-72901251"</span>,</span>
<span>               weights <span class="op">=</span> <span class="va">weights</span>,</span>
<span>               ens_db <span class="op">=</span> <span class="va">ens_db</span>, </span>
<span>               use_LD <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>               cor_dir <span class="op">=</span> <span class="va">cor_dir</span>,</span>
<span>               pip_thresh <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span>               locus_range <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">71.6e6</span>,<span class="fl">72.4e6</span><span class="op">)</span>,</span>
<span>               gene_biotype<span class="op">=</span><span class="st">"protein_coding"</span>,</span>
<span>               legend.position <span class="op">=</span> <span class="st">"right"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 2024-05-30 20:27:45 INFO::focus gene: HPR</span></span>
<span><span class="co">## 2024-05-30 20:27:45 INFO::focus id: ENSG00000261701.6|expression_Liver</span></span>
<span><span class="co">## 2024-05-30 20:27:45 INFO::locus_range: 71600000,72400000</span></span>
<span><span class="co">## 2024-05-30 20:27:45 INFO::Load correlation matrices for region 16:71020125-72901251</span></span>
<span><span class="co">## 2024-05-30 20:27:47 INFO::QTL positions: 72063820,72063928</span></span></code></pre>
<div class="figure" style="text-align: center">
<img src="summarizing_ctwas_results_files/figure-html/locus_plot-1.png" alt="&amp;nbsp;" width="1200"><p class="caption">
Â 
</p>
</div>
<p>If we do not have precomputed correlation matrices, we could compute those by specifying <code>region_data</code>, <code>LD_info</code> and <code>snp_info</code>. Or we could skip coloring correlations by setting <code>use_LD = FALSE</code>.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Siming Zhao, Wesley Crouse, Sheng Qian, Kaixuan Luo, Matthew Stephens, Xin He.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
