<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Causal inference for TWAS — ctwas • ctwas</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="Causal inference for TWAS — ctwas" />
<meta property="og:description" content="Causal inference for TWAS" />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ctwas</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.9</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/ctwas_idv_wkfl.html">Run on individual-level data</a>
    </li>
    <li>
      <a href="../articles/ctwas_rss_wkfl.html">Run on summary statistics data</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Causal inference for TWAS</h1>
    
    <div class="hidden name"><code>ctwas.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Causal inference for TWAS</p>
    </div>

    <pre class="usage"><span class='fu'>ctwas</span><span class='op'>(</span><span class='va'>pgenfs</span>, <span class='va'>exprfs</span>, <span class='va'>Y</span>, ld_regions <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/c.html'>c</a></span><span class='op'>(</span><span class='st'>"EUR"</span>, <span class='st'>"ASN"</span>, <span class='st'>"AFR"</span><span class='op'>)</span>,
  ld_regions_custom <span class='op'>=</span> <span class='cn'>NULL</span>, thin <span class='op'>=</span> <span class='fl'>1</span>, prob_single <span class='op'>=</span> <span class='fl'>0.8</span>,
  niter1 <span class='op'>=</span> <span class='fl'>3</span>, niter2 <span class='op'>=</span> <span class='fl'>30</span>, L <span class='op'>=</span> <span class='fl'>5</span>, group_prior <span class='op'>=</span> <span class='cn'>NULL</span>,
  group_prior_var <span class='op'>=</span> <span class='cn'>NULL</span>, estimate_group_prior <span class='op'>=</span> <span class='cn'>T</span>,
  estimate_group_prior_var <span class='op'>=</span> <span class='cn'>T</span>, use_null_weight <span class='op'>=</span> <span class='cn'>T</span>, coverage <span class='op'>=</span> <span class='fl'>0.95</span>,
  stardardize <span class='op'>=</span> <span class='cn'>T</span>, ncore <span class='op'>=</span> <span class='fl'>1</span>, outputdir <span class='op'>=</span> <span class='fu'><a href='https://rdrr.io/r/base/getwd.html'>getwd</a></span><span class='op'>(</span><span class='op'>)</span>, outname <span class='op'>=</span> <span class='cn'>NULL</span>,
  logfile <span class='op'>=</span> <span class='cn'>NULL</span><span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>pgenfs</th>
      <td><p>A character vector of .pgen or .bed files. One file for one
chromosome, in the order of 1 to 22. Therefore, the length of this vector
needs to be 22. If .pgen files are given, then .pvar and .psam are assumed
to present in the same directory. If .bed files are given, then .bim and
.fam files are assumed to present in the same directory.</p></td>
    </tr>
    <tr>
      <th>exprfs</th>
      <td><p>A character vector of .`expr` or `.expr.gz` files. One file for
one chromosome, in the order of 1 to 22. Therefore, the length of this vector
needs to be 22.  `.expr.gz` file is gzip compressed `.expr` files. `.expr` is
a matrix of imputed expression values, row is for each sample, column is for
each gene. Its sample order is same as in files provided by `.pgenfs`. We also
assume corresponding `.exprvar` files are present in the same directory.
`.exprvar` files are just tab delimited text files, with columns:</p><dl>
  <dt>chrom</dt><dd><p>chromosome number, numeric</p></dd>
  <dt>p0</dt><dd><p>gene boundary position, the smaller value</p></dd>
  <dt>p1</dt><dd><p>gene boundary position, the larger value</p></dd>
  <dt>id</dt><dd><p>gene id</p></dd>

</dl><p>Its rows should be in the same order as the columns for corresponding `.expr`
files.</p></td>
    </tr>
    <tr>
      <th>Y</th>
      <td><p>a vector of length n, phenotype, the same order as provided
by `.pgenfs` (defined in .psam or .fam files).</p></td>
    </tr>
    <tr>
      <th>ld_regions</th>
      <td><p>A string representing the population to use for defining
LD regions. These LD regions are defined by ldetect. The user can also
provide custom LD regions matching genotype data, see
<code>ld_regions_custom</code>.</p></td>
    </tr>
    <tr>
      <th>ld_regions_custom</th>
      <td><p>A bed format file defining LD regions. The default
is <code>NULL</code>; when specified, <code>ld_regions</code> will be ignored.</p></td>
    </tr>
    <tr>
      <th>prob_single</th>
      <td><p>blocks with probility more than this value will be
used for parameter estimation</p></td>
    </tr>
    <tr>
      <th>outname</th>
      <td><p>a string, the output name</p></td>
    </tr>
    <tr>
      <th>logfile</th>
      <td><p>the log file, if NULL will print log info on screen</p></td>
    </tr>
    </table>


  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Siming Zhao.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


