<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Estimate cTWAS parameters — est_param • ctwas</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Estimate cTWAS parameters — est_param"><meta property="og:description" content="Estimate cTWAS parameters"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ctwas</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.36</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorial
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/multigroup_ctwas_pipeline.html">Pipepline of using multigroup cTWAS with summary statistics</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/xinhe-lab/ctwas/tree/multigroup_test" class="external-link">Source</a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Estimate cTWAS parameters</h1>
    <small class="dont-index">Source: <a href="https://github.com/xinhe-lab/ctwas/blob/multigroup_test/R/ctwas_est_parameters.R" class="external-link"><code>R/ctwas_est_parameters.R</code></a></small>
    <div class="hidden name"><code>est_param.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Estimate cTWAS parameters</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">est_param</span><span class="op">(</span></span>
<span>  <span class="va">z_snp</span>,</span>
<span>  <span class="va">region_info</span>,</span>
<span>  regionlist <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  z_gene <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  gene_info <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  weights <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  thin <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  prob_single <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span>  niter1 <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  niter2 <span class="op">=</span> <span class="fl">30</span>,</span>
<span>  group_prior <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  group_prior_var <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  group_prior_var_structure <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"independent"</span>, <span class="st">"shared_all"</span>, <span class="st">"shared+snps"</span>,</span>
<span>    <span class="st">"shared_QTLtype"</span><span class="op">)</span>,</span>
<span>  use_null_weight <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  coverage <span class="op">=</span> <span class="fl">0.95</span>,</span>
<span>  min_abs_corr <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  ncore <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  logfile <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>z_snp</dt>
<dd><p>a data frame with four columns: "id", "A1", "A2", "z".
giving the z scores for SNPs. "A1" is effect allele. "A2" is the other allele.</p></dd>


<dt>region_info</dt>
<dd><p>a data frame of region definition and associated file names</p></dd>


<dt>regionlist</dt>
<dd><p>a list object indexing regions, variants and genes.</p></dd>


<dt>z_gene</dt>
<dd><p>a data frame with two columns: "id", "z". giving the z scores for genes.
Optionally, a "type" column can also be supplied; this is for using multiple sets of weights</p></dd>


<dt>gene_info</dt>
<dd><p>a data frame of gene information obtained from <code>compute_gene_z</code></p></dd>


<dt>thin</dt>
<dd><p>The proportion of SNPs to be used for the parameter estimation and
initial screening region steps.
Smaller <code>thin</code> parameters reduce runtime at the expense of accuracy.
The fine mapping step is rerun using full SNPs for regions with strong gene signals.</p></dd>


<dt>prob_single</dt>
<dd><p>Blocks with probability greater than <code>prob_single</code> of
having 1 or fewer effects will be used for parameter estimation</p></dd>


<dt>niter1</dt>
<dd><p>the number of iterations of the E-M algorithm to perform during the initial parameter estimation step</p></dd>


<dt>niter2</dt>
<dd><p>the number of iterations of the E-M algorithm to perform during the complete parameter estimation step</p></dd>


<dt>group_prior</dt>
<dd><p>a vector of two prior inclusion probabilities for SNPs and genes.</p></dd>


<dt>group_prior_var</dt>
<dd><p>a vector of two prior variances for SNPs and gene effects.</p></dd>


<dt>group_prior_var_structure</dt>
<dd><p>a string indicating the structure to put on the prior variance parameters.
"independent" is the default and allows all groups to have their own separate variance parameters.
"shared" allows all groups to share the same variance parameter.
"shared+snps" allows all groups to share the same variance parameter, and this variance parameter is also shared with SNPs.
"shared_type" allows all groups in one molecular QTL type to share the same variance parameter.</p></dd>


<dt>use_null_weight</dt>
<dd><p>TRUE/FALSE. If TRUE, allow for a probability of no effect in susie</p></dd>


<dt>coverage</dt>
<dd><p>A number between 0 and 1 specifying the “coverage” of the estimated confidence sets</p></dd>


<dt>min_abs_corr</dt>
<dd><p>Minimum absolute correlation allowed in a
credible set. The default, 0.5, corresponds to a squared
correlation of 0.25, which is a commonly used threshold for
genotype data in genetic studies.</p></dd>


<dt>ncore</dt>
<dd><p>The number of cores used to parallelize computation over regions</p></dd>


<dt>logfile</dt>
<dd><p>the log file, if NULL will print log info on screen</p></dd>


<dt>weight_list</dt>
<dd><p>a list of weights by chromosome</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>a list with estimated parameters, and updated region_info with correlation file names for each region.</p>
    </div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Siming Zhao, Wesley Crouse, Sheng Qian, Kaixuan Luo, Matthew Stephens, Xin He.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

