<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Function reference â€¢ ctwas</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Function reference"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-index">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ctwas</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.2.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorial
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/multigroup_ctwas_pipeline.html">Pipepline of using multigroup cTWAS with summary statistics</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/xinhe-lab/ctwas/tree/multigroup_test" class="external-link">Source</a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index"><colgroup><col class="alias"><col class="title"></colgroup><tbody><tr><th colspan="2">
          <h2 id="all-functions">All functions <a href="#all-functions" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="EM.html">EM()</a></code> </p>
        </td>
        <td><p>Run EM to estimate parameters.
Iteratively run susie and estimate parameters - RSS version</p></td>
      </tr><tr><td>
          <p><code><a href="add_z_to_regionlist.html">add_z_to_regionlist()</a></code> </p>
        </td>
        <td><p>add z-scores from z_snp and z_gene to regionlist</p></td>
      </tr><tr><td>
          <p><code><a href="adjust_boundary_genes.html">adjust_boundary_genes()</a></code> </p>
        </td>
        <td><p>adjust regionlist for boundary genes</p></td>
      </tr><tr><td>
          <p><code><a href="anno_susie.html">anno_susie()</a></code> </p>
        </td>
        <td><p>annotate susie results with SNP and gene information</p></td>
      </tr><tr><td>
          <p><code><a href="assign_region_ids.html">assign_region_ids()</a></code> </p>
        </td>
        <td><p>assign gene and SNP IDs for regions in the regioninfo</p></td>
      </tr><tr><td>
          <p><code><a href="coef.susie.html">coef(<i>&lt;susie&gt;</i>)</a></code> </p>
        </td>
        <td><p>Extract regression coefficients from susie fit</p></td>
      </tr><tr><td>
          <p><code><a href="combine_z.html">combine_z()</a></code> </p>
        </td>
        <td><p>Combine SNP and gene z-scores</p></td>
      </tr><tr><td>
          <p><code><a href="compute_gene_z.html">compute_gene_z()</a></code> </p>
        </td>
        <td><p>Compute gene z-scores</p></td>
      </tr><tr><td>
          <p><code><a href="compute_region_cor.html">compute_region_cor()</a></code> </p>
        </td>
        <td><p>Compute correlation matrices for a region</p></td>
      </tr><tr><td>
          <p><code><a href="compute_ss.html">compute_ss()</a></code> </p>
        </td>
        <td><p>Compute sufficient statistics for susie.</p></td>
      </tr><tr><td>
          <p><code><a href="convert_geno_to_LD_matrix.html">convert_geno_to_LD_matrix()</a></code> </p>
        </td>
        <td><p>Convert PLINK genotype data to LD matrices and SNP info files,
save LD matrices as .RDS files and SNP info as .Rvar files</p></td>
      </tr><tr><td>
          <p><code><a href="ctwas_locus_plot.html">ctwas_locus_plot()</a></code> </p>
        </td>
        <td><p>Plot the output of cTWAS at a single locus</p></td>
      </tr><tr><td>
          <p><code><a href="ctwas_locus_plot_genetrack.html">ctwas_locus_plot_genetrack()</a></code> </p>
        </td>
        <td><p>Create the gene track for the cTWAS locus plot</p></td>
      </tr><tr><td>
          <p><code><a href="ctwas_sumstats.html">ctwas_sumstats()</a></code> </p>
        </td>
        <td><p>Causal inference for TWAS using summary statistics</p></td>
      </tr><tr><td>
          <p><code><a href="ctwas_susie_rss.html">ctwas_susie_rss()</a></code> </p>
        </td>
        <td><p>Run cTWAS version of susie_rss for a single region</p></td>
      </tr><tr><td>
          <p><code><a href="detect_ld_mismatch_susie.html">detect_ld_mismatch_susie()</a></code> </p>
        </td>
        <td><p>Detect LD mismatches using SuSiE RSS</p></td>
      </tr><tr><td>
          <p><code><a href="est_param.html">est_param()</a></code> </p>
        </td>
        <td><p>Estimate cTWAS parameters</p></td>
      </tr><tr><td>
          <p><code><a href="estimate_s_rss.html">estimate_s_rss()</a></code> </p>
        </td>
        <td><p>Estimate s in <code>susie_rss</code> Model Using Regularized LD</p></td>
      </tr><tr><td>
          <p><code><a href="expand_regionlist.html">expand_regionlist()</a></code> </p>
        </td>
        <td><p>Expand regionlist with full SNPs</p></td>
      </tr><tr><td>
          <p><code><a href="finemap_region.html">finemap_region()</a></code> </p>
        </td>
        <td><p>run cTWAS finemapping for a single region</p></td>
      </tr><tr><td>
          <p><code><a href="finemap_regions.html">finemap_regions()</a></code> </p>
        </td>
        <td><p>run cTWAS finemapping for multiple regions</p></td>
      </tr><tr><td>
          <p><code><a href="finemap_regions_L1.html">finemap_regions_L1()</a></code> </p>
        </td>
        <td><p>run cTWAS finemapping for multiple regions with L = 1</p></td>
      </tr><tr><td>
          <p><code><a href="get_gene_info.html">get_gene_info()</a></code> </p>
        </td>
        <td><p>get gene info from weights</p></td>
      </tr><tr><td>
          <p><code><a href="get_gene_regions.html">get_gene_regions()</a></code> </p>
        </td>
        <td><p>get regions for each gene</p></td>
      </tr><tr><td>
          <p><code><a href="get_problematic_regions.html">get_problematic_regions()</a></code> </p>
        </td>
        <td><p>Get regions with problematic high PIP SNPs or genes</p></td>
      </tr><tr><td>
          <p><code><a href="get_regionlist.html">get_regionlist()</a></code> </p>
        </td>
        <td><p>Get regionlist for each region in region_info</p></td>
      </tr><tr><td>
          <p><code><a href="harmonize_wgt_ld.html">harmonize_wgt_ld()</a></code> </p>
        </td>
        <td><p>Harmonize weight to match ld reference genotypes.
Flip signs when reverse complement matches, remove ambiguous variants from the prediction models</p></td>
      </tr><tr><td>
          <p><code><a href="harmonize_z_ld.html">harmonize_z_ld()</a></code> </p>
        </td>
        <td><p>Harmonize z scores from GWAS to match ld reference genotypes.
Flip signs when reverse complement matches, remove strand ambiguous SNPs</p></td>
      </tr><tr><td>
          <p><code><a href="initiate_group_priors.html">initiate_group_priors()</a></code> </p>
        </td>
        <td><p>set pi_prior and V_prior based on init_group_prior and init_group_prior_var</p></td>
      </tr><tr><td>
          <p><code><a href="kriging_rss.html">kriging_rss()</a></code> <code><a href="kriging_rss.html">kriging_rss()</a></code> </p>
        </td>
        <td><p>Compute Distribution of z-scores of Variant j Given Other z-scores, and Detect Possible Allele Switch Issue</p></td>
      </tr><tr><td>
          <p><code><a href="label_merged_regions.html">label_merged_regions()</a></code> </p>
        </td>
        <td><p>Function to identify consecutive regions and label them</p></td>
      </tr><tr><td>
          <p><code><a href="load_LD.html">load_LD()</a></code> </p>
        </td>
        <td><p>Load LD matrix</p></td>
      </tr><tr><td>
          <p><code><a href="load_weights.html">load_weights()</a></code> </p>
        </td>
        <td><p>Load PredictDB or Fusion weights</p></td>
      </tr><tr><td>
          <p><code><a href="predict.susie.html">predict(<i>&lt;susie&gt;</i>)</a></code> </p>
        </td>
        <td><p>Predict outcomes or extract coefficients from susie fit.</p></td>
      </tr><tr><td>
          <p><code><a href="prep_pgen.html">prep_pgen()</a></code> </p>
        </td>
        <td><p>Read .pgen file into R</p></td>
      </tr><tr><td>
          <p><code><a href="prep_pvar.html">prep_pvar()</a></code> </p>
        </td>
        <td><p>Prepare .pvar file</p></td>
      </tr><tr><td>
          <p><code><a href="preprocess_weights.html">preprocess_weights()</a></code> </p>
        </td>
        <td><p>Preprocess PredictDB/Fusion weights and harmonize with LD reference</p></td>
      </tr><tr><td>
          <p><code><a href="preprocess_z_snp.html">preprocess_z_snp()</a></code> </p>
        </td>
        <td><p>Preprocess GWAS z-scores, harmonize GWAS z-scores with LD reference</p></td>
      </tr><tr><td>
          <p><code><a href="read_LD_SNP_file.html">read_LD_SNP_file()</a></code> </p>
        </td>
        <td><p>read a single LD SNP info file as a data frame</p></td>
      </tr><tr><td>
          <p><code><a href="read_LD_SNP_files.html">read_LD_SNP_files()</a></code> </p>
        </td>
        <td><p>read all LD SNP info files as a data frame</p></td>
      </tr><tr><td>
          <p><code><a href="read_bim.html">read_bim()</a></code> </p>
        </td>
        <td><p>Read .bim file into R</p></td>
      </tr><tr><td>
          <p><code><a href="read_pgen.html">read_pgen()</a></code> </p>
        </td>
        <td><p>Read pgen file into R</p></td>
      </tr><tr><td>
          <p><code><a href="read_pvar.html">read_pvar()</a></code> </p>
        </td>
        <td><p>Read .pvar file into R</p></td>
      </tr><tr><td>
          <p><code><a href="read_var_info.html">read_var_info()</a></code> </p>
        </td>
        <td><p>Read variant information from .pvar or .bim file into R</p></td>
      </tr><tr><td>
          <p><code><a href="region2core.html">region2core()</a></code> </p>
        </td>
        <td><p>assign regions to cores</p></td>
      </tr><tr><td>
          <p><code><a href="screen_regions.html">screen_regions()</a></code> </p>
        </td>
        <td><p>Screen regions</p></td>
      </tr><tr><td>
          <p><code><a href="select_highPIP_regions.html">select_highPIP_regions()</a></code> </p>
        </td>
        <td><p>Select regions with high non-SNP PIPs</p></td>
      </tr><tr><td>
          <p><code><a href="select_single_effect_regions.html">select_single_effect_regions()</a></code> </p>
        </td>
        <td><p>Select single effect regions</p></td>
      </tr><tr><td>
          <p><code><a href="summarize_param.html">summarize_param()</a></code> </p>
        </td>
        <td><p>Summarize and plot cTWAS parameter estimates</p></td>
      </tr><tr><td>
          <p><code><a href="susie.html">susie()</a></code> <code><a href="susie.html">susie_suff_stat()</a></code> </p>
        </td>
        <td><p>Sum of Single Effects (SuSiE) Regression</p></td>
      </tr><tr><td>
          <p><code><a href="susie_get_methods.html">susie_get_objective()</a></code> <code><a href="susie_get_methods.html">susie_get_posterior_mean()</a></code> <code><a href="susie_get_methods.html">susie_get_posterior_sd()</a></code> <code><a href="susie_get_methods.html">susie_get_niter()</a></code> <code><a href="susie_get_methods.html">susie_get_prior_variance()</a></code> <code><a href="susie_get_methods.html">susie_get_residual_variance()</a></code> <code><a href="susie_get_methods.html">susie_get_lfsr()</a></code> <code><a href="susie_get_methods.html">susie_get_posterior_samples()</a></code> <code><a href="susie_get_methods.html">susie_get_cs()</a></code> <code><a href="susie_get_methods.html">susie_get_pip()</a></code> </p>
        </td>
        <td><p>Inferences From Fitted SuSiE Model</p></td>
      </tr><tr><td>
          <p><code><a href="susie_init_coef.html">susie_init_coef()</a></code> </p>
        </td>
        <td><p>Initialize a susie object using regression coefficients</p></td>
      </tr><tr><td>
          <p><code><a href="susie_rss.html">susie_rss()</a></code> </p>
        </td>
        <td><p>Sum of Single Effects (SuSiE) Regression using summary statistics</p></td>
      </tr><tr><td>
          <p><code><a href="trim_regionlist.html">trim_regionlist()</a></code> </p>
        </td>
        <td><p>Remove SNPs from regionlist if the total number of SNPs exceeds limit</p></td>
      </tr><tr><td>
          <p><code><a href="univariate_regression.html">univariate_regression()</a></code> </p>
        </td>
        <td><p>Perform Univariate Linear Regression Separately for Columns of X</p></td>
      </tr></tbody></table></div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Siming Zhao, Wesley Crouse, Sheng Qian, Kaixuan Luo, Matthew Stephens, Xin He.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

