<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Rerun cTWAS finemapping using summary statistics with L = 1 for problematic regions — rerun_finemap_regions_L1 • ctwas</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Rerun cTWAS finemapping using summary statistics with L = 1 for problematic regions — rerun_finemap_regions_L1"><meta property="og:description" content="Rerun cTWAS finemapping using summary statistics with L = 1 for problematic regions"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">ctwas</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.36</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Tutorial
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/multigroup_ctwas_pipeline.html">Pipepline of using multigroup cTWAS with summary statistics</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/xinhe-lab/ctwas/tree/multigroup_test" class="external-link">Source</a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Rerun cTWAS finemapping using summary statistics with L = 1 for problematic regions</h1>
    <small class="dont-index">Source: <a href="https://github.com/xinhe-lab/ctwas/blob/multigroup_test/R/ctwas_postprocess.R" class="external-link"><code>R/ctwas_postprocess.R</code></a></small>
    <div class="hidden name"><code>rerun_finemap_regions_L1.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Rerun cTWAS finemapping using summary statistics with L = 1 for problematic regions</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">rerun_finemap_regions_L1</span><span class="op">(</span></span>
<span>  <span class="va">z_snp</span>,</span>
<span>  <span class="va">z_gene</span>,</span>
<span>  <span class="va">weight</span>,</span>
<span>  problematic_snps <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  rerun_region_tags <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  pip_thresh <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  ctwas_outputdir <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  ctwas_outname <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  ld_R_dir <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  outputdir <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  outname <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  group_prior <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  group_prior_var <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  use_null_weight <span class="op">=</span> <span class="cn">T</span>,</span>
<span>  coverage <span class="op">=</span> <span class="fl">0.95</span>,</span>
<span>  min_abs_corr <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  ncore <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  logfile <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>z_snp</dt>
<dd><p>A data frame with four columns: "id", "A1", "A2", "z".
giving the z scores for snps. "A1" is effect allele. "A2" is the other allele.</p></dd>


<dt>z_gene</dt>
<dd><p>A data frame with two columns: "id", "z". giving the z scores for genes.
Optionally, a "type" column can also be supplied; this is for using multiple sets of weights</p></dd>


<dt>weight</dt>
<dd><p>a string, weight filename used in cTWAS</p></dd>


<dt>problematic_snps</dt>
<dd><p>a character vector of problematic SNP rsIDs</p></dd>


<dt>rerun_region_tags</dt>
<dd><p>a character vector of region tags to rerun</p></dd>


<dt>pip_thresh</dt>
<dd><p>Minimum PIP value to select regions</p></dd>


<dt>ctwas_outputdir</dt>
<dd><p>a string, the directory to cTWAS output</p></dd>


<dt>ctwas_outname</dt>
<dd><p>a string, the directory to cTWAS output name</p></dd>


<dt>outputdir</dt>
<dd><p>a string, the directory to store output</p></dd>


<dt>outname</dt>
<dd><p>a string, the output name</p></dd>


<dt>group_prior</dt>
<dd><p>a vector of two prior inclusion probabilities for SNPs and genes. This is ignored
if <code>estimate_group_prior = T</code></p></dd>


<dt>group_prior_var</dt>
<dd><p>a vector of two prior variances for SNPs and gene effects. This is ignored
if <code>estimate_group_prior_var = T</code></p></dd>


<dt>use_null_weight</dt>
<dd><p>TRUE/FALSE. If TRUE, allow for a probability of no effect in susie</p></dd>


<dt>coverage</dt>
<dd><p>A number between 0 and 1 specifying the “coverage” of the estimated confidence sets</p></dd>


<dt>min_abs_corr</dt>
<dd><p>Minimum absolute correlation allowed in a
  credible set. The default, 0.5, corresponds to a squared
  correlation of 0.25, which is a commonly used threshold for
  genotype data in genetic studies.
#'</p></dd>


<dt>ncore</dt>
<dd><p>The number of cores used to parallelize over regions</p></dd>


<dt>logfile</dt>
<dd><p>the log file, if NULL will print log info on screen</p></dd>


<dt>LD_R_dir</dt>
<dd><p>a string, pointing to a directory containing all LD matrix files and variant information.</p></dd>

</dl></div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Siming Zhao, Wesley Crouse, Sheng Qian, Kaixuan Luo, Matthew Stephens, Xin He.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

